---
title: "exRNA_disease_biomarkers_nb"
output: html_notebook
date: "`r sys.Date()`"
author:
- name: "Roman E. Reggiardo"
  affiliation: "UC Santa Cruz Dept. of Biomolecular Engineering"
  email: "rreggiar@ucsc.edu"
editor_options: 
  chunk_output_type: console
---
# nb_setup
```{r, setup, include=FALSE}
knitr::opts_chunk$set(fig.width = 8, collapse = T)
knitr::opts_chunk$set(message = F)
here::here()
suppressPackageStartupMessages(library(tidyverse))
```

# global_init
## ggplot_theme
```{r}
library(ggplot2)
library(ggthemes)
base_size = 7

ggplot2::theme_set(ggthemes::theme_foundation(base_size = base_size,
                                              base_family = 'Helvetica') + 
                     theme(plot.title = element_text(size = base_size, face = 'bold'),
                           panel.background = element_rect(colour = NA),
                           plot.background = element_rect(colour = NA),
                           panel.border = element_rect(colour = NA), 
                           axis.line = element_line(), 
                           axis.line.x = NULL, 
                           axis.line.y = NULL, 
                           axis.text = element_text(size = rel(0.95)), 
                           axis.text.x = element_text(margin = 
                                                        margin(t = 0.8 * base_size/4)), 
                           axis.text.x.top = element_text(margin = 
                                                            margin(b = 0.8 * base_size/4), 
                                                          vjust = 0), 
                           axis.text.y = element_text(margin = 
                                                        margin(r = 0.5 * base_size/4), 
                                                      hjust = 1),
                           axis.text.y.right = element_text(margin = 
                                                              margin(l = 0.5 * base_size/4), 
                                                            hjust = 0), 
                           axis.ticks = element_line(), 
                           axis.ticks.length = unit(base_size/2.5, "pt"), axis.ticks.length.x = NULL, 
                           axis.ticks.length.x.top = NULL, axis.ticks.length.x.bottom = NULL, 
                           axis.ticks.length.y = NULL, axis.ticks.length.y.left = NULL, 
                           axis.ticks.length.y.right = NULL,
                           strip.text = element_text(size = rel(0.8), face = 'bold'),
                           strip.background = element_rect(size = 0.5),
                           legend.key.size= unit(0.08, "in"),
                           legend.spacing = unit(0, "in"),
                           legend.key = element_rect(colour = NA),
                           legend.title = element_text(face="italic"),
                           legend.title.align=0.5,
                           legend.text = element_text(face = 'bold'),
                           legend.justification = c("right", "top"),
                           legend.background = element_rect(size = 0.25, colour = 'black'),
                           legend.box.just = "right",
                           legend.margin = margin(6, 6, 6, 6),
                           plot.tag = element_text(size = base_size, face = 'bold'),
                           plot.margin=margin(0.04,
                                              0.04,
                                              0.04,
                                              0.04,
                                              unit = "in"),
                           legend.box.spacing = unit(-0.02, 'in'),
                           panel.grid.major = element_blank(),
                           panel.grid.minor = element_blank()
                           )
                   )

theme_set_fun <- function(){
  
  ggplot2::theme_set(ggthemes::theme_foundation(base_size = base_size,
                                              base_family = 'Helvetica') + 
                     theme(plot.title = element_text(size = base_size, face = 'bold'),
                           panel.background = element_rect(colour = NA),
                           plot.background = element_rect(colour = NA),
                           panel.border = element_rect(colour = NA), 
                           axis.line = element_line(), 
                           axis.line.x = NULL, 
                           axis.line.y = NULL, 
                           axis.text = element_text(size = rel(0.95)), 
                           axis.text.x = element_text(margin = 
                                                        margin(t = 0.8 * base_size/4)), 
                           axis.text.x.top = element_text(margin = 
                                                            margin(b = 0.8 * base_size/4), 
                                                          vjust = 0), 
                           axis.text.y = element_text(margin = 
                                                        margin(r = 0.5 * base_size/4), 
                                                      hjust = 1),
                           axis.text.y.right = element_text(margin = 
                                                              margin(l = 0.5 * base_size/4), 
                                                            hjust = 0), 
                           axis.ticks = element_line(), 
                           axis.ticks.length = unit(base_size/2.5, "pt"), axis.ticks.length.x = NULL, 
                           axis.ticks.length.x.top = NULL, axis.ticks.length.x.bottom = NULL, 
                           axis.ticks.length.y = NULL, axis.ticks.length.y.left = NULL, 
                           axis.ticks.length.y.right = NULL,
                           strip.text = element_text(size = rel(0.8), face = 'bold'),
                           strip.background = element_rect(size = 0.5),
                           legend.key.size= unit(0.08, "in"),
                           legend.spacing = unit(0, "in"),
                           legend.key = element_rect(colour = NA),
                           legend.title = element_text(face="italic"),
                           legend.text = element_text(face = 'bold'),
                           legend.justification = c("right", "top"),
                           legend.background = element_rect(size = 0.25, colour = 'black'),
                           legend.box.just = "right",
                           legend.margin = margin(6, 6, 6, 6),
                           plot.tag = element_text(size = base_size, face = 'bold'),
                           plot.margin=margin(0.04,
                                              0.04,
                                              0.04,
                                              0.04,
                                              unit = "in"),
                           legend.box.spacing = unit(-0.02, 'in'),
                           panel.grid.major = element_blank(),
                           panel.grid.minor = element_blank()
                           )
                   )
  
}


txt.mm_to_pts <- function(ratio){ base_size * ratio * (25.4 / 72.27) }
```
## data_path_obj
```{r}
figure.dir <- here::here('manuscript/figures')
input_data.dir <- here::here('data/input_data')
output_data.dir <- here::here('data/output_data')
qc_data.dir <- here::here('data/output_data/rna_qc')
meta_data.dir <- here::here('data/meta_data')
reference.dir <- here::here('reference')
r_code.dir <- here::here('R')
```
## figure_export
```{r}
figure_single_column_width = unit(90, 'mm')
figure_double_column_width = unit(180, 'mm')
figure_depth = unit(170, 'mm')
panel_height_1 = unit(170/3, 'mm')

# install.packages("import")

import::from(.directory = r_code.dir, .from = 'helper_analysisFunctions.R', save.manuscript.panel)
```
## analysis_const
```{r}
chr_order <- c('chr1', 'chr2', 'chr3', 
               'chr4', 'chr5', 'chr6', 
               'chr7', 'chr8', 'chr9',
               'chr10', 'chr11', 'chr12', 
               'chr13', 'chr14', 'chr15', 
               'chr16', 'chr17',
               'chr18', 'chr19', 'chr20', 
               'chr21', 'chr22', 'chrX', 'chrY')


gencode_ver <- 39

three_color_pal <- RColorBrewer::brewer.pal(n = 3, 'Dark2')
eight_color_pal <- RColorBrewer::brewer.pal(n = 8, 'Dark2')
viridis_three_color_pal <- viridisLite::plasma(20)[c(1,9,16)]



orig_identity_color_pal <- c('ctrl' = viridis_three_color_pal[[1]],
                             'panc' = viridis_three_color_pal[[2]],
                             'covid' = viridis_three_color_pal[[3]])

disc_identity_color_pal <- c('normal' = viridis_three_color_pal[[1]],
                             'panc' = viridis_three_color_pal[[2]])

te_aware_annotation_levels <- c('GENCODE_coding',
                                 'GENCODE_lncRNA',
                                 'LINE',
                                 'SINE',
                                 'LTR',
                                 'DNA',
                                 'Simple_repeat',
                                 'other')

te_color_pal <- c('grey', RColorBrewer::brewer.pal(n = 7, 'Dark2'))
names(te_color_pal) <- te_aware_annotation_levels

te_alone_annotation_levels <- c('LINE',
                                'SINE',
                                'LTR',
                                'DNA',
                                'Simple_repeat',
                                'rRNA',
                                'tRNA')
te_alone_color_pal <- RColorBrewer::brewer.pal(n = 8, 'Dark2')[2:8]
names(te_alone_color_pal) <- te_alone_annotation_levels

import::from(.from = 'SummarizedExperiment',
             rowRanges)
disc_identity_shape_pal <- c('Male' = 23, 'Female' = 25)


```

# meta_data_construction
## reference meta data
```{r}

## reference meta data
if (!file.exists(file.path(meta_data.dir, 'reference_meta_data_df.rds'))) {
  reference_meta_data <- tibble::lst()
  
  reference_meta_data$ucsc_rmsk_insert_info.df <-
    readr::read_tsv(
      file.path(reference.dir, 'ucsc.rmsk.insert.info.txt'),
      col_names = c('gene', 'clade', 'family')
    )
  
  reference_meta_data$gencode_tx_to_gene.df <-
    readr::read_csv(file.path(
      reference.dir,
      paste0('gencode.v', gencode_ver, '.tx.to.gene.csv')
    ),
    col_names = c('tx', 'gene')) %>%
    tidyr::separate(
      tx,
      sep = '\\|',
      into = c('enst', 'ensg', NA, NA, 'tx', 'gene', 'len', 'biotype')
    )
  
  reference_meta_data$gencode_tx_to_gene.df %>%
    dplyr::select(-enst) %>%
    dplyr::group_by(biotype) %>%
    dplyr::tally() -> reference_meta_data$gencode_biotype_count.df
  
  reference_meta_data$gencode_tx_to_gene.df %>%
    dplyr::select(ensg, biotype) %>%
    dplyr::group_by(ensg) %>%
    dplyr::filter(all(biotype == 'lncRNA')) %>%
    dplyr::distinct() -> reference_meta_data$gencode_lncRNA_gene_names.df
  
  reference_meta_data$ucsc_rmsk_tx_to_gene.df <-
    readr::read_csv(
      file.path(reference.dir,
                'ucsc.rmsk.insert.analysis.tx.to.gene.csv'),
      col_names = c('insert', 'info', 'gene', 'len', 'biotype')
    ) %>%
    mutate(enst = insert, ensg = gene, tx = insert) %>%
    select(enst, ensg, tx, gene, len, biotype,-insert,-info)
  
  reference_meta_data$total_tx_to_gene.df <-
    rbind(
      reference_meta_data$gencode_tx_to_gene.df,
      reference_meta_data$ucsc_rmsk_tx_to_gene.df
    )
  
  saveRDS(reference_meta_data,
          file.path(meta_data.dir, 'reference_meta_data_df.rds'))
  
} else {
  reference_meta_data <-
    readRDS(file.path(meta_data.dir, 'reference_meta_data_df.rds'))
}
```
## original sample metadata
```{r}

## sample_meta_data
if (!file.exists(file.path(meta_data.dir, 'sample_meta_data_df.rds'))) {
  list.files(meta_data.dir, pattern = '*_meta_data.csv') -> meta_file.list
  names(meta_file.list) <- meta_file.list
  
  meta_file.list %>%
    purrr::imap(function(name, file) {
      readr::read_csv(file.path(meta_data.dir, file))
    }) -> sample_meta_data
  
  saveRDS(sample_meta_data,
          file.path(meta_data.dir, 'sample_meta_data_df.rds'))
  
} else {
  sample_meta_data <-
    readRDS(file.path(meta_data.dir, 'sample_meta_data_df.rds'))
}

original_roster_names <- read_csv(
  file.path(meta_data.dir,
            "original_roster_data_2020-09-30.csv"),
  col_names = c('sample', 'name')
  ) %>%
  mutate(sample = str_split_fixed(sample, '_S', 2)[, 1])

original_panc_staged_data <-
  read_csv(file.path(meta_data.dir, 'bioIVT_panc_staged_metadata_2023-03-06.csv'))


original_panc_meta_data <-
  sample_meta_data$`2022.01.14_bioIVTAndPitts_meta_data.csv` %>% 
  filter(condition == 'panc') %>% 
  dplyr::rename(sample = patient, diagnosis = condition) %>% 
  merge(original_panc_staged_data %>% select(sample, stage), by ='sample') %>% 
  mutate(tmp = sample, dataset = 'pilot') %>% 
  mutate(input_vol = as.numeric(
      str_split_fixed(
      string = str_split_fixed(string = tmp, pattern = '[.]', n = 2)[,2],
      pattern = '[.]',
      n = 2)[,2]
    ) * 1000) # convert to uL

original_covid_meta_data <- 
  sample_meta_data$`2022.01.14_bioIVTCovid_meta_data.csv` %>% 
  dplyr::rename(sample = patient) %>% 
  mutate(tmp = sample, diagnosis = 'covid', dataset = 'pilot') %>% 
  mutate(input_vol = as.numeric(
      str_split_fixed(
      string = str_split_fixed(string = tmp, pattern = '[.]', n = 2)[,2],
      pattern = '[.]',
      n = 2)[,2]
    ) * 1000) #convert to uL


original_meta_data <- 
  sample_meta_data$`2022.01.14_combined_meta_data.csv` %>% 
  dplyr::rename(sample = patient, diagnosis = condition, gender = sex) %>% 
  mutate(tmp = sample, dataset = 'pilot') %>% 
  mutate(input_vol = as.numeric(
      str_split_fixed(
      string = str_split_fixed(string = tmp, pattern = '[.]', n = 2)[,2],
      pattern = '[.]',
      n = 2)[,2]
    ) * 1000)

```
## discovery cohort metadata
```{r}
discovery_meta_data <- read_csv(file.path(
  meta_data.dir,
  'discovery_meta_data_w_inputVol_2023-02-10.csv'
)) %>% 
  mutate(dataset = 'discovery')

colnames(discovery_meta_data) <-
  str_remove(str_remove_all(str_remove_all(colnames(
    discovery_meta_data
  ), pattern = 'sample_sku_'), pattern = 'patient_'),
  pattern = '_at_collection')
```
## external cohorts metadata
```{r}

gut_meta_data <-
  read_csv(file.path(meta_data.dir, 'ext_meta_data', 'GSE133684_meta.txt')) %>%
  select(sample = Run,
         diagnosis = disease_state) %>% 
  mutate(input_vol = 1000, dataset = 'gut')

elife_meta_data <-
  read_csv(file.path(meta_data.dir, 'ext_meta_data', 'GSE174302_meta.txt')) %>% 
  select(
    sample = Run,
    age = Age,
    diagnosis = disease,
    gender
  ) %>%
  mutate(diagnosis = str_replace_all(
    string = diagnosis,
    pattern = ' patient',
    replacement = ''
  )) %>% 
  mutate(input_vol = 1000, dataset = 'elife')

npj_meta_data <- 
  read_csv(file.path(meta_data.dir, 'ext_meta_data', 'GSE182824_meta.txt')) %>% 
  select(
    sample = Run,
    age = Age,
    cohort = Cohort,
    diagnosis = source_name,
    gender
  ) %>%
  mutate(diagnosis = str_replace_all(
    string = diagnosis,
    pattern = ' plasma',
    replacement = ''
  )) %>% 
  mutate(diagnosis = str_replace_all(
    string = diagnosis,
    pattern = 'Human ',
    replacement = ''
  )) %>% 
  mutate(diagnosis = str_replace_all(
    string = diagnosis,
    pattern = ' donor',
    replacement = ''
  )) %>% 
  mutate(input_vol = 3000, dataset = 'precision_onc')
```
## united metadata
```{r}
import::from(.directory = r_code.dir, 
             .from = 'helper_analysisFunctions.R', 
             matching_rbind)

united_metadata_df <- matching_rbind(
  'gut_meta_data',
  'elife_meta_data',
  'npj_meta_data',
  'discovery_meta_data',
  'original_panc_meta_data',
  'original_covid_meta_data'
)

original_discovery_meta_data <- 
  matching_rbind('original_meta_data',
                 'discovery_meta_data')

original_normal_meta_data <- 
  matching_rbind('original_panc_meta_data',
                 'discovery_meta_data')

origCovid_normal_meta_data <-
  matching_rbind('original_covid_meta_data',
                 'discovery_meta_data') %>% 
  filter(diagnosis %in% c('covid', 'normal'))

origCovid_normal_meta_data <-
  matching_rbind('original_covid_meta_data',
                 'discovery_meta_data') %>% 
  filter(diagnosis %in% c('covid', 'normal'))


```

# import_data
## load_tximeta_obj
```{r}
import::from(
  .directory = r_code.dir,
  .from = 'helper_analysisFunctions.R',
  load.tximeta.object.list,
  parse.tximeta.quant.metadata
)


internal_rmsk_quant <- load.tximeta.object.list('internal_rmsk_quant',
                                                 'rmsk')

internal_rmsk_quant <-
  parse.tximeta.quant.metadata(internal_rmsk_quant, extra_merge = original_roster_names)

internal_gencode_quant <- load.tximeta.object.list('internal_rmsk_quant',
                                                    'txonly')
internal_gencode_quant <-
  parse.tximeta.quant.metadata(internal_gencode_quant, extra_merge = original_roster_names)

# npj_rmsk_quant <-
#   load.tximeta.object.list('npj_rmsk_quant',
#                            'rmsk')
# npj_rmsk_quant <-
#   parse.tximeta.quant.metadata(npj_rmsk_quant)

elife_rmsk_quant <-
  load.tximeta.object.list('elife_rmsk_quant',
                           'rmsk')
elife_rmsk_quant <-
  parse.tximeta.quant.metadata(elife_rmsk_quant)

elife_gencode_quant <-
  load.tximeta.object.list('elife_rmsk_quant',
                           'txonly')
elife_gencode_quant <-
  parse.tximeta.quant.metadata(elife_gencode_quant)

gut_rmsk_quant <-
  load.tximeta.object.list('gut_rmsk_quant',
                           'rmsk')
gut_rmsk_quant <-
  parse.tximeta.quant.metadata(gut_rmsk_quant)

```

# filter data
## internal
```{r}
import::from(.directory = r_code.dir,
             .from = 'helper_analysisFunctions.R',
             subset.tximeta.se)

## create merged df that matches colData names to metaData names
merge(internal_rmsk_quant$rmsk$quant_meta,
      discovery_meta_data,
      by = 'sample') -> discovery_merge_meta_for_filt

qc_fails <- c('panc.6.2.5', 'panc.7.2.5', 'panc.9.2.5', 
              'panc.10.2.5', 'covid.427.0.80', 'covid.381.0.80')

orig_covid_samples <-
  internal_rmsk_quant$rmsk$quant_meta %>% 
  filter(grepl('^covid', sample)) %>% 
  pull(sample)

orig_panc_samples <-
  internal_rmsk_quant$rmsk$quant_meta %>%
  filter(grepl('^panc', sample)) %>%
  pull(sample)

unknown_samples <-
  internal_rmsk_quant$rmsk$quant_meta %>% 
  filter(grepl('^unknown', sample)) %>% 
  pull(sample)

pitts_samples <-
  internal_rmsk_quant$rmsk$quant_meta %>% 
  filter(grepl('^ctrl.', sample)) %>% 
  pull(sample)

esoph_samples <-
  internal_rmsk_quant$rmsk$quant_meta %>% 
  filter(grepl('^e', sample)) %>% 
  pull(sample)

normal_samples <-
  internal_rmsk_quant$rmsk$quant_meta %>% 
  filter(grepl('^n', sample)) %>% 
  pull(sample)

stage_IandII_select_list <- 
  discovery_merge_meta_for_filt %>% 
    filter(!stage %in% c('I-B', 'II-B', 'II'),
           diagnosis == 'panc') %>% 
  pull(sample)

stage_III_select_list <- 
  discovery_merge_meta_for_filt %>% 
    filter(!stage %in% c('III'),
           diagnosis == 'panc') %>% 
  pull(sample)

stage_IV_select_list <- 
  discovery_merge_meta_for_filt %>% 
    filter(!stage %in% c('IV', 'IV-B'),
           diagnosis == 'panc') %>% 
  pull(sample)

stage_IandII_select_list <- 
  discovery_merge_meta_for_filt %>% 
    filter(!stage %in% c('I-B', 'II-B', 'II'),
           diagnosis == 'panc') %>% 
  pull(sample)

stage_III_select_list <- 
  discovery_merge_meta_for_filt %>% 
    filter(!stage %in% c('III'),
           diagnosis == 'panc') %>% 
  pull(sample)

stage_IV_select_list <- 
  discovery_merge_meta_for_filt %>% 
    filter(!stage %in% c('IV', 'IV-B'),
           diagnosis == 'panc') %>% 
  pull(sample)

discovery_male_select_list <-
  discovery_merge_meta_for_filt %>% 
    filter(gender != 'Male' & diagnosis %in% c('panc', 'normal')) %>% 
  pull(sample)
  
discovery_female_select_list <-
  discovery_merge_meta_for_filt %>% 
    filter(gender != 'Female' & diagnosis %in% c('panc', 'normal')) %>% 
  pull(sample)

preTx_select_list <-
  discovery_merge_meta_for_filt %>%
  filter(diagnosis == 'panc') %>% 
  filter(if_any(
    c(
      treatment_status_radiation_treatment_status,
      treatment_status_resection_treatment_status
    ),
    ~ . != 'Pre Tx'
  )) %>% 
  pull(sample)

postTx_select_list <-
  discovery_merge_meta_for_filt %>%
  filter(diagnosis == 'panc') %>% 
  filter(if_all(
    c(
      treatment_status_radiation_treatment_status,
      treatment_status_resection_treatment_status
    ),
    ~ . == 'Pre Tx'
  )) %>% 
  pull(sample)

internal_rmsk_quant$panc_normal <-
  subset.tximeta.se(
    internal_rmsk_quant$rmsk,
    filter_list = c(
      qc_fails,
      orig_covid_samples,
      unknown_samples,
      esoph_samples,
      stage_IandII_select_list,
      stage_III_select_list,
      stage_IV_select_list,
      pitts_samples
    )
  )

internal_gencode_quant$panc_normal <-
  subset.tximeta.se(
    internal_gencode_quant$txonly,
    filter_list = c(
      qc_fails,
      orig_covid_samples,
      unknown_samples,
      esoph_samples,
      stage_IandII_select_list,
      stage_III_select_list,
      stage_IV_select_list,
      pitts_samples
    )
  )

internal_rmsk_quant$covid_ctrl <-
  subset.tximeta.se(
    internal_rmsk_quant$rmsk,
    filter_list = c(
      qc_fails,
      normal_samples,
      unknown_samples,
      orig_panc_samples,
      esoph_samples,
      stage_IandII_select_list,
      stage_III_select_list,
      stage_IV_select_list
    )
  )

rowRanges(internal_rmsk_quant$rmsk$gxi) %>%
  as.data.frame() %>%
  filter(seqnames == 'chrM') %>%
  select(group_name) %>%
  unlist() %>% unname() -> chrM.filter.list


```
## external elife
```{r}

merge(
  elife_rmsk_quant$rmsk$quant_meta,
    elife_meta_data,
    by = 'sample'
  ) -> elife_merge_meta_for_filt


elife_merge_meta_for_filt %>%
  filter(!diagnosis %in% c('Healthy donor', 'Colorectal Cancer')) %>%
  pull(sample) -> elife_colo_select

elife_merge_meta_for_filt %>%
  filter(!diagnosis %in% c('Healthy donor', 'Esophagus Cancer')) %>%
  pull(sample) -> elife_esoph_select

elife_merge_meta_for_filt %>%
  filter(!diagnosis %in% c('Healthy donor', 'Liver Cancer')) %>%
  pull(sample) -> elife_liver_select

elife_merge_meta_for_filt %>%
  filter(!diagnosis %in% c('Healthy donor', 'Lung Cancer')) %>%
  pull(sample) -> elife_lung_select

elife_merge_meta_for_filt %>%
  filter(!diagnosis %in% c('Healthy donor', 'Stomach Cancer')) %>% 
  pull(sample) -> elife_stomach_select

elife_rmsk_quant$colorectal_healthy <-
  subset.tximeta.se(elife_rmsk_quant$rmsk,
                    filter_list = c(elife_colo_select))

elife_rmsk_quant$esoph_healthy <-
  subset.tximeta.se(elife_rmsk_quant$rmsk,
                    filter_list = c(elife_esoph_select))

elife_rmsk_quant$liver_healthy <-
  subset.tximeta.se(elife_rmsk_quant$rmsk,
                    filter_list = c(elife_liver_select))

elife_rmsk_quant$lung_healthy <-
  subset.tximeta.se(elife_rmsk_quant$rmsk,
                    filter_list = c(elife_lung_select))

elife_rmsk_quant$stomach_healthy <-
  subset.tximeta.se(elife_rmsk_quant$rmsk,
                    filter_list = c(elife_stomach_select))
############

elife_gencode_quant$colorectal_healthy <-
  subset.tximeta.se(elife_gencode_quant$txonly,
                    filter_list = c(elife_colo_select))

elife_gencode_quant$esoph_healthy <-
  subset.tximeta.se(elife_gencode_quant$txonly,
                    filter_list = c(elife_esoph_select))

elife_gencode_quant$liver_healthy <-
  subset.tximeta.se(elife_gencode_quant$txonly,
                    filter_list = c(elife_liver_select))

elife_gencode_quant$lung_healthy <-
  subset.tximeta.se(elife_gencode_quant$txonly,
                    filter_list = c(elife_lung_select))

elife_gencode_quant$stomach_healthy <-
  subset.tximeta.se(elife_gencode_quant$txonly,
                    filter_list = c(elife_stomach_select))
```
## external npj
```{r}
merge(
  npj_rmsk_quant$rmsk$quant_meta,
    npj_meta_data,
    by = 'sample'
  ) -> npj_merge_meta_for_filt

npj_select_list <-
  filter(
    external_quant$ucsc_rmsk_quant$quant_meta,
    !sample %in% npj_merge_meta_for_filt$sample
  ) %>% 
  pull(sample)

external_quant$npj <-
  subset.tximeta.se(external_quant$ucsc_rmsk_quant,
                    filter_list = c(npj_select_list))

external_quant$npj_liverCancer <-
  subset.tximeta.se(external_quant$ucsc_rmsk_quant,
                    filter_list = c(npj_select_list,
                                    npj_meta_data %>% 
                                      filter(!diagnosis %in% c('liver cancer', 'non-cancer')) %>% 
                                      pull(sample)))

external_quant$npj_liverCirrhosis <-
  subset.tximeta.se(external_quant$ucsc_rmsk_quant,
                    filter_list = c(npj_select_list,
                                    npj_meta_data %>% 
                                      filter(!diagnosis %in% c('liver cirrhosis', 'non-cancer')) %>% 
                                      pull(sample)))


```
## external gut
```{r}
merge(
  gut_rmsk_quant$rmsk$quant_meta,
    gut_meta_data,
    by = 'sample'
  ) -> gut_merge_meta_for_filt

```

# merge metadata
## internal
```{r}
internal_rmsk_quant$panc_normal$quant_meta <-
  merge(
    internal_rmsk_quant$panc_normal$quant_meta,
    original_normal_meta_data,
    by = 'sample'
  )

internal_gencode_quant$panc_normal$quant_meta <-
  merge(
    internal_gencode_quant$panc_normal$quant_meta,
    original_normal_meta_data,
    by = 'sample'
  )

internal_rmsk_quant$covid_ctrl$quant_meta <-
  merge(
    internal_rmsk_quant$covid_ctrl$quant_meta,
    original_discovery_meta_data,
    by = 'sample'
  )

```
## external elife
```{r}
elife_rmsk_quant$rmsk$quant_meta <-
  merge(elife_rmsk_quant$rmsk$quant_meta,
        elife_meta_data,
        by = 'sample')

elife_rmsk_quant$colorectal_healthy$quant_meta <-
  merge(elife_rmsk_quant$colorectal_healthy$quant_meta,
        elife_meta_data,
        by = 'sample')

elife_rmsk_quant$esoph_healthy$quant_meta <-
  merge(elife_rmsk_quant$esoph_healthy$quant_meta,
        elife_meta_data,
        by = 'sample')

elife_rmsk_quant$liver_healthy$quant_meta <-
  merge(elife_rmsk_quant$liver_healthy$quant_meta,
        elife_meta_data,
        by = 'sample')

elife_rmsk_quant$lung_healthy$quant_meta <-
  merge(elife_rmsk_quant$lung_healthy$quant_meta,
        elife_meta_data,
        by = 'sample')

elife_rmsk_quant$stomach_healthy$quant_meta <-
  merge(elife_rmsk_quant$stomach_healthy$quant_meta,
        elife_meta_data,
        by = 'sample')

############

elife_gencode_quant$colorectal_healthy$quant_meta <-
  merge(elife_gencode_quant$colorectal_healthy$quant_meta,
        elife_meta_data,
        by = 'sample')

elife_gencode_quant$esoph_healthy$quant_meta <-
  merge(elife_gencode_quant$esoph_healthy$quant_meta,
        elife_meta_data,
        by = 'sample')

elife_gencode_quant$liver_healthy$quant_meta <-
  merge(elife_gencode_quant$liver_healthy$quant_meta,
        elife_meta_data,
        by = 'sample')

elife_gencode_quant$lung_healthy$quant_meta <-
  merge(elife_gencode_quant$lung_healthy$quant_meta,
        elife_meta_data,
        by = 'sample')

elife_gencode_quant$stomach_healthy$quant_meta <-
  merge(elife_gencode_quant$stomach_healthy$quant_meta,
        elife_meta_data,
        by = 'sample')

```
## external npj
```{r}
external_quant$npj$quant_meta <- 
  merge(
    external_quant$npj$quant_meta,
    npj_meta_data,
    by = 'sample'
  )
```
## external gut
```{r}
gut_rmsk_quant$rmsk$quant_meta <- 
  merge(
    gut_rmsk_quant$rmsk$quant_meta,
    gut_meta_data,
    by = 'sample'
  )
```

## construct_analysis_sets
```{r}
import::from(.directory = r_code.dir, 
             .from = 'helper_analysisFunctions.R', 
             build.analysis.set)


united_analysis_set <- list()

rbind(
  internal_quant$filtered_data$quant_meta,
  external_quant$elife$quant_meta %>% select(sample, contains('salmon')),
  external_quant$npj$quant_meta %>% select(sample, contains('salmon')),
  gut_quant$ucsc_rmsk_quant$quant_meta %>% select(sample, contains('salmon'))
) -> united_quant_meta

united_quant_meta %>%
  merge(united_metadata_df, by = 'sample') %>%
  mutate(diagnosis = ifelse(
    diagnosis %in% c('Lung Cancer', 'non-cancer', 'PDAC', 'Liver Cancer'),
    case_when(
      diagnosis == 'Lung Cancer' ~ 'lung cancer',
      diagnosis == 'non-cancer' ~ 'healthy',
      diagnosis == 'PDAC' ~ 'panc',
      diagnosis == 'Liver Cancer' ~ 'liver cancer'
    ),
    diagnosis
  )) -> united_analysis_set$quant_meta

rm(united_quant_meta)

united_gxi_tmp <-
      SparseSummarizedExperiment::cbind(internal_quant$filtered_data$gxi,
                                        external_quant$ucsc_rmsk_quant$gxi)

united_analysis_set$gxi <-
      SparseSummarizedExperiment::cbind(united_gxi_tmp,
                                        gut_quant$ucsc_rmsk_quant$gxi)

united_analysis_set$gxi <-
  SummarizedExperiment::subset(
    united_analysis_set$gxi,
    select = SummarizedExperiment::colData(united_analysis_set$gxi)$names %in% united_analysis_set$quant_meta$sample
  )

rm(united_gxi_tmp)

##
# salmon_quant$analysis_set$panc_ctrl <- 
#   build.analysis.set(se_list = salmon_quant, 
#                      se_1 = 'bioIvt_panc_salmon', 
#                      se_2 = 'pittsburgh_ctrl_salmon',
#                      sample_meta_in = sample_meta_data$`2022.01.14_bioIVTAndPitts_meta_data.csv`)
# 
# salmon_quant$analysis_set$covid_ctrl <- 
#   build.analysis.set(se_list = salmon_quant, 
#                      se_1 = 'bioIvt_covid_salmon', 
#                      se_2 = 'pittsburgh_ctrl_salmon',
#                      sample_meta_in = sample_meta_data$`2022.01.14_combined_meta_data.csv`)
# 
# salmon_quant$analysis_set$ctrl_panc_covid <- 
#   build.analysis.set(se_list = salmon_quant, 
#                      se_1 = 'bioIvt_panc_salmon', 
#                      analysis_set_2 = salmon_quant$analysis_set$covid_ctrl,
#                      sample_meta_in = sample_meta_data$`2022.01.14_combined_meta_data.csv`)
# ##
# ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl <- 
#     build.analysis.set(se_list = ucsc.rmsk.salmon_quant, 
#                        se_1 = 'bioIvt_panc_ucsc.rmsk.salmon', 
#                        se_2 = 'pittsburgh_ctrl_ucsc.rmsk.salmon',
#                        sample_meta_in = sample_meta_data$`2022.01.14_bioIVTAndPitts_meta_data.csv`)
# 
# ucsc.rmsk.salmon_quant$analysis_set$covid_ctrl <- 
#     build.analysis.set(se_list = ucsc.rmsk.salmon_quant, 
#                        se_1 = 'bioIvt_covid_ucsc.rmsk.salmon', 
#                        se_2 = 'pittsburgh_ctrl_ucsc.rmsk.salmon',
#                        sample_meta_in = sample_meta_data$`2022.01.14_combined_meta_data.csv`)
# 
# ucsc.rmsk.salmon_quant$analysis_set$ctrl_panc_covid <- 
#     build.analysis.set(se_list = ucsc.rmsk.salmon_quant, 
#                        se_1 = 'bioIvt_panc_ucsc.rmsk.salmon', 
#                        analysis_set_2 = ucsc.rmsk.salmon_quant$analysis_set$covid_ctrl,
#                        sample_meta_in = sample_meta_data$`2022.01.14_combined_meta_data.csv`)
# ##
# process.aware.salmon_quant$analysis_set$panc_ctrl <- 
#     build.analysis.set(se_list = process.aware.salmon_quant, 
#                        se_1 = 'bioIvt_panc_process.aware.salmon', 
#                        se_2 = 'pittsburgh_ctrl_process.aware.salmon',
#                        sample_meta_in = sample_meta_data$`2022.01.14_bioIVTAndPitts_meta_data.csv`)
# 
# process.aware.salmon_quant$analysis_set$covid_ctrl <- 
#     build.analysis.set(se_list = process.aware.salmon_quant, 
#                        se_1 = 'bioIvt_covid_process.aware.salmon', 
#                        se_2 = 'pittsburgh_ctrl_process.aware.salmon',
#                        sample_meta_in = sample_meta_data$`2022.01.14_combined_meta_data.csv`)
# 
# process.aware.salmon_quant$analysis_set$ctrl_panc_covid <- 
#     build.analysis.set(se_list = process.aware.salmon_quant, 
#                        se_1 = 'bioIvt_panc_process.aware.salmon', 
#                        analysis_set_2 = process.aware.salmon_quant$analysis_set$covid_ctrl,
#                        sample_meta_in = sample_meta_data$`2022.01.14_combined_meta_data.csv`)
```

# analysis_INTERACTIVE
## run_de_seq
```{r}
import::from(.directory = r_code.dir, 
             .from = 'helper_analysisFunctions.R', 
              run.de.seq,
              run.de.seq.individual)


internal_rmsk_quant$panc_normal$deseq <- 
  run.de.seq.individual(input_se = internal_rmsk_quant$panc_normal,
                        type = 'gxi',
                        base_level = 'normal',
                        top_level = 'panc',
                        ref_type = 'ucsc.rmsk',
                        dds_formula = 
                          as.formula('~age+gender+input_vol+diagnosis'))

internal_gencode_quant$panc_normal$deseq <- 
  run.de.seq.individual(input_se = internal_gencode_quant$panc_normal,
                        type = 'gxi',
                        base_level = 'normal',
                        top_level = 'panc',
                        ref_type = 'salmon',
                        dds_formula = 
                          as.formula('~age+gender+input_vol+diagnosis'))

# gut$panc_normal$deseq <- 
#   run.de.seq.individual(input_se = internal_gencode_quant$panc_normal,
#                         type = 'gxi',
#                         base_level = 'normal',
#                         top_level = 'panc',
#                         ref_type = 'salmon',
#                         dds_formula = 
#                           as.formula('~age+gender+input_vol+diagnosis'))
```
## elife de
```{r}

elife_rmsk_quant$rmsk$deseq <- 
  run.de.seq.individual(input_se = elife_rmsk_quant$rmsk,
                        type = 'gxi',
                        base_level = 'Healthy donor',
                        top_level = 'Colorectal Cancer',
                        ref_type = 'ucsc.rmsk',
                        dds_formula = 
                          as.formula('~age+gender+diagnosis'))

elife_rmsk_quant$colorectal_healthy$deseq <- 
  run.de.seq.individual(input_se = elife_rmsk_quant$colorectal_healthy,
                        type = 'gxi',
                        base_level = 'Healthy donor',
                        top_level = 'Colorectal Cancer',
                        ref_type = 'ucsc.rmsk',
                        dds_formula = 
                          as.formula('~age+gender+diagnosis'))

elife_rmsk_quant$esoph_healthy$deseq <- 
  run.de.seq.individual(input_se = elife_rmsk_quant$esoph_healthy,
                        type = 'gxi',
                        base_level = 'Healthy donor',
                        top_level = 'Esophagus Cancer',
                        ref_type = 'ucsc.rmsk',
                        dds_formula = 
                          as.formula('~age+gender+diagnosis'))

elife_rmsk_quant$liver_healthy$deseq <- 
  run.de.seq.individual(input_se = elife_rmsk_quant$liver_healthy,
                        type = 'gxi',
                        base_level = 'Healthy donor',
                        top_level = 'Liver Cancer',
                        ref_type = 'ucsc.rmsk',
                        dds_formula = 
                          as.formula('~age+gender+diagnosis'))

elife_rmsk_quant$lung_healthy$deseq <- 
  run.de.seq.individual(input_se = elife_rmsk_quant$lung_healthy,
                        type = 'gxi',
                        base_level = 'Healthy donor',
                        top_level = 'Lung Cancer',
                        ref_type = 'ucsc.rmsk',
                        dds_formula = 
                          as.formula('~age+gender+diagnosis'))

elife_rmsk_quant$stomach_healthy$deseq <- 
  run.de.seq.individual(input_se = elife_rmsk_quant$stomach_healthy,
                        type = 'gxi',
                        base_level = 'Healthy donor',
                        top_level = 'Stomach Cancer',
                        ref_type = 'ucsc.rmsk',
                        dds_formula = 
                          as.formula('~age+gender+diagnosis'))

######

elife_gencode_quant$colorectal_healthy$deseq <- 
  run.de.seq.individual(input_se = elife_gencode_quant$colorectal_healthy,
                        type = 'gxi',
                        base_level = 'Healthy donor',
                        top_level = 'Colorectal Cancer',
                        ref_type = 'salmon',
                        dds_formula = 
                          as.formula('~age+gender+diagnosis'))

elife_gencode_quant$esoph_healthy$deseq <- 
  run.de.seq.individual(input_se = elife_gencode_quant$esoph_healthy,
                        type = 'gxi',
                        base_level = 'Healthy donor',
                        top_level = 'Esophagus Cancer',
                        ref_type = 'salmon',
                        dds_formula = 
                          as.formula('~age+gender+diagnosis'))

elife_gencode_quant$liver_healthy$deseq <- 
  run.de.seq.individual(input_se = elife_gencode_quant$liver_healthy,
                        type = 'gxi',
                        base_level = 'Healthy donor',
                        top_level = 'Liver Cancer',
                        ref_type = 'salmon',
                        dds_formula = 
                          as.formula('~age+gender+diagnosis'))

elife_gencode_quant$lung_healthy$deseq <- 
  run.de.seq.individual(input_se = elife_gencode_quant$lung_healthy,
                        type = 'gxi',
                        base_level = 'Healthy donor',
                        top_level = 'Lung Cancer',
                        ref_type = 'salmon',
                        dds_formula = 
                          as.formula('~age+gender+diagnosis'))

elife_gencode_quant$stomach_healthy$deseq <- 
  run.de.seq.individual(input_se = elife_gencode_quant$stomach_healthy,
                        type = 'gxi',
                        base_level = 'Healthy donor',
                        top_level = 'Stomach Cancer',
                        ref_type = 'salmon',
                        dds_formula = 
                          as.formula('~age+gender+diagnosis'))


```


## run_pca
```{r}
import::from(.directory = r_code.dir, 
             .from = 'helper_analysisFunctions.R', 
              run.pca, extract.meta.pca.correlates)

internal_gencode_quant$panc_normal$pca <- 
  run.pca(input_de = internal_gencode_quant$panc_normal$deseq)

internal_gencode_quant$panc_normal$pca$pca_out.df %>% 
  extract.meta.pca.correlates() %>% 
  filter(pval < 0.05, abs(cor) > 0.5) -> internal_gencode_quant$panc_normal$pca$pc_sig_corr

internal_rmsk_quant$panc_normal$pca <- 
  run.pca(input_de = internal_rmsk_quant$panc_normal$deseq)

internal_rmsk_quant$panc_normal$pca$pca_out.df %>% 
  extract.meta.pca.correlates() %>% 
  filter(pval < 0.05, abs(cor) > 0.6) -> internal_rmsk_quant$panc_normal$pca$pc_sig_corr



# internal_rmsk_quant$panc_normal$quant_meta %>%
#   mutate(reference = 'Repeat-aware') %>% 
#   rbind(internal_gencode_quant$panc_normal$quant_meta %>% 
#           mutate(reference = 'Repeat-naive')) %>% 
#   ggplot(aes(input_vol, salmon_percent_mapped, shape = gender, color = reference)) +
#   scale_color_grey(end = 0.6) +
#   scale_shape_manual(values = disc_identity_shape_pal) +
#   geom_point(aes(fill = after_scale(alpha(color, 0.4)))) 
# 
# internal_rmsk_quant$panc_normal$pca$pc_sig_corr %>% 
#   mutate(reference = 'Repeat-aware') %>% 
#   rbind(internal_gencode_quant$panc_normal$pca$pc_sig_corr %>% 
#           mutate(reference = 'Repeat-naive')) %>% 
#   mutate(var = str_remove_all(var, 'salmon_')) %>% 
#   ggplot(aes(reorder(var, -cor), abs(cor), color = reference)) +
#   geom_col(aes(fill = after_scale(alpha(color, 0.4))), position = position_dodge()) +
#   scale_color_grey(end = 0.6) +
#   ggforce::facet_row(~pc, scales = 'free_x')
  

```
## 1vsall
```{r}
PAAD_1vAll <- read_csv(file.path(output_data.dir, 'PAAD_vs_all.results'), skip = 7) %>% filter(log2FoldChange != 0)
PAAD_v_Pancreas <- read_csv(file.path(output_data.dir, 'Pancreas_vs_all.results.lfcShrink'), skip = 6)


# PAAD_1vAll %>% 
#   # filter(name %in% reference_meta_data$ucsc_rmsk_insert_info.df$gene) %>% 
#   ggplot(aes(log2FoldChange, -log10(padj), label = name)) + 
#   geom_point()
#   ggrepel::geom_text_repel()

PAAD_1vAll %>% 
  filter(! name %in% Pancreas_1vAll$name) %>% 
  merge(reference_meta_data$total_tx_to_gene.df %>% select(ensg, 'name' = gene) %>% distinct(), by = 'name') -> filt_PAAD_1vAll

```


## run_fgsea
```{r}
import::from(.directory = r_code.dir, 
             .from = 'helper_analysisFunctions.R', 
              make.fgsea.ranks)

import::from(.from = 'msigdbr',
             msigdbr)

msig.df <- 
  msigdbr(species = 'Homo sapiens', category = 'H') %>% 
  dplyr::select(gene_symbol, gs_name) %>% 
  split(x = .$gene_symbol, f = .$gs_name)

internal_rmsk_quant$panc_normal$deseq$fgsea <- 
  make.fgsea.ranks(de.seq = 
                     internal_rmsk_quant$panc_normal$deseq$de_out %>% 
                     filter(gene %in% reference_meta_data$gencode_tx_to_gene.df$gene))

salmon_quant$analysis_set$ctrl_panc_covid$deseq$fgsea$covid_v_ctrl <- 
  make.fgsea.ranks(de.seq = salmon_quant$analysis_set$covid_ctrl$deseq$de_out)

```

## run_process_aware
```{r}
import::from(.directory = r_code.dir, 
             .from = 'helper_analysisFunctions.R', 
             generate.unsplice.data)

process.aware.salmon_quant$analysis_set$ctrl_panc_covid$unsplice_data.df <- 
  generate.unsplice.data(process.aware.salmon_quant$analysis_set$ctrl_panc_covid)
```

## meta table re number
```{r}
salmon_quant$analysis_set$ctrl_panc_covid$quant_meta %>% 
  gather(var, value, -sample, -condition, -sex, -context) %>% 
  mutate(value = as.numeric(value),
         condition = factor(condition, levels = c('panc', 'ctrl', 'covid'))) %>% 
  filter(var %in% c('age', 'input_vol')) %>% 
  spread(var, value) %>% 
  filter(! sample %in% qc_fails) %>% 
  select(-context, -input_vol) %>% 
  mutate(num = ifelse(grepl('covid', sample),
                      row_number(),
                      str_split_fixed(string = sample, pattern = '[.]', n = 4)[, 2]),
         old_name = sample,
         num = ifelse(condition == 'ctrl', as.numeric(num)-3, num),
         sample = paste0(condition, ' ', num)) -> meta_grob_input_table
```

```{r}
tmp <- read_csv(file.path(output_data.dir, 'unmelted_cohort_data.csv'))

tmp %>% 
  mutate(Human = TRUE) %>% 
  mutate(insertion_status = str_replace(insertion_status, pattern = '\\[', replacement = ''),
         insertion_status = str_replace(insertion_status, pattern = '\\]', replacement = ''),
         insertion_status = str_replace_all(insertion_status, pattern = ' ', replacement = ''),
         insertion_status = as.list(as.vector(insertion_status))) %>% 
  unnest(insertion_status) %>% 
  separate(insertion_status, sep = '[,]', into = c('Chimp', 'Gorilla', 'Orangutan', 'Bonobo')) %>% 
  select(insertion = insertion_y, Human, Chimp, Gorilla, Orangutan, Bonobo, `panc.1.2.3`:`ctrl.9.1.2`) %>%
  mutate(Chimp = as.logical(Chimp),
         Gorilla = as.logical(Gorilla),
         Orangutan = as.logical(Orangutan),
         Bonobo = as.logical(Bonobo)) %>% 
  pivot_longer(names_to = 'sample', cols = `panc.1.2.3`:`ctrl.9.1.2`) %>% 
  select(insertion:Bonobo) %>% 
  pivot_longer(names_to = 'species', cols = Human:Bonobo) %>% 
  filter(value == T) -> tmp_insert_df

tmp %>% 
  mutate(Human = TRUE) %>% 
  mutate(insertion_status = str_replace(insertion_status, pattern = '\\[', replacement = ''),
         insertion_status = str_replace(insertion_status, pattern = '\\]', replacement = ''),
         insertion_status = str_replace_all(insertion_status, pattern = ' ', replacement = ''),
         insertion_status = as.list(as.vector(insertion_status))) %>% 
  unnest(insertion_status) %>% 
  separate(insertion_status, sep = '[,]', into = c('Chimp', 'Gorilla', 'Orangutan', 'Bonobo')) %>% 
  select(insertion = insertion_y, Human, Chimp, Gorilla, Orangutan, Bonobo, `panc.1.2.3`:`ctrl.9.1.2`) %>% 
  mutate(Chimp = as.logical(Chimp),
         Gorilla = as.logical(Gorilla),
         Orangutan = as.logical(Orangutan),
         Bonobo = as.logical(Bonobo)) %>% 
  pivot_longer(names_to = 'sample', cols = `panc.1.2.3`:`ctrl.9.1.2`) %>% 
  select(insertion, sample, value) %>% 
  separate(sample, sep = '[.]', into = c('condition', NA, NA, NA)) -> tmp_value_df

tmp %>% 
  select(-`...1`) %>% 
  mutate(Human = TRUE) %>% 
  select(insertion = insertion_x, everything(), -insertion_y) %>% 
  filter_if(is.numeric, any_vars(. > 15)) %>% 
  pivot_longer(names_to = 'sample', cols = `panc.1.2.3`:`ctrl.9.1.2`) %>% 
  select(insertion:Bonobo) %>% 
  pivot_longer(names_to = 'species', cols = Human:Bonobo) %>% 
  filter(value == TRUE) %>% 
  distinct() -> tmp_insert_df

tmp %>% 
  select(-`...1`) %>% 
  mutate(Human = TRUE) %>% 
  select(insertion = insertion_x, everything(), -insertion_y) %>% 
  filter_if(is.numeric, any_vars(. > 15)) %>% 
  pivot_longer(names_to = 'sample', cols = `panc.1.2.3`:`ctrl.9.1.2`) %>% 
  select(insertion, sample, value) %>% 
  distinct() %>% 
  separate(sample, sep = '[.]', into = c('condition', NA, NA, NA)) -> tmp_value_df

merge(tmp_insert_df, 
      tmp_value_df,
      by = 'insertion') -> tmp_merge_df
  

tmp_merge_df %>% 
  merge(reference_meta_data$ucsc_rmsk_tx_to_gene.df, by.x = 'insertion',
        by.y = 'enst') %>% 
  mutate(species = factor(species, levels = c('Human', 'Gorilla', 'Chimp', 'Orangutan', 'Bonobo'))) %>% 
  group_by(insertion, condition) %>% 
  summarize(species = unique(list(species)), 
            mean_exp = mean(log1p(value.y)/n()), 
            sd_exp = sd(log1p(value.y))/n()) %>%
            # mean_exp = sum(value.y/n())) %>%
            # mean_exp = mean(log1p(value.y)), sd_exp = sd(log1p(value.y))) %>%
  ggplot(aes(x = species,y = mean_exp, color = condition)) + 
  scale_color_manual(values = identity_color_pal) +
  # geom_errorbar(position = position_dodge(width = 0.45)) +
  geom_point(position = position_jitterdodge(dodge.width = 0.75, jitter.width = 0.15)) + 
  ggupset::scale_x_upset(order_by = 'degree') + 
  ylab('avg. log expression per insertion')

tmp_merge_df %>% 
  group_by(insertion, condition) %>% 
  summarize(species = unique(list(species)), mean_exp = mean(log1p(value.y)), sd_exp = sd(log1p(value.y))) %>%
  filter(grepl('Alu', insertion)) %>% nrow()
  filter(mean_exp > 0) %>% 
  ggplot(aes(x = species, fill = condition)) + 
  scale_fill_manual(values = identity_color_pal) +
  geom_bar(position = position_dodge()) +
  ggupset::scale_x_upset(order_by = 'degree') + ylab('number of detected Alu insertions')

```

# Figures
# Fig_1
## Fig_flowchart_1A
```{r 1A}
# install.packages('pdftools')
# install.packages('magick')
flow_chart.plt <- magick::image_read_pdf(path = file.path(figure.dir, 'fig.1', 'plasma_schematic_v6.pdf'))
flowchart_wrap_fig1A.plt <- patchwork::wrap_elements(full = magick::image_ggplot(flow_chart.plt))

```
## Fig_1B/C
```{r 1B/C}
internal_gencode_quant$panc_normal$pca$pca_out.df %>%
  ggplot(aes(PC1, PC2,
             color = diagnosis, shape = gender)) +
  geom_point(size = txt.mm_to_pts(0.8),
             aes(fill = after_scale(alpha(color, 0.4))),
             show.legend = F) +
  geom_rug(show.legend = F) +
  xlab(paste('PC1', round(
    cumsum(internal_gencode_quant$panc_normal$pca$pc_props)[1],
    digits = 3
  ) * 100, '%', sep = ' ')) +
  ylab(paste('PC2', round(
    cumsum(internal_gencode_quant$panc_normal$pca$pc_props)[2] -
      cumsum(internal_gencode_quant$panc_normal$pca$pc_props)[1],
    digits = 3
  ) * 100, '%', sep = ' ')) +
  geom_vline(
    xintercept = 0,
    linetype = 'dotted',
    alpha = 0.3,
    size = 0.25
  ) +
  geom_hline(
    yintercept = 0,
    linetype = 'dotted',
    alpha = 0.3,
    size = 0.25
  ) +
  ggforce::geom_mark_ellipse(
    aes(group = stage, label = stage),
    label.fontsize = txt.mm_to_pts(3),
    expand = 0.01,
    show.legend = F
  ) +
  scale_color_manual(values = disc_identity_color_pal) +
  scale_shape_manual(values = disc_identity_shape_pal) -> fig.1B.plt

internal_rmsk_quant$panc_normal$pca$pca_out.df %>%
  ggplot(aes(PC1, PC2,
             color = diagnosis, shape = gender)) +
  geom_point(size = txt.mm_to_pts(0.8),
             aes(fill = after_scale(alpha(color, 0.4))),
             show.legend = F) +
  geom_rug(show.legend = F) +
  xlab(paste('PC1', round(
    cumsum(internal_rmsk_quant$panc_normal$pca$pc_props)[1],
    digits = 3
  ) * 100, '%', sep = ' ')) +
  ylab(paste('PC2', round(
    cumsum(internal_rmsk_quant$panc_normal$pca$pc_props)[2] -
      cumsum(internal_rmsk_quant$panc_normal$pca$pc_props)[1],
    digits = 3
  ) * 100, '%', sep = ' ')) +
  geom_vline(
    xintercept = 0,
    linetype = 'dotted',
    alpha = 0.3,
    size = 0.25
  ) +
  geom_hline(
    yintercept = 0,
    linetype = 'dotted',
    alpha = 0.3,
    size = 0.25
  ) +
  ggforce::geom_mark_ellipse(
    aes(group = stage, filter = stage != 'none', label = stage), 
    label.fontsize = txt.mm_to_pts(3),
    expand = 0.01,
    show.legend = F
  ) +
  scale_color_manual(values = disc_identity_color_pal) +
  scale_shape_manual(values = disc_identity_shape_pal) -> fig.1C.plt
    
    


```
## Fig_1D
```{r 1D}

internal_rmsk_quant$panc_normal$quant_meta %>%
  mutate(reference = 'Repeat-aware') %>%
  rbind(
    internal_gencode_quant$panc_normal$quant_meta %>%
      mutate(reference = 'Repeat-naive')
  ) %>%
  mutate(reference = factor(reference, levels = c('Repeat-naive', 'Repeat-aware')),
         sample = factor(sample)) %>%
  ggplot(aes(reference, salmon_percent_mapped, shape = gender, color = diagnosis)) +
  geom_boxplot(fill = NA, outlier.color = NULL, aes(group = reference), show.legend = F) +
  geom_point(size = txt.mm_to_pts(0.8), aes(fill = after_scale(alpha(color, 0.4)))) +
  geom_line(
    aes(group = sample),
    alpha = 0.2,
    show.legend = F
  ) +
  ylab('% of reads mapped') + xlab('') + 
  scale_shape_manual(values = disc_identity_shape_pal) +
  scale_color_manual(values = disc_identity_color_pal) +
  ggforce::facet_row( ~ diagnosis, strip.position = 'bottom') +
  ggpubr::stat_compare_means(comparisons = list(c('Repeat-naive', 'Repeat-aware')), 
                           method = 'wilcox', size = txt.mm_to_pts(1.2), 
                           method.args = list(exact = TRUE, paired = T), label = 'p.signif') -> fig.1D.plt




internal_rmsk_quant$panc_normal$deseq$norm_counts.df %>%
  rownames_to_column('gene') %>%
  merge(reference_meta_data$ucsc_rmsk_insert_info.df,
        by = 'gene',
        all.x = T) %>%
  mutate(
    family = ifelse(
      gene %in% reference_meta_data$gencode_lncRNA_gene_names.df$ensg,
      'GENCODE_lncRNA',
      family
    ),
    family = ifelse(is.na(family), 'GENCODE_coding', family),
    clade = ifelse(grepl('GENCODE', family), family, clade),
    clade = ifelse(
      clade %in% c(
        'GENCODE_coding',
        'GENCODE_lncRNA',
        'LINE',
        'SINE',
        'LTR',
        'DNA',
        'Simple_repeat'
      ),
      clade,
      'other'
    ),
    type = ifelse(grepl('GENCODE', family), family, 'TE')
  ) %>%
  group_by(type) %>%
  summarize(across(where(is.numeric), ~ .x / sum(.x))) %>%
  pivot_longer(names_to = 'sample',
               values_to = 'frequency',
               cols = where(is.numeric)) %>%
  group_by(sample, type) %>%
  filter(frequency > 0) %>%
  summarize(total_non_zero = n()) %>%
  merge(original_normal_meta_data, by = 'sample') %>%
  ggplot(aes(
    diagnosis,
    total_non_zero,
    shape = gender,
    color = diagnosis,
    fill = after_scale(alpha(color, 0.4))
  )) +
  geom_boxplot(fill = NA,
               outlier.shape = NA,
               aes(group = diagnosis),
               show.legend = F) +
  geom_jitter(size = txt.mm_to_pts(0.8), width = 0.125, show.legend = T) +
  # scale_y_continuous(breaks = c(1000, 5000, 10000)) +
  scale_shape_manual(values = disc_identity_shape_pal) +
  scale_color_manual(values = disc_identity_color_pal) +
  ylab('detected genes') + xlab('') +
  theme(legend.position = 'none') +
  ggforce::facet_row(~ type, strip.position = 'bottom', scales = 'free_y') +
  ggpubr::stat_compare_means(comparisons = list(c('panc', 'normal')), 
                         method = 'wilcox', size = txt.mm_to_pts(1.2), 
                         method.args = list(exact = TRUE, paired = F), label = 'p.signif') -> fig.1E.plt
```
### Fig_1_patchwork
```{r Fig1}
import::from(.from = 'patchwork', plot_annotation, wrap_plots, guide_area, plot_spacer, wrap_elements)


fig_1_layout <- "
AB
AC
EE
DD
"

wrap_plots(flowchart_wrap_fig1A.plt,
           fig.1D.plt+ coord_cartesian(clip = 'off'),
           fig.1E.plt+ coord_cartesian(clip = 'off'),
           # fig.1B.plt + coord_cartesian(clip = 'off'),
           # fig.1C.plt + coord_cartesian(clip = 'off'),
           plot_spacer(),
           guide_area(),
           design = fig_1_layout, 
           guides = 'collect',
           heights = c(0.8,0.8,0.2,0.8)) + plot_annotation(tag_levels = 'A') & 
  theme(plot.tag.position = c(0.01, 0.99), 
        legend.position = 'bottom',
        legend.title = element_text(vjust = 1.25),
        legend.justification = c(0.9, 0.5)) -> fig.1.patchwork


save.manuscript.panel(figure = 1,
                      figure.dir = figure.dir,
                      name = 'patchwork_full_fig_1',
                      width = figure_double_column_width,
                      height = figure_depth,
                      plot.in = fig.1.patchwork)


```

# Fig_2
## Fig_2A
```{r}
internal_rmsk_quant$panc_normal$deseq$norm_counts.df %>%
  rownames_to_column('gene') %>%
  merge(reference_meta_data$ucsc_rmsk_insert_info.df,
        by = 'gene',
        all.x = T) %>%
  mutate(
    family = ifelse(
      gene %in% reference_meta_data$gencode_lncRNA_gene_names.df$ensg,
      'GENCODE_lncRNA',
      family
    ),
    family = ifelse(is.na(family), 'GENCODE_coding', family),
    clade = ifelse(grepl('GENCODE', family), family, clade),
    clade = ifelse(
      clade %in% c(
        'GENCODE_coding',
        'GENCODE_lncRNA',
        'LINE',
        'SINE',
        'LTR',
        'DNA',
        'Simple_repeat'
      ),
      clade,
      'other'
    ),
    type = ifelse(grepl('GENCODE', family), 'GENCODE', 'TE'),
    clade = factor(clade, levels = te_aware_annotation_levels)
  ) %>%
  filter(!gene %in% chrM.filter.list) -> total_annotated_rmsk_counts

total_annotated_rmsk_counts %>%
  gather(sample, count, -gene, -clade, -family, -type) %>%
  filter(count > 0) %>%
  merge(internal_rmsk_quant$panc_normal$quant_meta,
        by = 'sample') -> tidy_total_annotated_rmsk_counts

tidy_total_annotated_rmsk_counts %>%
  group_by(sample, diagnosis, clade) %>%
  summarize(clade_sum = sum(count)) %>%
  mutate(clade_frac = clade_sum / sum(clade_sum)) %>%
  merge(internal_rmsk_quant$panc_normal$quant_meta %>% select(-diagnosis),
        by = 'sample') %>%
  distinct() %>%
  ggplot(aes(sample, clade_frac, fill = clade, color = diagnosis)) +
  geom_col(color = 'white', show.legend = T) +
  scale_fill_manual(values = te_color_pal, name = 'biotype/clade') +
  ggforce::facet_row(~ diagnosis + stage, scales = 'free_x', space = 'free') +
  scale_y_continuous(expand = c(0, 0)) +
  xlab('N = {normal: 22, panc: 10}') + ylab('normalized count fraction') +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank()) -> fig.2A.plt
```
## Fig_2B
```{r}
import::from(.directory = r_code.dir, 
             .from = 'helper_analysisFunctions.R', 
              calc_entropy)

calc_entropy(
  internal_rmsk_quant$panc_normal$deseq$norm_counts.df,
  original_discovery_meta_data,
  grouping_level = 'clade'
) %>%
  filter(
    clade %in% c(
      'GENCODE_coding',
      'GENCODE_lncRNA',
      'SINE',
      'LTR',
      'Simple_repeat'
    )
  ) %>%
  ggplot(aes(
    diagnosis,
    entropy,
    # shape = gender,
    color = diagnosis,
    fill = after_scale(alpha(color, 0.4))
  )) +
  geom_boxplot(
    fill = NA,
    outlier.shape = NA,
    aes(group = diagnosis),
    show.legend = F
  ) +
  geom_jitter(width = 0.125, show.legend = F, shape = 21) +
  ggforce::facet_row( ~ clade, space = 'free', strip.position = 'top') +
  ylim(0, 12) + ylab('Shannon Entropy') + xlab('') +
  # scale_shape_manual(values = disc_identity_shape_pal) +
  scale_color_manual(values = disc_identity_color_pal) +
  ggpubr::stat_compare_means(
    comparisons = list(c('normal', 'panc')),
    method = 'wilcox',
    size = 2.5,
    method.args = list(exact = TRUE, paired = F),
    label = 'p.signif'
  ) +
  theme(axis.text.x = element_text(angle = 40, hjust = 1),
        strip.text = element_text(size = txt.mm_to_pts(1.5))) -> fig.2B.plt
```
## Fig_2C
```{r}
internal_rmsk_quant$panc_normal$deseq$de_out %>%
  merge(reference_meta_data$ucsc_rmsk_insert_info.df,
        by = 'gene',
        all.x = T) %>%
  mutate(
    family = ifelse(
      ensg %in% reference_meta_data$gencode_lncRNA_gene_names.df$ensg,
      'GENCODE_lncRNA',
      family
    ),
    family = ifelse(is.na(family), 'GENCODE_coding', family),
    clade = ifelse(grepl('GENCODE', family), family, clade),
    clade = ifelse(
      clade %in% c(
        'GENCODE_coding',
        'GENCODE_lncRNA',
        'LINE',
        'SINE',
        'LTR',
        'DNA',
        'Simple_repeat'
      ),
      clade,
      'other'
    ),
    type = ifelse(grepl('GENCODE', family), 'GENCODE', 'TE'),
    clade = factor(clade, levels = te_aware_annotation_levels)
  ) %>%
  filter(!ensg %in% chrM.filter.list,
         !grepl('^MT', gene)) -> panc_annotated_rmsk_de


panc_label_list <- panc_annotated_rmsk_de %>%
  filter(grepl('LTR|^L1|Alu|^SSU', gene), padj < 0.01) %>%
  top_n(7, wt = log2FoldChange)


panc_label_gencode_list_dn <- panc_annotated_rmsk_de %>%
  filter(clade == 'GENCODE_coding', padj < 0.01) %>%
  top_n(5, wt = -log2FoldChange)

panc_label_gencode_list_up <- panc_annotated_rmsk_de %>%
  filter(clade == 'GENCODE_coding', padj < 0.01) %>%
  top_n(8, wt = log2FoldChange)

panc_label_list <-
  rbind(panc_label_list, panc_label_gencode_list_dn, panc_label_gencode_list_up)

panc_annotated_rmsk_de %>%
  filter(clade == 'GENCODE_coding') %>%
  ggplot(aes(log2FoldChange, -log10(padj), color = clade)) +
  geom_point(size = 2,
             alpha = 0.6,
             show.legend = FALSE) -> gencode_layer.plt

gencode_layer.plt +
  geom_point(
    data = filter(panc_annotated_rmsk_de, clade != 'GENCODE_coding') %>%
      mutate(clade = factor(clade, levels = te_aware_annotation_levels)),
    aes(log2FoldChange, -log10(padj), color = clade),
    size = 2,
    alpha = 0.8,
    show.legend = F
  ) +
  geom_hline(
    yintercept = -log10(0.01),
    linetype = 'dashed',
    alpha = 0.4
  ) +
  geom_vline(xintercept = 0,
             linetype = 'dashed',
             alpha = 0.4) +
  scale_color_manual(values = te_color_pal, name = 'biotype/clade') +
  ggrepel::geom_text_repel(
    data = panc_annotated_rmsk_de %>%
      mutate(clade = factor(clade, levels = te_aware_annotation_levels)),
    aes(label = ifelse(gene %in% panc_label_list$gene, gene, '')),
    show.legend = F,
    color = 'black',
    size = txt.mm_to_pts(1),
    max.overlaps = 10000,
    box.padding = 0.3,
    segment.colour = 'grey'
  ) +
  xlim(-5, 5) #-> fig.2C.plt

# panc_annotated_rmsk_de %>%
#   mutate(sig = factor(ifelse(padj < 0.01, T, F), levels = c(T, F))) %>%
#   filter(sig == F) %>%
#   ggplot(aes(log(baseMean), log2FoldChange)) +
#   geom_point(color = 'grey', alpha = 0.2) -> unsig_plt
# 
# unsig_plt +
#   geom_point(
#     data = panc_annotated_rmsk_de %>%
#       mutate(sig = factor(ifelse(padj < 0.01, T, F), levels = c(T, F))) %>%
#       filter(sig == T),
#     aes(color = clade, fill = after_scale(alpha(color, 0.4))),
#     show.legend = F
#   ) +
#   ggrepel::geom_text_repel(
#     data = panc_annotated_rmsk_de %>%
#       mutate(sig = factor(ifelse(padj < 0.01, T, F), levels = c(T, F))) %>%
#       filter(sig == T),
#     aes(label = ifelse(sig == T, gene, '')) ,
#     size = txt.mm_to_pts(1),
#     box.padding = 0.3,
#     show.legend = F,
#     max.overlaps = 25
#   ) +
#   geom_hline(yintercept = 0, linetype = 'dashed') +
#   ylim(-5, 5) +
#   scale_color_manual(values = te_color_pal, name = 'biotype/clade') +
#   theme(legend.position = c(0.99, 0.25)) -> fig.2E.plt

```
## Fig_2D
```{r}
# unique(c(filter(ucsc.rmsk.salmon_quant$analysis_set$covid_ctrl$deseq$de_out,
#                 abs(log2FoldChange) > 3.5)$ensg,
#   filter(ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out,
#          abs(log2FoldChange) > 1.5)$ensg)) -> de_ensg_names


internal_rmsk_quant$panc_normal$deseq$norm_counts.df %>%
  # sample_n(100) %>% 
  rownames_to_column('ensg') %>%
  filter(ensg %in% filter(reference_meta_data$ucsc_rmsk_insert_info.df,
                          clade %in% c('SINE', 'Simple_repeat'))$gene) %>%
  filter(across(where(is.numeric), ~ any(.x != 0))) %>%
  rowwise() %>%
  mutate(sum = sum(c_across(where(is.numeric)))) %>%
  filter(sum > 34*5) %>% select(-sum) %>%
  ungroup() %>%
  column_to_rownames('ensg') %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column('ensg') %>% 
  mutate_if(is.numeric, ~ scale(log2(. + 1), center = T)) %>% 
  column_to_rownames('ensg') %>% as.matrix() %>% t() -> hm_count_data

hm_meta_data <-
  internal_rmsk_quant$panc_normal$deseq$scaled_quant_meta_for_de.df

top_hma <-
  ComplexHeatmap::columnAnnotation(
    df = hm_meta_data %>% select(condition),
    col = list(condition = disc_identity_color_pal),
    show_annotation_name = F,
    show_legend = F,
    simple_anno_size = unit(.3, "cm"),
    annotation_legend_param =
      list(title_gp = grid::gpar(
        fontsize = 6, fontface = 'bold'
      ))
  )

  

reference_meta_data$ucsc_rmsk_insert_info.df %>% 
  filter(gene %in% rownames(hm_count_data)) %>% 
  select(gene, clade) %>% 
  # filter(clade != 'Unknown', !grepl('\\?', clade)) %>% 
  mutate(clade = ifelse(clade %in% names(te_color_pal), clade, 'other')) %>%
  distinct() %>% 
  # group_by(gene) %>% 
  # filter(!gene %in% c('(GAATG)n', '(CATTC)n')) %>%
  column_to_rownames('gene') -> hm_te_data

hm_te_data <- hm_te_data[match(rownames(hm_count_data), rownames(hm_te_data)), , drop = F]

left_hma <- ComplexHeatmap::rowAnnotation(
  df = hm_te_data,
  col = list(clade = te_color_pal),
  show_annotation_name = T,
  annotation_name_side = 'top',
  annotation_name_gp = grid::gpar(fontsize = 8, fontface = 'italic'),
  show_legend = F,
  simple_anno_size = unit(.275, "cm")
)

set.seed(2023-03-07)
ComplexHeatmap::Heatmap(hm_count_data, 
                        show_row_names = F, 
                        show_column_names = F, 
                        left_annotation = left_hma, 
                        row_dend_width = unit(.45, "cm"),
                        column_dend_height = unit(.45, "cm"),
                        heatmap_legend_param = 
                          list(legend_gp = grid::gpar(fontsize = 1),
                               legend_direction = "horizontal", 
                               legend_height = unit(.35, "cm"),
                               title_position = "lefttop",
                               title_gp = grid::gpar(fontsize = 6, fontface = 'bold')),
                        top_annotation = top_hma,
                        column_names_gp = grid::gpar(fontsize = 6),
                        name = 'Z-score') -> fig.2D.chm

patchwork::wrap_elements(full = grid::grid.grabExpr(ComplexHeatmap::draw(fig.2D.chm, heatmap_legend_side = "bottom"),
                                                    merge_legend = T)) -> fig.2D.chm
  
```
### Fig_2_patchwork
```{r Fig1}
import::from(.from = 'patchwork', plot_annotation, wrap_plots, guide_area, plot_spacer)


fig_2_layout <- "
AA
EE
BD
CD
"

wrap_plots(fig.2A.plt,
           fig.2B.plt,
           fig.2C.plt,
           fig.2D.chm,
           # fig.2C.plt + labs(title = 'covid'),
           guide_area(),
           # plot_spacer(),
           # plot_spacer(),
           # patchwork::wrap_elements(full = grid::grid.grabExpr(ComplexHeatmap::draw(fig.2D.chm), 
           #                                                      width = figure_double_column_width)),
           design = fig_2_layout, 
           guides = 'collect',
           heights = c(1.25,0.225,1,1, .75)) + plot_annotation(tag_levels = 'A') & 
  theme(plot.tag.position = c(0.01, 0.99), 
        legend.position = 'bottom',
        legend.title = element_text(vjust = 1),
        legend.justification = c(0.5, 0.5)) -> fig.2.patchwork

save.manuscript.panel(figure = 2,
                      figure.dir = figure.dir,
                      name = 'patchwork_full_fig_2',
                      width = figure_double_column_width,
                      height = figure_depth,
                      plot.in = fig.2.patchwork)


```

# New Fig_3
## Fig_3A
```{r}

imap(elife_rmsk_quant[names(elife_rmsk_quant) != 'rmsk'], function(analysis_set, name) {
  analysis_set$deseq$de_out %>%
    filter(
      padj < 0.01,
      log2FoldChange <= -0.75,
      gene %in% reference_meta_data$ucsc_rmsk_insert_info.df$gene
    ) %>%
    mutate(set = name)
  
}) %>% bind_rows() %>%
  merge(
    reference_meta_data$ucsc_rmsk_insert_info.df %>% select(gene, clade) %>% distinct(),
    by = 'gene'
  ) %>%
  mutate(clade = ifelse(clade %in% names(te_color_pal), clade, 'other'),
       set = str_remove_all(set, '_healthy')) %>% 
  group_by(gene, clade) %>%
  summarize(set = list(set)) %>%
  ggplot(aes(set, fill = clade)) +
  geom_bar(show.legend = F, color = 'white', position = 'dodge') +
  geom_text(
    stat = 'count',
    aes(label = after_stat(count), color = clade),
    vjust = -.5,
    size = txt.mm_to_pts(1),
    position = position_dodge(width = 0.9),
    show.legend = F
  ) +
  scale_fill_manual(values = te_color_pal) +
  scale_color_manual(values = te_color_pal) +
  xlab('') + ylab('') +
  ggupset::scale_x_upset() -> fig.3E.plt


imap(elife_rmsk_quant[names(elife_rmsk_quant) != 'rmsk'], function(analysis_set, name) {
  analysis_set$deseq$de_out %>%
    filter(
      padj < 0.01,
      log2FoldChange >= 0.75,
      gene %in% reference_meta_data$ucsc_rmsk_insert_info.df$gene
    ) %>%
    mutate(set = name)
  
}) %>% bind_rows() %>%
  merge(
    reference_meta_data$ucsc_rmsk_insert_info.df %>% select(gene, clade) %>% distinct(),
    by = 'gene'
  ) %>%
  mutate(clade = ifelse(clade %in% names(te_color_pal), clade, 'other'),
         set = str_remove_all(set, '_healthy')) %>% 
  group_by(gene, clade) %>%
  summarize(set = list(set)) %>% 
  ggplot(aes(set, fill = clade)) +
  geom_bar(show.legend = F, color = 'white', position = 'dodge') +
  geom_text(
    stat = 'count',
    aes(label = after_stat(count), color = clade),
    vjust = -.5,
    size = txt.mm_to_pts(1),
    position = position_dodge(width = 0.9),
    show.legend = F
  ) +
  scale_fill_manual(values = te_color_pal) +
  scale_color_manual(values = te_color_pal) +
  xlab('') + ylab('') +
  ggupset::scale_x_upset() -> fig.3F.plt


calc_entropy(elife_rmsk_quant$rmsk$deseq$norm_counts.df,
             elife_meta_data,
             grouping_level = 'clade') %>%
  filter(clade %in% te_aware_annotation_levels) %>%
  ggplot(aes(diagnosis,
             entropy,
             # shape = gender,
             color = diagnosis,
             fill = after_scale(alpha(color, 0.4)))) +
  geom_boxplot(
    fill = NA,
    outlier.shape = NA,
    aes(group = diagnosis),
    show.legend = F
  ) + 
  ggforce::facet_row( ~ clade, space = 'free', strip.position = 'top')

filt_elife_de_list <- bind_rows(elife_de_list[names(elife_de_list) != 'rmsk'])

elife_rmsk_quant$rmsk$deseq$norm_counts.df %>%
  # sample_n(100) %>%
  rownames_to_column('ensg') %>%
  filter(ensg %in% filt_elife_de_list$ensg) %>%
  # filter(across(where(is.numeric), ~ any(.x != 0))) %>%
  # rowwise() %>%
  # mutate(sum = sum(c_across(where(is.numeric)))) %>%
  # filter(sum > 224 * 10) %>% select(-sum) %>%
  # ungroup() %>%
  column_to_rownames('ensg') %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column('ensg') %>%
  mutate_if(is.numeric, ~ scale(log2(. + 1), center = T)) %>%
  column_to_rownames('ensg') %>% as.matrix() %>% t() -> hm_count_data


cor(hm_count_data) -> hm_cor_data




elife_identity_color_pal <-
  setNames(
    RColorBrewer::brewer.pal(n = 6, 'Set2'),
    elife_rmsk_quant$rmsk$deseq$scaled_quant_meta_for_de.df %>%
      pull(condition) %>% unique()
  )

hm_meta_data <-
  elife_rmsk_quant$rmsk$deseq$scaled_quant_meta_for_de.df

top_hma <-
  ComplexHeatmap::columnAnnotation(
    df = hm_meta_data %>% select(condition),
    col = list(condition = elife_identity_color_pal),
    show_annotation_name = F,
    show_legend = T,
    simple_anno_size = unit(.3, "cm"),
    annotation_legend_param =
      list(title_gp = grid::gpar(
        fontsize = 6, fontface = 'bold'
      ))
  )

left_hma <-
  ComplexHeatmap::rowAnnotation(
    df = hm_meta_data %>% select(condition),
    col = list(condition = elife_identity_color_pal),
    show_annotation_name = F,
    show_legend = F,
    simple_anno_size = unit(.3, "cm"),
    annotation_legend_param =
      list(title_gp = grid::gpar(
        fontsize = 6, fontface = 'bold'
      ))
  )

set.seed(2023 - 03 - 07)
ComplexHeatmap::Heatmap(
  hm_cor_data,
  show_row_names = F,
  show_column_names = F,
  left_annotation = left_hma,
  row_dend_width = unit(.25, "cm"),
  column_dend_height = unit(.25, "cm"),
  heatmap_legend_param =
    list(
      legend_gp = grid::gpar(fontsize = 1),
      legend_direction = "horizontal",
      legend_height = unit(.35, "cm"),
      title_position = "lefttop",
      title_gp = grid::gpar(fontsize = 6, fontface = 'bold')
    ),
  top_annotation = top_hma,
  column_names_gp = grid::gpar(fontsize = 6),
  name = 'Z-score'
) 



reference_meta_data$ucsc_rmsk_insert_info.df %>%
  filter(gene %in% rownames(hm_count_data)) %>%
  select(gene, clade) %>%
  mutate(clade = ifelse(clade %in% names(te_color_pal), clade, 'other')) %>%
  distinct() %>% 
  group_by(gene) %>% 
  mutate(count = n()) %>% 
  filter(n() > 1) %>% 
  filter(if (count==2) clade != 'other' else .) -> dupe_filt
  
reference_meta_data$ucsc_rmsk_insert_info.df %>%
  filter(gene %in% rownames(hm_count_data)) %>%
  select(gene, clade) %>%
  mutate(clade = ifelse(clade %in% names(te_color_pal), clade, 'other')) %>%
  distinct() %>% 
  filter(!gene %in% dupe_filt$gene) %>% 
  rbind(dupe_filt %>% select(gene, clade)) %>% 
  column_to_rownames('gene') -> hm_te_data

hm_te_data <-
  hm_te_data[match(rownames(hm_count_data), rownames(hm_te_data)), , drop = F]

left_hma <- ComplexHeatmap::rowAnnotation(
  df = hm_te_data,
  col = list(clade = te_color_pal),
  show_annotation_name = T,
  annotation_name_side = 'top',
  annotation_name_gp = grid::gpar(fontsize = 8, fontface = 'italic'),
  show_legend = F,
  simple_anno_size = unit(.275, "cm")
)

set.seed(2023 - 03 - 07)
ComplexHeatmap::Heatmap(
  hm_count_data,
  show_row_names = F,
  show_column_names = F,
  left_annotation = left_hma,
  row_dend_width = unit(.25, "cm"),
  column_dend_height = unit(.25, "cm"),
  heatmap_legend_param =
    list(
      legend_gp = grid::gpar(fontsize = 1),
      legend_direction = "horizontal",
      legend_height = unit(.35, "cm"),
      title_position = "lefttop",
      title_gp = grid::gpar(fontsize = 6, fontface = 'bold')
    ),
  top_annotation = top_hma,
  column_names_gp = grid::gpar(fontsize = 6),
  name = 'Z-score'
) #-> fig.2D.chm

patchwork::wrap_elements(full = grid::grid.grabExpr(
  ComplexHeatmap::draw(fig.2D.chm, heatmap_legend_side = "bottom"),
  merge_legend = T
)) #-> fig.2D.chm
```
## Fig_3B
```{r}
elife_rmsk_quant$rmsk$pca <- 
  run.pca(input_de = elife_rmsk_quant$rmsk$deseq, te_only = FALSE)

elife_rmsk_quant$rmsk$pca$pca_out.df %>%
  ggplot(aes(PC3, PC4,
             color = diagnosis)) +
  geom_point(size = txt.mm_to_pts(0.8),
             aes(fill = after_scale(alpha(color, 0.4))),
             show.legend = F) +
  geom_rug(show.legend = F) +
  xlab(paste('PC1', round(
    cumsum(elife_rmsk_quant$rmsk$pca$pc_props)[1],
    digits = 3
  ) * 100, '%', sep = ' ')) +
  ylab(paste('PC2', round(
    cumsum(elife_rmsk_quant$rmsk$pca$pc_props)[2] -
      cumsum(elife_rmsk_quant$rmsk$pca$pc_props)[1],
    digits = 3
  ) * 100, '%', sep = ' ')) +
  geom_vline(
    xintercept = 0,
    linetype = 'dotted',
    alpha = 0.3,
    size = 0.25
  ) +
  geom_hline(
    yintercept = 0,
    linetype = 'dotted',
    alpha = 0.3,
    size = 0.25
  ) +
  scale_color_manual(values = elife_identity_color_pal)
```
## Fig_3C
```{r}

de_out_list <- list(elife_rmsk_quant$liver_healthy$deseq$de_out,
                 elife_rmsk_quant$lung_healthy$deseq$de_out,
                 elife_rmsk_quant$esoph_healthy$deseq$de_out,
                 elife_rmsk_quant$colorectal_healthy$deseq$de_out,
                 elife_rmsk_quant$stomach_healthy$deseq$de_out)

lapply(c(de_out_list), function(de_out) {
  
  de_out %>% 
    merge(reference_meta_data$ucsc_rmsk_insert_info.df,
          by = 'gene',
          all.x = T) %>%
    mutate(
      family = ifelse(
        ensg %in% reference_meta_data$gencode_lncRNA_gene_names.df$ensg,
        'GENCODE_lncRNA',
        family
      ),
      family = ifelse(is.na(family), 'GENCODE_coding', family),
      clade = ifelse(grepl('GENCODE', family), family, clade),
      clade = ifelse(
        clade %in% c(
          'GENCODE_coding',
          'GENCODE_lncRNA',
          'LINE',
          'SINE',
          'LTR',
          'DNA',
          'Simple_repeat'
        ),
        clade,
        'other'
      ),
      type = ifelse(grepl('GENCODE', family), 'GENCODE', 'TE'),
      clade = factor(clade, levels = te_aware_annotation_levels)
    ) %>%
    filter(!ensg %in% chrM.filter.list,
           !grepl('^MT', gene)) -> panc_annotated_rmsk_de

  panc_annotated_rmsk_de %>%
    filter(clade == 'GENCODE_coding') %>%
    ggplot(aes(log2FoldChange, -log10(padj), color = clade)) +
    geom_point(size = 2,
               alpha = 0.6,
               show.legend = FALSE) -> gencode_layer.plt

  gencode_layer.plt +
    geom_point(
      data = filter(panc_annotated_rmsk_de, clade != 'GENCODE_coding') %>%
        mutate(clade = factor(clade, levels = te_aware_annotation_levels)),
      aes(log2FoldChange, -log10(padj), color = clade, fill = after_scale(alpha(color,  0.4))),
      size = 2,
      alpha = 0.8
    ) +
    geom_hline(
      yintercept = -log10(0.01),
      linetype = 'dashed',
      alpha = 0.4
    ) +
    geom_vline(xintercept = 0,
               linetype = 'dashed',
               alpha = 0.4) +
    scale_color_manual(values = te_color_pal, name = 'biotype/clade') -> plt
  
  return(plt)
  
}) -> plt_list



```
### Fig_3_patchwork
```{r}
import::from(.from = 'patchwork', plot_annotation, wrap_plots, guide_area, plot_spacer, wrap_elements)

fig_3_layout <- "
AB
CD
EF
GH
"

wrap_plots(plt_list[[1]] + ggtitle('liver cancer'),
           plt_list[[2]] + ggtitle('lung cancer'),
           plt_list[[3]] + ggtitle('esophagus cancer'),
           plt_list[[4]] + ggtitle('colorectal cancer'),
           plt_list[[5]] + ggtitle('stomach cancer'),
           guide_area(),
           wrap_elements(full = fig.3F.plt + ylim(0,95) + ggtitle('upregulated TEs')),
           wrap_elements(full = fig.3E.plt + ylim(0,95) + ggtitle('downregulated TEs')),
           heights = c(0.75,0.75,0.75,1.75),
           # widths = c(0.75, 1.25),
           guides = 'collect',
           design = fig_3_layout) +
  plot_annotation(tag_levels = 'A') &
  theme(plot.tag.position = c(0.01, 0.99), 
        legend.position = 'bottom',
        legend.title = element_text(vjust = 0.5),
        legend.justification = c(0.5, 0.5)) -> fig.3.patchwork

save.manuscript.panel(figure = 3,
                      figure.dir = figure.dir,
                      name = 'patchwork_full_fig_3',
                      width = figure_double_column_width,
                      height = figure_depth,
                      plot.in = fig.3.patchwork)
```



# Fig_S1
## Fig_S1A/B
```{r 1B/C}

fig.S1A.plt <- 
  salmon_quant$analysis_set$ctrl_panc_covid$deseq$pca$pca_1v2.plt + xlim(-280,280)
  
fig.S1B.plt <- 
  salmon_quant$analysis_set$ctrl_panc_covid$deseq$pca$pca_2v3.plt

```
## Fig_S1C
```{r S1A}
salmon_quant$analysis_set$ctrl_panc_covid$quant_meta %>% 
  gather(var, value, -sample, -condition, -sex, -context) %>% 
  mutate(value = as.numeric(value),
         condition = factor(condition, levels = c('panc', 'ctrl', 'covid'))) -> tidy_quant_meta.df

meta_of_interest <- c('star_multimapped_percent',
                      'star_uniquely_mapped_percent')

tidy_quant_meta.df %>%  
  filter(var %in% meta_of_interest) %>% 
  mutate(var = sub('star_', '', var),
         var = case_when(var == 'uniquely_mapped_percent' ~ 'uniquely-mapped', 
                         var == 'multimapped_percent' ~ 'multi-mapped'),
         var = factor(var, levels = c('uniquely-mapped', 'multi-mapped'))) %>% 
  ggplot(aes(condition, value, color = condition)) + 
  geom_boxplot(fill = NA, outlier.colour = NA, show.legend = F) + 
  geom_point(position = position_jitter(width = 0.2), size = rel(1), alpha = 0.8) +
  # scale_shape_manual(values = c(21,25)) +
  ylab('percent mapped') + xlab('') +
  ggforce::facet_row(~var, scales = 'free_y', strip.position = 'bottom') + 
  scale_color_manual(values = identity_color_pal) + 
  theme(legend.position = 'right', legend.justification = 'center') +
  ggpubr::stat_compare_means(comparisons = list(c('panc', 'covid'),
                                                c('ctrl', 'covid'), 
                                                c('ctrl', 'panc')), 
                             method = 'wilcox.test', size = 2.5, method.args = list(exact =T), label = 'p.signif') -> mapping_rate_dist.plt

fig.S1C.plt <- mapping_rate_dist.plt


```
## Fig_S1D
```{r S1B}
s1_flow_chart.plt <- magick::image_read_pdf(path = file.path(figure.dir, 'fig.1', 'complete-seq-diagram-no-process.pdf'))
flowchart_wrap_figS1D.plt <- wrap_elements(full = magick::image_ggplot(s1_flow_chart.plt))
```
### Fig_S1_patchwork
```{r}
import::from(.from = 'patchwork', plot_annotation, wrap_plots, guide_area, plot_spacer, wrap_elements)


fig_S1_layout <- "
AB
CC
DD
EE
"

wrap_plots(fig.S1A.plt,
           fig.S1B.plt,
           guide_area(),
           fig.S1C.plt + coord_cartesian(clip = 'off'),
           flowchart_wrap_figS1B.plt,
           design = fig_S1_layout, 
           guides = 'collect',
           heights = c(0.55,0.145,1.1,1.3)) + plot_annotation(tag_levels = 'A') & 
  theme(plot.tag.position = c(0.01, 0.99), 
        legend.position = 'bottom',
        legend.title = element_text(vjust = 1.25),
        legend.justification = c(0.5, 0.5)) -> fig.S1.patchwork


save.manuscript.panel(figure = 1,
                      figure.dir = figure.dir,
                      name = 'patchwork_full_fig_S1',
                      width = figure_double_column_width,
                      height = figure_depth,
                      plot.in = fig.S1.patchwork)
```

# Fig_S2
## Fig_S2A
```{r S2A}
salmon_quant$analysis_set$ctrl_panc_covid$deseq$pca$pca_corr_out.df %>% 
  group_by(pc) %>% 
  mutate(pc = paste0('PC', pc)) %>% 
  filter(!grepl('num', var),
         ! var %in% c('input_vol', 'age'),
         grepl('map', var), grepl('percent', var)) %>% 
  top_n(7, wt = abs(cor)) %>% 
  mutate(var = ifelse(var == 'salmon_percent_mapped', 'salmon_mapped_percent', var)) %>% 
  dplyr::rename('gencode' = 'cor') -> salmon_pca_corr_top_7.df

ucsc.rmsk.salmon_quant$analysis_set$ctrl_panc_covid$deseq$pca$pca_corr_out.df %>% 
  group_by(pc) %>% 
  mutate(pc = paste0('PC', pc)) %>%
  mutate(var = ifelse(var == 'salmon_percent_mapped', 'salmon_mapped_percent', var)) %>% 
  dplyr::rename('Repeat-aware' = 'cor') -> ucsc.rmsk.salmon_pca_corr_compare.df

salmon_pca_corr_top_7.df %>% 
  merge(ucsc.rmsk.salmon_pca_corr_compare.df, by = c('pc', 'var')) %>% 
  gather(reference, cor, -pc, -var) %>% 
  filter(reference == 'Repeat-aware') %>% 
  ggplot(aes(reorder(var, -abs(cor)),cor, fill = reference)) + 
  geom_col(color = 'white', position = position_dodge()) +
  # coord_flip() + 
  ggforce::facet_col(~pc) + 
  xlab('') + ylab('correlation between metric and PC (pearson)') + 
  geom_hline(yintercept = c(-0.5, 0.5), linetype = 'dashed', alpha = 0.4) + 
  theme(legend.position = 'top', legend.justification = 'center', 
        axis.text.x = element_text(angle = 60, hjust = 1)) + 
  scale_fill_manual(values = c('black', 'grey')) -> pca_correlation_top7.plt
  
fig.S2A.plt <- pca_correlation_top7.plt
```
## Fig_S2B
```{r S2B}


meta_grob_input_table %>% 
  select(-num, -old_name) %>% 
  gridExtra::tableGrob(theme = gridExtra::ttheme_minimal(base_size = 8,
                                                         core = list(padding=unit(c(1, 3.25), "mm"))), 
                       rows = NULL) -> fig.S2B.meta_table.grob

```
### Fig_S2_patchwork
```{r Fig1}
import::from(.from = 'patchwork', plot_annotation, wrap_plots, guide_area, plot_spacer)


fig_S2_layout <- "
AB
AB
"

wrap_plots(fig.S2A.plt,
           patchwork::wrap_elements(full = fig.S2B.meta_table.grob, ignore_tag = F),
           design = fig_S2_layout, 
           widths = c(1,1)) + plot_annotation(tag_levels = 'A') -> fig.S2.patchwork

save.manuscript.panel(figure = 2,
                      figure.dir = figure.dir,
                      name = 'patchwork_full_fig_S2',
                      width = figure_double_column_width,
                      height = figure_depth,
                      plot.in = fig.S2.patchwork)


```

# Fig_3
## tcga & gtex load
```{r}
tcga_data_path <- file.path(output_data.dir, 'tcga_gtex')
gtex_colData <-
  read_csv(
    file.path(tcga_data_path, 'panc_gtex_colData.csv'),
    col_names = c('sample', 'name', 'tissue_1', 'tissue_2', 'sex', 'age')
  )

tcga_colData <-
  read_tsv(file.path(tcga_data_path, 'paad_tcga_colData.tsv')) %>%
  mutate(
    submitter_id.samples = str_replace_all(
      string = submitter_id.samples,
      pattern = 'TCGA',
      replacement = 'PAAD'
    ),
    submitter_id.samples = str_replace_all(
      string = submitter_id.samples,
      pattern = '01A',
      replacement = 'TP'
    ),
    submitter_id.samples = str_replace_all(
      string = submitter_id.samples,
      pattern = '11A',
      replacement = 'NT'
    )
  )

if (!file.exists(file.path(tcga_data_path, 'paad_tcga_gene_readNum_matrix.tsv')) &
    !file.exists(file.path(tcga_data_path, 'paad_gtex_gene_readNum_matrix.tsv'))) {
  tcga_paad_counts <-
    vroom::vroom(file.path(tcga_data_path, 'paad_tcga_readNum_matrix.tsv.gz'))
  tx_names <- tcga_paad_counts$Name
  
  enst_names <- tx_names[grepl('^ENST', tx_names), drop = F]
  enst_tx2gene <-
    tibble(tx = enst_names,
           gene = str_split_fixed(
             string = enst_names,
             pattern = '\\|',
             n = Inf
           )[, 6])
  
  te_names <- tx_names[grepl('^hg38', tx_names), drop = F]
  te_tx2gene <-
    tibble(tx = te_names,
           gene = str_split_fixed(
             string = te_names,
             pattern = '_',
             n = Inf
           )[, 3])
  
  tx2gene <- rbind(enst_tx2gene, te_tx2gene)
  
  tx2gene
  vroom::vroom_write(x = tx2gene, file.path(tcga_data_path, 'tcga_gtex_tx2gene.tsv'))
  
  tcga_paad_counts %>%
    merge(tx2gene, by.x = 'Name', by.y = 'tx') -> tcga_paad_counts
  
  tcga_paad_counts %>%
    group_by(gene) %>%
    summarize(across(is.numeric, ~ sum(.x))) -> tcga_paad_gene_counts
  
  vroom::vroom_write(x = tcga_paad_gene_counts,
                     file.path(tcga_data_path, 'paad_tcga_gene_readNum_matrix.tsv'))
  
  gtex_paad_counts <-
    vroom::vroom(file.path(tcga_data_path, 'paad_gtex_readNum_matrix.tsv.gz'))
  
  gtex_paad_counts %>%
    merge(tx2gene, by.x = 'Name', by.y = 'tx') -> gtex_paad_counts
  
  gtex_paad_counts %>%
    group_by(gene) %>%
    summarize(across(is.numeric, ~ sum(.x))) -> gtex_paad_gene_counts
  
  rm(gtex_paad_counts)
  
  vroom::vroom_write(x = gtex_paad_gene_counts,
                     file.path(tcga_data_path, 'paad_gtex_gene_readNum_matrix.tsv'))
  
} else {
  message('loading....')
  tx2gene <-
    read_tsv(file.path(tcga_data_path, 'tcga_gtex_tx2gene.tsv'))
  gtex_paad_gene_counts <-
    read_tsv(file.path(tcga_data_path, 'paad_gtex_gene_readNum_matrix.tsv'))
  tcga_paad_gene_counts <-
    read_tsv(file.path(tcga_data_path, 'paad_tcga_gene_readNum_matrix.tsv'))
  
}
```
## Fig_TCGA_Surv_plt
```{r 3 prep}
library(survminer)
library(survival)
library(viridis)
tcga_paad_gene_counts %>% 
  column_to_rownames('gene') %>% as.matrix() %>% t() -> tcga_paad_te_counts

tcga_paad_meta_data <- tibble('sample' = rownames(tcga_paad_te_counts), 
                              'condition' = ifelse(grepl('PAAD', sample), 'PAAD', 'GTEx'))

t(tcga_paad_te_counts) %>% 
  as.data.frame() %>%
  mutate(across(where(is.numeric), ~round(.x))) -> tcga_paad_te_count_data_for_DEseq

tcga_de_dds <- DESeq2::DESeqDataSetFromMatrix(countData = tcga_paad_te_count_data_for_DEseq, 
                                              colData = tcga_paad_meta_data, 
                                              design = as.formula('~1'), tidy = F)

tcga_de_dds <- DESeq2::estimateSizeFactors(tcga_de_dds)

tcga_paad_dds_norm_counts.df <- as.data.frame(DESeq2::counts(tcga_de_dds, normalized=T))

tcga_colData %>% 
  filter(submitter_id.samples %in% colnames(tcga_paad_dds_norm_counts.df)) %>% 
  select('sample' = submitter_id.samples, 'type' = sample_type.samples) %>% 
  filter(type != 'Metastatic') %>% 
  mutate(patient = str_replace_all(string = sample, 
                                   pattern = '-TP', replacement = ''),
         patient = str_replace_all(string = patient, 
                                   pattern = '-NT', replacement = ''),
         patient = str_replace_all(string = patient,
                                   pattern = 'PAAD-',
                                   replacement = 'TCGA-')) -> tcga_primary_tumors

tcga_primary_tumors %>% 
  group_by(patient) %>% 
  filter(n() > 1) -> tcga_matched_samples 
  
```
### TCGA PCA
```{r}
input_de$norm_counts.df %>% 
  rownames_to_column('ensg') %>% 
  merge(reference_meta_data$total_tx_to_gene.df %>% select(ensg, gene) %>% distinct(), by = 'ensg') %>% 
  select(-ensg) %>% 
  group_by(gene) %>% 
  summarise(across(everything(), ~mean(.x))) %>% 
  column_to_rownames('gene') -> gene_norm_counts

# tcga_pca_input_data <- t(tcga_de_dds_norm_counts.df)[!grepl('-TM$|-NT$', rownames(hm_count_data)), ]
tcga_pca_input_data <- t(tcga_paad_dds_norm_counts.df[colnames(tcga_paad_dds_norm_counts.df) %in% 
                                                         tcga_matched_samples$sample, ])
# tcga_pca_input_meta_data <- hm_meta_data[match(rownames(tcga_pca_input_data), hm_meta_data$sample), ]
tcga_pca_input_meta_data <- tcga_matched_samples 
    
  

tcga_pca_tmp <- tcga_pca_input_data[, colnames(tcga_pca_input_data) %in% rownames(gene_norm_counts), drop = F]
gene_norm_counts_tmp <- gene_norm_counts[rownames(gene_norm_counts) %in% colnames(tcga_pca_tmp), ,drop = F]

match_tcga_pca_tmp <- tcga_pca_tmp[, match(colnames(tcga_pca_tmp), rownames(gene_norm_counts_tmp))]

match_tcga_pca_tmp[, !colnames(match_tcga_pca_tmp) %in%
                           c(ucsc.rmsk.salmon_quant$ucsc_rmsk_quant$deseq$sex_filter.df$group_name), drop = F] ->
  tcga_pca_tmp_in

# input_de$vst_counts.df[rownames(input_de$vst_counts.df) %in% 
#                            c(filt_PAAD_1vAll$ensg), ] %>% 
#     t() -> pca_tmp_in 
  
tcga_pca_tmp_in_sd <- apply(tcga_pca_tmp_in, 2, sd)
tcga_pca_tmp_in_zero_sd <- tcga_pca_tmp_in[,tcga_pca_tmp_in_sd!=0]
tcga_non_zero_pca <- prcomp(tcga_pca_tmp_in_zero_sd, center = T, scale. = T, rank. = 50)

summary(tcga_non_zero_pca)
pcs.of.interest <- apply(tcga_non_zero_pca$x, 2, var)
pcs.props <-  pcs.of.interest/sum(pcs.of.interest)
cumsum(pcs.props)[c(1,2)]
print(as.data.frame(pcs.props) %>% 
        rownames_to_column('pc') %>% 
        mutate(pc = as.numeric(sub('PC', '', pc))) %>% 
        ggplot(aes(pc, pcs.props)) + 
        geom_col())

pca.out <- as.data.frame(tcga_non_zero_pca$x) %>% 
  rownames_to_column('sample') %>% 
  merge(tcga_pca_input_meta_data, by = 'sample')

pca.out.summary <- 
  as.data.frame(summary(non_zero_pca)$importance) %>% 
  rownames_to_column('metric') %>% 
  gather(pc, value, -metric) %>% 
  spread(metric, value)

pca.out %>% 
  ggplot(aes(PC1, PC2,
             fill = type, shape = type)) +
  geom_point(size = txt.mm_to_pts(0.8), alpha = 0.8) +
  scale_shape_manual(values = c(21,25)) +
  xlab(paste('PC1', round(cumsum(pcs.props)[1], digits = 3)*100,'%',sep = ' ')) +
  ylab(paste('PC2', round(cumsum(pcs.props)[2] - cumsum(pcs.props)[1], digits = 3)*100,'%',sep = ' ')) +
  geom_vline(xintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) +
  geom_hline(yintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) + 
  scale_fill_manual(values = c('grey', 'black'))

sample_input_pred <- as.data.frame(t(gene_norm_counts_tmp))

sample_input_pred_filt <- sample_input_pred[, colnames(sample_input_pred) %in% colnames(tcga_pca_tmp_in_zero_sd)]

sample_input_pred_filt_match <- sample_input_pred_filt[, match(colnames(sample_input_pred_filt),
                                                               colnames(tcga_pca_tmp_in_zero_sd))]


sample_pred_pca <- predict(object = tcga_non_zero_pca, newdata = sample_input_pred_filt_match)


sample_pred_pca %>% 
  as.data.frame() %>% 
  rownames_to_column('sample') %>% 
  mutate(sample = str_split_fixed(sample, '_S', 2)[, 1]) %>%
  # merge(original_roster_names, by = 'sample', all.x = T) %>% 
  # mutate(sample = ifelse(is.na(name), sample, name)) %>% 
  # select(-name) %>% 
  merge(input_de$scaled_quant_meta_for_de.df %>% rownames_to_column('sample'), by = 'sample') -> sample_pred_pca.out


sample_pred_pca.out %>% 
  mutate(stage = factor(stage, 
                        levels = c(NA, 'I-B', 'II', 'II-B', 'III', 'IV', 'IV-B'),
                        exclude = NULL)) %>% 
  ggplot(aes(PC3, PC4,
             shape = condition, fill = stage)) +
  geom_point(size = txt.mm_to_pts(0.8), alpha = 0.8) +
  scale_shape_manual(values = c(21,25)) +
  xlab(paste('PC1', round(cumsum(pcs.props)[1], digits = 3)*100,'%',sep = ' ')) +
  ylab(paste('PC2', round(cumsum(pcs.props)[2] - cumsum(pcs.props)[1], digits = 3)*100,'%',sep = ' ')) +
  geom_vline(xintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) +
  geom_hline(yintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) + 
  scale_fill_grey(na.value = 'lightgreen', end = 0.25, start = 1)

```

### adjCurves
```{r}
library(survival)
library(adjustedCurves)
devtools::install_github("tagteam/riskRegression")

gene_set <- filter(
  internal_rmsk_quant$panc_normal$deseq$de_out,
  padj < 0.05,
  abs(log2FoldChange) > 0.25,
  gene %in% reference_meta_data$ucsc_rmsk_insert_info.df$gene
)$gene

t(tcga_paad_dds_norm_counts.df[rownames(tcga_paad_dds_norm_counts.df) %in% gene_set, ]) %>%
  as.data.frame() %>%
  rownames_to_column('sample') %>%
  filter(sample %in% tcga_primary_tumors$sample) -> surv_counts_df

paad_survival <-
  read_tsv(file.path(output_data.dir, 'tcga_gtex', 'PAAD_survival.txt'))

paad_tcga_pheno_merge <-
  tcga_primary_tumors %>% merge(paad_survival %>% rename('patient' = '_PATIENT'), by = 'patient')

paad_tcga_pheno_merge %>%
  filter(sample.x %in% colnames(tcga_paad_dds_norm_counts.df)) %>%
  select(sample = sample.x, everything(), -patient, -sample.y) -> filt_paad_surv


filt_paad_surv %>%
  merge(surv_counts_df, by = 'sample') %>%
  merge(tcga_colData %>% mutate('sample' = submitter_id.samples), by = 'sample') %>%
  mutate('gender' = gender.demographic,
         'stage' = tumor_stage.diagnoses,
         'age' = age_at_initial_pathologic_diagnosis) -> merge_luad_surv

pivot_exclude <-
  c(colnames(filt_paad_surv), colnames(tcga_colData))

merge_luad_surv %>%
  pivot_longer(
    names_to = 'gene',
    values_to = 'count',
    cols = colnames(surv_counts_df)[-1]
  ) %>%
  group_by(sample, age, gender, stage, PFI, PFI.time, OS, OS.time) %>%
  summarize(avg_count = mean(count)) %>%
  ungroup() %>%
  mutate(signature = ifelse(avg_count > median(avg_count), 'high', 'low')) -> merge_luad_surv_final

adj_surv_final <- merge_luad_surv_final %>%
  mutate(stage = case_when(
    stage %in% c('stage i', 'stage ia', 'stage ib') ~ 1,
    stage %in% c('stage iia', 'stage iib', 'stage iii') ~ 2,
  )) %>%
  filter(!is.na(stage)) %>%
  mutate(stage = ifelse(stage == 1, 'stage 1', 'stage 2/3')) %>%
  mutate(signature = factor(signature, levels = c('low', 'high')),
         age = scale(age)[, 1])

adj_surv_final -> tmp_surv

cox_fit <-
  coxph(Surv(time = OS.time, event = OS) ~ age + gender + signature,
        data = tmp_surv,
        x = T)

adjsurv <- adjustedsurv(
  data = tmp_surv,
  variable = "signature",
  ev_time = "OS.time",
  event = "OS",
  method = "direct",
  outcome_model = cox_fit,
  conf_int = TRUE,
  bootstrap = TRUE,
  n_boot = 500
)

tmp_pval <-
  adjusted_curve_test(adjsurv,
                      from = 0,
                      to = max(adjsurv$adjsurv$time),
  )$p_value
adjsurv$adjsurv %>%
  group_by(group) %>%
  ggplot(aes(
    time,
    y = surv,
    group = group,
    fill = group,
    color = group
  )) +
  geom_step() +
  pammtools::geom_stepribbon(aes(ymin = ci_lower, ymax = ci_upper),
                             alpha = 0.085,
                             linetype = 0) +
  scale_color_manual('panc DN signature',
                     values = c(viridis::viridis(3)[1], viridis::viridis(3)[2])) +
  scale_fill_manual(values = c(viridis::viridis(3)[1], viridis::viridis(3)[2]),
                    guide = 'none') +
  annotate(
    'text',
    x = 600,
    y = 0.07,
    label = paste0('pval = ',
                   round(tmp_pval,
                         8)),
    size = txt.mm_to_pts(0.8)
  ) +
  ylab('Overall Survival Probability') +
  xlab('Days') +
  scale_x_continuous(breaks = c(0, 500, 1000, 2000)) +
  theme(
    legend.position = c(0.99, 0.99),
    legend.title = element_text(size = rel(0.8)),
    legend.text = element_text(size = rel(0.6))
  )
 
```

### UP Surv
```{r 3A}
tcga_padd_stratified_counts <-
  tcga_paad_dds_norm_counts.df %>%
  t() %>% as.data.frame()

tcga_padd_stratified_counts %>% 
  rownames_to_column('sample') %>% 
  filter(sample %in% tcga_primary_tumors$sample) %>% 
  gather(gene, count, -sample) %>% 
  filter(gene %in% filter(internal_rmsk_quant$panc_normal$deseq$de_out, 
                          padj < 0.01, log2FoldChange > 0.75)$gene) %>% 
  filter(!gene %in% reference_meta_data$ucsc_rmsk_insert_info.df$gene) %>%
  select(sample, count) %>% 
  group_by(sample) %>%
  summarize(avg.count = mean(count)) %>%
  mutate(gene.set = 'panc_RNA_signal') %>%
  group_by(gene.set) %>%
  mutate(stratum = ifelse(avg.count >= quantile(avg.count, 0.66),
                          'top-third', ifelse(avg.count <= quantile(avg.count, 0.33),
                                              'bottom-third', 'middle'))) %>%
  filter(stratum != 'middle') %>%
  spread(gene.set, stratum) -> tcga_paad_stratified_counts

paad_survival <- read_tsv(file.path(output_data.dir, 'tcga_gtex', 'PAAD_survival.txt'))

paad_tcga_pheno_merge <- 
  tcga_primary_tumors %>% merge(paad_survival, by.x = 'patient', by.y = '_PATIENT')

surv.obj <- Surv(time = filter(paad_tcga_pheno_merge %>% select(sample.x, OS.time, OS) %>% distinct(), 
                               sample.x %in% tcga_paad_stratified_counts$sample)$OS.time,
                 event = filter(paad_tcga_pheno_merge %>% select(sample.x, OS.time, OS) %>% distinct(), 
                                sample.x %in% tcga_paad_stratified_counts$sample)$OS)

surv.fit <- survfit(as.formula(surv.obj ~ panc_RNA_signal), 
                    data = tcga_paad_stratified_counts)
ggsurvplot(surv.fit, 
           ylab = 'O.S. Probability',
           xlab = 'Overall Survival Time -- days',
           pval = T, 
           pval.size = txt.mm_to_pts(1),
           legend.title = "",
           ggtheme = theme_set_fun() + theme(legend.position = c(0.45,0.95)),
           pval.coord = c(0, 0.08),
           palette = viridis(3)[c(2,1)]) -> surv.plt

paste0('panc GENCODE up') -> km_title

surv.plt$plot + ggtitle(km_title) -> plt_surv_A.plt



# ------------------------------------------------------------------------------
filter(ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out, 
                        padj < 0.05, log2FoldChange > 0.75) %>%
  filter(!gene %in% reference_meta_data$ucsc_rmsk_insert_info.df$gene) %>%
  top_n(wt =-padj, 5) %>% dplyr::pull(gene) -> this_top_10

ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$norm_counts.df %>% 
  rownames_to_column('ensg') %>% 
  merge(reference_meta_data$total_tx_to_gene.df %>% select(ensg, gene) %>% distinct(), by = 'ensg') %>% 
  filter(gene %in% this_top_10) %>% 
  select(-ensg) %>% 
  gather(patient, count, -gene) %>% 
  merge(sample_meta_data$`2022.01.14_bioIVTAndPitts_meta_data.csv`, by = 'patient') -> count_data


count_data %>% 
  ggplot(aes(condition, count, color = condition)) + 
  geom_boxplot(fill = NA, width = 0.5, position = position_dodge(width = 0.2), outlier.colour = NULL) + 
  geom_jitter(size = 1, alpha = 0.4, position = position_dodge(width = 0.2)) + 
  ggforce::facet_row(~gene, strip.position = 'bottom') + 
  ggpubr::stat_compare_means(comparisons = list(c('ctrl', 'panc')), ref.group = 'ctrl',
                             method = 'wilcox.test', size = 2.5, method.args = list(exact =T), label = 'p.signif') +
  scale_color_manual(values = identity_color_pal[1:2]) + xlab('') -> plt_exp_A.plt


```
### DN TEs
```{r}
tcga_padd_stratified_counts <-
  tcga_paad_dds_norm_counts.df %>%
  t() %>% as.data.frame()

tcga_padd_stratified_counts %>% 
  rownames_to_column('sample') %>% 
  filter(sample %in% tcga_primary_tumors$sample) %>% 
  gather(gene, count, -sample) %>% 
  filter(gene %in% filter(ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out, 
                          padj < 0.05, log2FoldChange < -1)$gene) %>% 
  filter(gene %in% reference_meta_data$ucsc_rmsk_insert_info.df$gene) %>%
  select(sample, count) %>% 
  group_by(sample) %>%
  summarize(avg.count = mean(count)) %>%
  mutate(gene.set = 'panc_RNA_signal') %>%
  group_by(gene.set) %>%
  mutate(stratum = ifelse(avg.count >= quantile(avg.count, 0.66),
                          'top-third', ifelse(avg.count <= quantile(avg.count, 0.33),
                                              'bottom-third', 'middle'))) %>%
  filter(stratum != 'middle') %>%
  spread(gene.set, stratum) -> tcga_paad_stratified_counts

paad_survival <- read_tsv(file.path(output_data.dir, 'tcga_gtex', 'PAAD_survival.txt'))

paad_tcga_pheno_merge <- 
  tcga_primary_tumors %>% merge(paad_survival, by.x = 'patient', by.y = '_PATIENT')

surv.obj <- Surv(time = filter(paad_tcga_pheno_merge %>% select(sample.x, OS.time, OS) %>% distinct(), 
                               sample.x %in% tcga_paad_stratified_counts$sample)$OS.time,
                 event = filter(paad_tcga_pheno_merge %>% select(sample.x, OS.time, OS) %>% distinct(), 
                                sample.x %in% tcga_paad_stratified_counts$sample)$OS)

surv.fit <- survfit(as.formula(surv.obj ~ panc_RNA_signal), 
                    data = tcga_paad_stratified_counts)
ggsurvplot(surv.fit, 
           ylab = 'O.S. Probability',
           xlab = 'Overall Survival Time -- days',
           pval = T, 
           pval.size = txt.mm_to_pts(1),
           legend.title = "", 
           ggtheme = theme_set_fun() + theme(legend.position = c(0.45,0.95)),
           pval.coord = c(0, 0.08),
           palette = viridis(3)[c(2,1)]) -> surv.plt

paste0('panc Repeat down') -> km_title

surv.plt$plot + ggtitle(km_title) -> plt_surv_B.plt

# ------------------------------------------------------------------------------

filter(ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out, 
                          padj < 0.05, log2FoldChange < -1) %>% 
  filter(gene %in% reference_meta_data$ucsc_rmsk_insert_info.df$gene) %>%
  top_n(wt =-padj, 5) %>% dplyr::pull(gene) -> this_top_10

ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$norm_counts.df %>% 
  rownames_to_column('ensg') %>% 
  merge(reference_meta_data$total_tx_to_gene.df %>% select(ensg, gene) %>% distinct(), by = 'ensg') %>% 
  filter(gene %in% this_top_10) %>% 
  select(-ensg) %>% 
  gather(patient, count, -gene) %>% 
  merge(sample_meta_data$`2022.01.14_bioIVTAndPitts_meta_data.csv`, by = 'patient') -> count_data


count_data %>% 
  ggplot(aes(condition, count, color = condition)) + 
  geom_boxplot(fill = NA, width = 0.5, position = position_dodge(width = 0.2), outlier.colour = NULL) + 
  geom_jitter(size = 1, alpha = 0.4, position = position_dodge(width = 0.2)) + 
  ggforce::facet_row(~gene, strip.position = 'bottom') + 
  ggpubr::stat_compare_means(comparisons = list(c('ctrl', 'panc')), ref.group = 'ctrl',
                             method = 'wilcox.test', size = 2.5, method.args = list(exact =T), label = 'p.signif') +
  scale_color_manual(values = identity_color_pal[1:2]) + xlab('') -> plt_exp_B.plt



```
### OX PHOS
```{r}
tcga_padd_stratified_counts <-
  tcga_paad_dds_norm_counts.df %>%
  t() %>% as.data.frame()

tcga_padd_stratified_counts %>% 
  rownames_to_column('sample') %>% 
  filter(sample %in% tcga_primary_tumors$sample) %>% 
  gather(gene, count, -sample) %>% 
  filter(gene %in% filter(ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out, 
                          padj < 0.05, log2FoldChange > 0.75)$gene) %>%
  filter(gene %in% msig.df$HALLMARK_OXIDATIVE_PHOSPHORYLATION) %>%
  select(sample, count) %>% 
  group_by(sample) %>%
  summarize(avg.count = mean(count)) %>%
  mutate(gene.set = 'panc_RNA_signal') %>%
  group_by(gene.set) %>%
  mutate(stratum = ifelse(avg.count >= quantile(avg.count, 0.66),
                          'top-third', ifelse(avg.count <= quantile(avg.count, 0.33),
                                              'bottom-third', 'middle'))) %>%
  filter(stratum != 'middle') %>%
  spread(gene.set, stratum) -> tcga_paad_stratified_counts

paad_survival <- read_tsv(file.path(output_data.dir, 'tcga_gtex', 'PAAD_survival.txt'))

paad_tcga_pheno_merge <- 
  tcga_primary_tumors %>% merge(paad_survival, by.x = 'patient', by.y = '_PATIENT')

surv.obj <- Surv(time = filter(paad_tcga_pheno_merge %>% select(sample.x, OS.time, OS) %>% distinct(), 
                               sample.x %in% tcga_paad_stratified_counts$sample)$OS.time,
                 event = filter(paad_tcga_pheno_merge %>% select(sample.x, OS.time, OS) %>% distinct(), 
                                sample.x %in% tcga_paad_stratified_counts$sample)$OS)

surv.fit <- survfit(as.formula(surv.obj ~ panc_RNA_signal), 
                    data = tcga_paad_stratified_counts)
ggsurvplot(surv.fit,
           ylab = 'O.S. Probability',
           xlab = 'Overall Survival Time -- days',
           pval = T, 
           pval.size = txt.mm_to_pts(1),
           legend.title = "", 
           ggtheme = theme_set_fun() + theme(legend.position = c(0.45,0.95)),
           pval.coord = c(0, 0.08),
           palette = viridis(3)[c(2,1)]) -> surv.plt

paste0('Oxidative Phosphorylation') -> km_title

surv.plt$plot + ggtitle(km_title) -> plt_surv_C.plt


# ------------------------------------------------------------------------------

filter(ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out, 
                          padj < 0.05, log2FoldChange > 0.75) %>%
  filter(gene %in% msig.df$HALLMARK_OXIDATIVE_PHOSPHORYLATION) %>%
  top_n(wt =-padj, 5) %>% dplyr::pull(gene) -> this_top_10

ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$norm_counts.df %>% 
  rownames_to_column('ensg') %>% 
  merge(reference_meta_data$total_tx_to_gene.df %>% select(ensg, gene) %>% distinct(), by = 'ensg') %>% 
  filter(gene %in% this_top_10) %>% 
  select(-ensg) %>% 
  gather(patient, count, -gene) %>% 
  merge(sample_meta_data$`2022.01.14_bioIVTAndPitts_meta_data.csv`, by = 'patient') -> count_data


count_data %>%
  ggplot(aes(condition, count, color = condition)) +
  geom_boxplot(fill = NA, width = 0.5, position = position_dodge(width = 0.2), outlier.colour = NULL) +
  geom_jitter(size = 1, alpha = 0.4, position = position_dodge(width = 0.2)) +
  ggforce::facet_row(~gene, strip.position = 'bottom') +
  ggpubr::stat_compare_means(comparisons = list(c('ctrl', 'panc')), ref.group = 'ctrl',
                             method = 'wilcox.test', size = 2.5, method.args = list(exact =T), label = 'p.signif') +
  scale_color_manual(values = identity_color_pal[1:2]) + xlab('') -> plt_exp_C.plt


```
### MYC
```{r}
tcga_padd_stratified_counts <-
  tcga_paad_dds_norm_counts.df %>%
  t() %>% as.data.frame()

tcga_padd_stratified_counts %>% 
  rownames_to_column('sample') %>% 
  filter(sample %in% tcga_primary_tumors$sample) %>% 
  gather(gene, count, -sample) %>% 
  filter(gene %in% filter(ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out, 
                          padj < 0.05, log2FoldChange > 0.75)$gene) %>%
  filter(gene %in% msig.df$HALLMARK_MYC_TARGETS_V1) %>%
  select(sample, count) %>% 
  group_by(sample) %>%
  summarize(avg.count = mean(count)) %>%
  mutate(gene.set = 'panc_RNA_signal') %>%
  group_by(gene.set) %>%
  mutate(stratum = ifelse(avg.count >= quantile(avg.count, 0.66),
                          'top-third', ifelse(avg.count <= quantile(avg.count, 0.33),
                                              'bottom-third', 'middle'))) %>%
  filter(stratum != 'middle') %>%
  spread(gene.set, stratum) -> tcga_paad_stratified_counts

paad_survival <- read_tsv(file.path(output_data.dir, 'tcga_gtex', 'PAAD_survival.txt'))

paad_tcga_pheno_merge <- 
  tcga_primary_tumors %>% merge(paad_survival, by.x = 'patient', by.y = '_PATIENT')

surv.obj <- Surv(time = filter(paad_tcga_pheno_merge %>% select(sample.x, OS.time, OS) %>% distinct(), 
                               sample.x %in% tcga_paad_stratified_counts$sample)$OS.time,
                 event = filter(paad_tcga_pheno_merge %>% select(sample.x, OS.time, OS) %>% distinct(), 
                                sample.x %in% tcga_paad_stratified_counts$sample)$OS)

surv.fit <- survfit(as.formula(surv.obj ~ panc_RNA_signal), 
                    data = tcga_paad_stratified_counts)
ggsurvplot(surv.fit, 
           ylab = 'O.S. Probability',
           xlab = 'Overall Survival Time -- days',
           pval = T, 
           pval.size = txt.mm_to_pts(1),
           legend.title = "", 
           ggtheme = theme_set_fun() + theme(legend.position = c(0.45,0.95)),
           pval.coord = c(0, 0.08),
           palette = viridis(3)[c(2,1)]) -> surv.plt

paste0('MYC Targets') -> km_title

surv.plt$plot + ggtitle(km_title) -> plt_surv_D.plt

# ------------------------------------------------------------------------------

filter(ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out, 
                          padj < 0.05, log2FoldChange > 0.75) %>%
  filter(gene %in% msig.df$HALLMARK_MYC_TARGETS_V1) %>%
  top_n(wt =-padj, 5) %>% dplyr::pull(gene) -> this_top_10

ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$norm_counts.df %>% 
  rownames_to_column('ensg') %>% 
  merge(reference_meta_data$total_tx_to_gene.df %>% select(ensg, gene) %>% distinct(), by = 'ensg') %>% 
  filter(gene %in% this_top_10) %>% 
  select(-ensg) %>% 
  gather(patient, count, -gene) %>% 
  merge(sample_meta_data$`2022.01.14_bioIVTAndPitts_meta_data.csv`, by = 'patient') -> count_data


count_data %>% 
  ggplot(aes(condition, count, color = condition)) + 
  geom_boxplot(fill = NA, width = 0.5, position = position_dodge(width = 0.2), outlier.colour = NULL) + 
  geom_jitter(size = 1, alpha = 0.4, position = position_dodge(width = 0.2)) + 
  ggforce::facet_row(~gene, strip.position = 'bottom') + 
  ggpubr::stat_compare_means(comparisons = list(c('ctrl', 'panc')), ref.group = 'ctrl',
                             method = 'wilcox.test', size = 2.5, method.args = list(exact =T), label = 'p.signif') +
  scale_color_manual(values = identity_color_pal[1:2]) + xlab('') -> plt_exp_D.plt

```
### plt_surv_patchwork
```{r}
import::from(.from = 'patchwork', plot_annotation, wrap_plots, guide_area, plot_spacer, wrap_elements)

fig_surv_layout <- "
AC
DE
FG
HI
BB
"

wrap_plots(plt_surv_A.plt + theme(legend.background = element_blank()),
           guide_area(),
           plt_exp_A.plt + coord_cartesian(clip = 'off') + ylab('norm. counts'),
           plt_surv_B.plt + theme(legend.background = element_blank()),
           plt_exp_B.plt + coord_cartesian(clip = 'off') + ylab('norm. counts'),
           plt_surv_C.plt + theme(legend.background = element_blank()),
           plt_exp_C.plt + coord_cartesian(clip = 'off') + ylab('norm. counts'),
           plt_surv_D.plt + theme(legend.background = element_blank()),
           plt_exp_D.plt + coord_cartesian(clip = 'off') + ylab('norm. counts'),
           heights = c(0.9,0.9,0.9,0.9,0.4),
           widths = c(0.75, 1.25),
           guides = 'collect',
           design = fig_surv_layout) +
  plot_annotation(tag_levels = 'A') &
  theme(plot.tag.position = c(0.01, 0.99), 
        legend.position = 'bottom',
        legend.title = element_text(vjust = 0.5),
        legend.justification = c(0.5, 0.5)) -> fig.surv.patchwork

save.manuscript.panel(figure = 3,
                      figure.dir = figure.dir,
                      name = 'patchwork_full_fig_3',
                      width = figure_double_column_width,
                      height = figure_depth,
                      plot.in = fig.surv.patchwork)
```

# Fig_S3
```{r}
SummarizedExperiment::assay(ucsc.rmsk.salmon_quant$analysis_set$ctrl_panc_covid$gxi, 
                            'counts') %>% as.data.frame() -> ucsc_rmsk_counts

ucsc_rmsk_counts <- ucsc_rmsk_counts[, !colnames(ucsc_rmsk_counts) %in% qc_fails]

ucsc_rmsk_counts %>% 
  rownames_to_column('gene') %>% 
  merge(reference_meta_data$ucsc_rmsk_insert_info.df, by = 'gene', all.x = T) %>% 
  mutate(family = ifelse(gene %in% reference_meta_data$gencode_lncRNA_gene_names.df$ensg,
                         'GENCODE_lncRNA', family),
         family = ifelse(is.na(family), 'GENCODE_coding', family),
         clade = ifelse(grepl('GENCODE', family), family, clade),
         clade = ifelse(clade %in% c('GENCODE_coding',
                                     'GENCODE_lncRNA',
                                     'LINE',
                                     'SINE',
                                     'LTR',
                                     'DNA',
                                     'Simple_repeat'), clade, 'other'),
         type = ifelse(grepl('GENCODE',family), 'GENCODE', 'TE'),
         clade = factor(clade, levels = te_aware_annotation_levels)) %>% 
    distinct() -> ucsc_rmsk_counts

ucsc_rmsk_counts %>% 
  pivot_longer(cols = panc.1.2.3:ctrl.9.1.2, names_to = 'sample', values_to = 'count') %>% 
  mutate(condition = case_when(grepl('panc', sample) ~ 'panc',
                               grepl('covid', sample) ~ 'covid',
                               grepl('ctrl', sample) ~ 'ctrl')) %>% 
  group_by(sample, clade, condition) %>% 
  filter(count >= 5) %>% 
  summarize(count = n()) -> ucsc_rmsk_biotype_summary


ucsc_rmsk_biotype_summary %>% 
  group_by(clade) %>% 
  filter(!grepl('\\?', clade)) %>% 
  ggplot(aes(condition, count, color = condition)) + 
  geom_boxplot(fill = NA, outlier.colour = NA) + 
  geom_jitter(width = 0.25) +
  ylab('detected genes/elements') + 
  xlab('') +
  scale_color_manual(values = identity_color_pal) + 
  facet_wrap(~clade, scales = 'free_y', ncol = 2, strip.position = 'bottom') +
  coord_cartesian(clip = 'off') +
  ggpubr::stat_compare_means(method = 'wilcox', 
                             comparisons = list(c('covid', 'ctrl'), c('panc', 'ctrl')),
                             size = txt.mm_to_pts(1), label = 'p.signif') +
  theme(axis.text.x = element_text(angle = 40, hjust = 1),
        legend.position = 'bottom', legend.direction = 'horizontal') -> biotype_count_fig_S3


```
### Fig_S3_patchwork
```{r}
import::from(.from = 'patchwork', plot_annotation, wrap_plots, guide_area, plot_spacer, wrap_elements)

wrap_plots(biotype_count_fig_S3) -> patchwork_fig_S3

save.manuscript.panel(figure = 3,
                      figure.dir = figure.dir,
                      name = 'patchwork_full_fig_S3',
                      width = figure_double_column_width,
                      height = figure_depth,
                      plot.in = patchwork_fig_S3)

```

# Fig_4
## read_len load
```{r}
panc.7.1.1_read_len <- read_csv(file.path(output_data.dir, 'nanopore', 
                                          'nanopore_read_len', 'panc.7.1.1_biotype_read_lengths.csv')) %>% 
  separate(sample, sep = '[.]', into = c(NA,NA,'biotype',NA)) %>% 
  mutate(biotype = sub('_extraction', '', biotype))

panc.6.1.1_read_len <- read_csv(file.path(output_data.dir, 'nanopore', 
                                          'nanopore_read_len', 'panc.6.1.1_biotype_read_lengths.csv')) %>% 
  separate(sample, sep = '[.]', into = c(NA,NA,'biotype',NA)) %>% 
  mutate(biotype = sub('_extraction', '', biotype))

ctrl.6.1.5_read_len <- read_csv(file.path(output_data.dir, 'nanopore',
                                          'nanopore_read_len', 'ctrl.6.1.5_biotype_read_lengths.csv')) %>%
  separate(sample, sep = '[.]', into = c(NA,NA,'biotype',NA)) %>%
  mutate(biotype = sub('_extraction', '', biotype))

panc.7.1.1_read_len %>% 
  mutate(sample = 'panc.7.1.1') %>% 
  rbind(panc.6.1.1_read_len %>%
         mutate(sample = 'panc.6.1.1')) %>% 
  rbind(ctrl.6.1.5_read_len %>%
         mutate(sample = 'ctrl.6.1.5')) %>%
  group_by(biotype) -> panc_nanopore_read_length.df

illumina_read_len <- list.files(file.path(output_data.dir, 'nanopore', 'illumina_read_len'), full.names = T)[1:31]
names(illumina_read_len) <- gsub(basename(illumina_read_len), pattern = '.csv', replacement = '')[1:31]

name_conversion <- data.frame('in_name' = stringr::str_split_fixed(names(illumina_read_len), '[.]', n = 2)[,1])

name_conversion$out_name <- c('panc.1.2.3', 'panc.2.2.7', 'panc.4.3.9', 
                              'panc.5.4.5', 'panc.6.2.5', 'panc.7.2.5',
                              'panc.9.2.5', 'panc.7.1.1', 'panc.9.1.1', 
                              'panc.10.1.2', 'panc.6.1.1', 'covid.400.0.56', 'covid.381.0.80',
                              'covid.396.0.68', 'covid.411.0.75', 'covid.412.0.64',
                              'covid.426.0.80', 'covid.432.0.58', 'covid.441.0.64',
                              'covid.442.0.61', 'covid.450.0.84', 'ctrl.12.1.1', 
                              'ctrl.5.1.3', 'ctrl.7.1.2', 'ctrl.9.1.2', 'ctrl.13.0.7', 
                              'ctrl.8.1.2', 'ctrl.10.0.4', 'ctrl.6.1.5', 'ctrl.11.1.1','ctrl.4.1.3')

illumina_roster_names <- read_csv(file.path(output_data.dir, 'nanopore', 'illumina_read_len', 'ROSTER.csv'),
                                  col_names = c('read', 'sample'))

# lapply(illumina_read_len, function(file) {
# 
#   read_len_file <- read_csv(file = file,
#                             col_names = c('read', 'read_length', 'seq_1',
#                                           'seq_2', 'originally_multisample',
#                                           'tx', 'sample'))
#   read_len_file %>%
#     mutate(tx = ifelse(grepl('^ENST', tx),
#             str_split_fixed(tx, pattern = '[|]', n = 10)[,1],
#             tx)) %>%
#     mutate(tx = ifelse(grepl('_range', tx),
#             str_split_fixed(tx, pattern = '::', n = 2)[,1],
#             tx)) %>%
#   mutate(sample = stringr::str_split_fixed(sample, '[.]', n = 2)[,1],
#          sample = gsub(pattern = 'panc_', replacement = 'panc', sample)) %>%
#     merge(name_conversion, by.x = 'sample', by.y = 'in_name') %>%
#     merge(reference_meta_data$total_tx_to_gene.df, by.x = 'tx', by.y = 'enst') %>%
#     dplyr::rename('sample_name' = 'out_name')
# 
# 
# }) -> illumina_read_len.list
# 
# 
# saveRDS(object = illumina_read_len.list, file = file.path(output_data.dir, 'nanopore', 'illumina_read_len.rds'))
# illumina_read_len.list <- readRDS(file = file.path(output_data.dir, 'nanopore', 'illumina_read_len.rds'))

# illumina_read_len.list %>%
#   bind_rows(.id = 'sample') %>%
#   filter(!sample %in% qc_fails) %>%
#   merge(sample_meta_data$`2022.01.14_combined_meta_data.csv`, by.x = 'sample_name', by.y = 'patient') %>%
#   group_by(condition, biotype) %>%
#   mutate(z_score = scale(abs(read_length))) %>%
#   filter(between(z_score, -2.5, 2.5)) -> illumina_read_len.df
# 
# saveRDS(object = illumina_read_len.df, file = file.path(output_data.dir, 'nanopore',
#                                                            'illumina_read_len.df.rds'))
# rm(illumina_read_len.list)

# illumina_read_len.df <- readRDS(file = file.path(output_data.dir, 'nanopore',
#                                                            'illumina_read_len.df.rds'))
# 
# illumina_read_len.df$seq_1 <- NULL
# illumina_read_len.df$seq_2 <- NULL
# 
# vroom::vroom_write(illumina_read_len.df, file = file.path(output_data.dir, 'nanopore','illumina_read_len_no_seq.tsv.gz'))

# illumina_read_len.df <-
#   vroom::vroom(file = file.path(output_data.dir,'nanopore','illumina_read_len_no_seq.tsv.gz'), num_threads = 16)
```
## Fig_NA
```{r}
table(illumina_read_len.df$biotype) %>% 
  as.data.frame() %>% 
  select(biotype = Var1) %>% 
  filter(biotype %in% c('LINE', 'lncRNA', 'LTR', 'protein_coding', 'SINE')) %>% 
  mutate(biolabel = c('LINE', 'GENCODE_lncRNA', 'LTR', 'GENCODE_coding', 'SINE')) -> illumina_biotype_labels

illumina_read_len.df %>% 
  filter(biotype %in% illumina_biotype_labels$biotype,
         ! sample %in% qc_fails) -> illumina_biotype_to.plt

illumina_read_len.df %>% 
  filter(! sample %in% qc_fails) %>% 
  group_by(sample, biotype, condition) %>% 
  summarize(read_length = median(abs(read_length))) -> tmp_mean_read_len.df
  

tmp_mean_read_len.df %>%
  group_by(biotype, condition) %>%
  filter(n() == 10, ! biotype %in% c('polymorphic_pseudogene', 'TR_C_gene', 'TR_V_gene', 'IG_V_gene', 'IG_C_gene',
                                   'TEC', 'non_stop_decay', 'nonsense_mediated_decay'),
         condition != 'covid') %>% 
  mutate(biotype = ifelse(biotype == 'lncRNA', 'gencode_lncRNA', biotype),
         biotype = ifelse(biotype == 'protein_coding', 'gencode_coding', biotype)) %>% 
  mutate(biotype = factor(biotype, levels = c('SINE', 'LINE', 'LTR', 'GENCODE_lncRNA', 
                                              'retained_intron', 'GENCODE_coding'))) %>% 
  ggplot(aes(condition, read_length, color = condition)) +
  geom_boxplot(outlier.colour = NA) + 
  geom_jitter(width = 0.24, alpha = 0.2) +
  facet_wrap(~biotype, scales = 'free', ncol = 3, strip.position = 'bottom') +
  scale_color_manual(values = identity_color_pal, guide = 'none') + 
  ggpubr::stat_compare_means(method = 'wilcox', comparisons = list(c('panc', 'ctrl')), size = txt.mm_to_pts(0.8)) +
  coord_cartesian(clip = 'off') + 
  theme(strip.placement = 'inside') +
  xlab('') + ylab('median fragment size (bp)') -> plt_NA.plt
```
## Fig_NB
```{r}
tmp_mean_read_len.df %>% 
  filter(condition != 'covid',
         ! sample %in% qc_fails) %>% 
  pivot_wider(names_from = biotype, values_from = read_length) %>% 
  select(-condition) %>% 
  column_to_rownames('sample') -> read_len_pca.tmp
  
pca_tmp_in_sd <- apply(read_len_pca.tmp, 2, sd)
pca_tmp_in_zero_sd <- read_len_pca.tmp[,pca_tmp_in_sd!=0 & !is.na(pca_tmp_in_sd)]
non_zero_pca <- prcomp(pca_tmp_in_zero_sd, center = T, scale. = T, rank. = 50)

summary(non_zero_pca)
pcs.of.interest <- apply(non_zero_pca$x, 2, var)
pcs.props <-  pcs.of.interest/sum(pcs.of.interest)
cumsum(pcs.props)[c(1,2)]
print(as.data.frame(pcs.props) %>% 
        rownames_to_column('pc') %>% 
        mutate(pc = as.numeric(sub('PC', '', pc))) %>% 
        ggplot(aes(pc, pcs.props)) + 
        geom_col())

pca.out <- as.data.frame(non_zero_pca$x) %>% 
  rownames_to_column('sample') %>% 
  merge(tmp_mean_read_len.df %>% select(sample, condition) %>% ungroup() %>% distinct(), by = 'sample') %>% 
  ungroup() %>% 
  select(-biotype) %>% 
  distinct()

pca.out.summary <- 
  as.data.frame(summary(non_zero_pca)$importance) %>% 
  rownames_to_column('metric') %>% 
  gather(pc, value, -metric) %>% 
  spread(metric, value)

pca.out %>% 
  ggplot(aes(PC1, PC2,
             color = condition)) +
  geom_point(size = rel(2), alpha = 1) + 
  # scale_shape_manual(values = c(21,25)) +
  xlab(paste('PC1', round(cumsum(pcs.props)[1], digits = 3), sep = ' ')) +
  ylab(paste('PC2', round(cumsum(pcs.props)[2] - cumsum(pcs.props)[1], digits = 3), sep = ' ')) +
  geom_vline(xintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) +
  geom_hline(yintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) + 
  scale_color_manual(values = identity_color_pal[1:2]) + 
  annotate(geom = 'text', x = -3.75, y = -3.25, label = 'size-aware', size = txt.mm_to_pts(1)) +
  theme(legend.position = c(0.95,1), legend.direction = 'horizontal') -> plt_NB.plt
```
## Fig_NC
```{r}
panc_nanopore_read_length.df %>% 
  group_by(biotype) %>% 
  mutate(z_score = scale(abs(read_length))) %>% 
  filter(between(z_score, -2.5, 2.5))-> panc_nanopore_read_length.df

table(panc_nanopore_read_length.df$biotype) %>% 
  as.data.frame() %>% 
  select(biotype = Var1) %>% 
  mutate(biolabel = c('LINE', 'GENCODE_lncRNA', 'LTR', 'GENCODE_coding', 'SINE')) -> nanopore_biotype_labels

panc_nanopore_read_length.df %>% 
  group_by(biotype) %>% 
  # mutate(biotype = factor(biotype, levels = c('SINE', ')))
  # filter(!grepl('ctrl', sample)) %>% 
  merge(nanopore_biotype_labels, by = 'biotype') %>% 
  mutate(biolabel = factor(biolabel, levels = te_aware_annotation_levels),
         sample = case_when(sample == 'panc.7.1.1' ~ 'panc 7', 
                            sample == 'panc.6.1.1' ~ 'panc 6',
                            sample == 'ctrl.6.1.5' ~ 'ctrl 6')) %>% 
  ggplot(aes(abs(read_length), fill = sample)) + 
  geom_histogram(aes(y=stat(density*width)), bins = 100) + 
  scale_y_continuous(labels = scales::number_format(accuracy = 0.01)) +
  ggforce::facet_col(~biolabel, scales = 'free') +
  scale_fill_manual(values = viridisLite::plasma(20)[c(1,8,10)]) +
  coord_cartesian(clip = 'off') +
  ylab('proportion of nanopore reads') + xlab('nanopore read length (bp)') +
  theme(legend.position = c(0.55,0.99), legend.direction = 'horizontal', plot.margin = margin(r = 3, unit = 'mm')) -> plt_NC.plt
```
## Fig_ND
```{r}
panc_nanopore_read_length.df %>% 
  ungroup() %>% 
  mutate(originally_multisample = ifelse(grepl('^ENST', originally_multisample), 
            str_split_fixed(originally_multisample, pattern = '[|]', n = 10)[,1],
            originally_multisample)) %>% 
  mutate(originally_multisample = ifelse(grepl('_range', originally_multisample),
            str_split_fixed(originally_multisample, pattern = '::', n = 2)[,1],
            originally_multisample)) %>%
  merge(reference_meta_data$total_tx_to_gene.df, by.x = 'originally_multisample', by.y = 'enst') %>% 
  mutate(len = as.numeric(len)) -> len_compare.df

len_compare.df %>%
  filter(biotype.x == 'SINE', sample == 'ctrl.6.1.5') -> ctrl.6.1.5_nanopore_hex_to.plt

len_compare.df %>% 
  filter(biotype.x == 'SINE', sample == 'panc.6.1.1') -> panc.6.1.1_nanopore_hex_to.plt

len_compare.df %>% 
  filter(biotype.x == 'SINE', sample == 'panc.7.1.1') -> panc.7.1.1_nanopore_hex_to.plt

# illumina_read_len.df %>% 
#   filter(! sample %in% qc_fails,
#          abs(read_length) <= 1000,
#          condition != 'covid') %>% 
#   mutate(read_length = abs(read_length)) %>% 
#   group_by(condition) %>% 
#   select(sample_name, biotype, read_length, condition) -> read_len_cdf_tmp
# 
# read_len_cdf_tmp %>% 
#   filter(sample_name == 'ctrl.6.1.5', biotype == 'SINE') %>% 
#   pull(read_length) %>% 
#   AnalystHelper::ecdf_fun(CI = T, CI.interval = 0.99) %>% 
#   mutate(sample = 'ctrl 6') %>% 
#   rbind(read_len_cdf_tmp %>% 
#     filter(sample_name == 'panc.6.1.1', biotype == 'SINE') %>% 
#     pull(read_length) %>% 
#     AnalystHelper::ecdf_fun(CI = T, CI.interval = 0.99) %>% 
#     mutate(sample = 'panc 6')) %>% 
#   rbind(read_len_cdf_tmp %>% 
#     filter(sample_name == 'panc.7.1.1', biotype == 'SINE') %>% 
#     pull(read_length) %>% 
#     AnalystHelper::ecdf_fun(CI = T, CI.interval = 0.99) %>% 
#     mutate(sample = 'panc 7')) %>% 
#   ggplot(aes(x = value, y = proportion, min = lwr.CI, ymax = upr.CI, color = sample)) + 
#   geom_ribbon(alpha = 0.05) + 
#   geom_step() +
#   # xlim(130,315) +
#   scale_color_manual(values = c(three_color_pal[3], three_color_pal[1], three_color_pal[2])) + 
#   ylab('cumulative proportion (99% CI)') + 
#   xlab('read length') + 
#   theme(legend.position = c(0.25, 0.99)) -> plt_ND_1.plt
# devtools::install_github("SwampThingPaul/AnalystHelper")
ctrl.6.1.5_nanopore_hex_to.plt %>% 
  pull(read_length) %>% 
  AnalystHelper::ecdf_fun(CI = T, CI.interval = 0.99) %>% 
  mutate(sample = 'ctrl 6') %>% 
  rbind(panc.6.1.1_nanopore_hex_to.plt %>% 
        pull(read_length) %>% 
        AnalystHelper::ecdf_fun(CI = T, CI.interval = 0.99) %>% 
          mutate(sample = 'panc 6')) %>% 
  rbind(panc.7.1.1_nanopore_hex_to.plt %>% 
        pull(read_length) %>% 
        AnalystHelper::ecdf_fun(CI = T, CI.interval = 0.99) %>% 
          mutate(sample = 'panc 7')) %>% 
  ggplot(aes(x = value, y = proportion, min = lwr.CI, ymax = upr.CI, color = sample)) + 
  geom_ribbon(alpha = 0.05) + 
  geom_step() + 
  xlim(130,315) +
  scale_color_manual(values = viridisLite::plasma(20)[c(1,8,10)], guide = 'none') +
  ylab('cumulative prop. (99% CI)') + 
  xlab('nanopore SINE read length (bp)') + 
  theme(legend.position = c(0.25, 0.99)) -> plt_ND_2.plt

# p4 <- ggplot(data.frame(l = plt_ND_2.plt$labels$y, x = 1, y = 1)) +
#       geom_text(aes(x, y, label = l), angle = 90) + 
#       theme_void() +
#       coord_cartesian(clip = "off")
# 
# panel_ND_layout <- "
# CA
# CB
# "
# 
# wrap_plots(plt_ND_1.plt + ylab('') + xlab(''),
#            plt_ND_2.plt + ylab(''),
#            p4,
#            widths = c(0.02,1),
#            design = panel_ND_layout) -> panel.ND.patchwork
```
## Fig_NE
```{r}
# install.packages('hexbin')
panc.6.1.1_nanopore_hex_to.plt %>% 
  group_by(tx, len, biotype.x, sample) %>% 
  summarize(read_length = mean(read_length)) %>% 
  merge(nanopore_biotype_labels, by.x = 'biotype.x', by.y = 'biotype') %>% 
  mutate(biolabel = factor(biolabel, levels = te_aware_annotation_levels),
         condition = ifelse(grepl('ctrl', sample), 'ctrl', 'panc')) %>% 
  # ggplot(aes(read_length, len, fill = scale(..count../(sum(..count..) * 1e6)))) + 
  ggplot(aes(read_length, len, fill = scales::rescale((..count../(sum(..count..) * 1e6))))) + 
  geom_hex(bins = 120) +
  # geom_vline(xintercept = median(panc.6.1.1_nanopore_hex_to.plt$read_length), color = 'grey', linetype = 'dashed') +
  # annotate(geom = 'text', x = 275, y = 120, label = 'panc 6') +
  xlab('avg. observed SINE read length (bp)') + ylab('expected SINE insert length (bp)') +
  # ggforce::facet_col(~sample) +
  ggtitle('panc 6') + 
  scale_fill_viridis_c(option = 'viridis', direction = 1, name = 'scaled cpm', breaks = c(0,1)) +
  scale_y_continuous(breaks = c(100,200,300,400), limits = c(75,425)) +
  scale_x_continuous(breaks = c(100,200,300,400), limits = c(75,370)) +
  theme(legend.position = c(0.98,0.45), legend.direction = 'horizontal', legend.key.width = unit(1, 'mm')) +
  guides(fill = guide_colorbar(title.position = 'left', title.vjust = 1)) -> plt_NE.plt

ctrl.6.1.5_nanopore_hex_to.plt %>%
  group_by(tx, len, biotype.x, sample) %>%
  summarize(read_length = mean(read_length)) %>%
  merge(nanopore_biotype_labels, by.x = 'biotype.x', by.y = 'biotype') %>%
  mutate(biolabel = factor(biolabel, levels = te_aware_annotation_levels),
         condition = ifelse(grepl('ctrl', sample), 'ctrl', 'panc')) %>%
  ggplot(aes(read_length, len, fill = scales::rescale((..count../(sum(..count..) * 1e6))))) +
  geom_hex(bins = 120) +
  # annotate(geom = 'text', x = 275, y = 150, label = 'ctrl 6') +
  # geom_vline(xintercept = median(ctrl.6.1.5_nanopore_hex_to.plt$read_length), 
  #            color = 'grey', linetype = 'dashed') +
  xlab('avg. observed read length (bp)') + ylab('expected SINE insert length (bp)') +
  # ggforce::facet_col(~sample) +
  ggtitle('ctrl 6') + 
  scale_fill_viridis_c(option = 'viridis', direction = 1, name = 'scaled cpm', guide = 'none') +
  # theme(legend.position = c(0.95,1), legend.direction = 'horizontal', legend.key.width = unit(1, 'mm'))  +
  # guides(fill = guide_colorbar(title.position = 'left', title.vjust = 1)) +
  scale_y_continuous(breaks = c(100,200,300,400), limits = c(75,425)) +
  scale_x_continuous(breaks = c(100,200,300,400), limits = c(75,370)) -> plt_SNA_1.plt

panc.7.1.1_nanopore_hex_to.plt %>% 
  group_by(tx, len, biotype.x, sample) %>% 
  summarize(read_length = mean(read_length)) %>% 
  merge(nanopore_biotype_labels, by.x = 'biotype.x', by.y = 'biotype') %>% 
  mutate(biolabel = factor(biolabel, levels = te_aware_annotation_levels),
         condition = ifelse(grepl('ctrl', sample), 'ctrl', 'panc')) %>% 
  ggplot(aes(read_length, len, fill = scales::rescale((..count../(sum(..count..) * 1e6))))) + 
  geom_hex(bins = 120) +
  # geom_vline(xintercept = median(panc.7.1.1_nanopore_hex_to.plt$read_length), color = 'grey', linetype = 'dashed') +
  # annotate(geom = 'text', x = 275, y = 150, label = 'panc 7') +
  xlab('avg. observed SINE read length (bp)') + ylab('expected SINE insert length (bp)') +
  # ggforce::facet_col(~sample) +
  ggtitle('panc 7') + 
  scale_fill_viridis_c(option = 'viridis', direction = 1, name = 'scaled cpm', guide = 'none') +
  scale_y_continuous(breaks = c(100,200,300,400), limits = c(75,425)) +
  scale_x_continuous(breaks = c(100,200,300,400), limits = c(75,370)) -> plt_SNA_3.plt

panel_SNA_layout <- "
AA
BB
CC
"

wrap_plots(plt_SNA_1.plt + xlab('') + ylab(''),
           plt_NE.plt + xlab(''),
           plt_SNA_3.plt + ylab(''),
           design = panel_SNA_layout) -> panel.SNA.patchwork
  
```
### Fig_Nanopore_patchwork
```{r}
import::from(.from = 'patchwork', plot_annotation, wrap_plots, guide_area, plot_spacer, wrap_elements)


fig_N_layout <- "
AB
AB
AB
AD
CC
"

wrap_plots(plt_NC.plt,
           wrap_elements(full = panel.SNA.patchwork, clip = F),
           guide_area(),
           plt_ND_2.plt + coord_cartesian(clip = 'off'),
           # wrap_elements(full = plt_ND_2.plt),
           design = fig_N_layout, 
           guides = 'collect',
           heights = c(1,1,1,1,0.17)) +
  plot_annotation(tag_levels = 'A') &
  theme(legend.justification = 'center', legend.position = 'bottom') -> fig.N.patchwork


# fig.N.patchwork

save.manuscript.panel(figure = 4,
                      figure.dir = figure.dir,
                      name = 'patchwork_full_fig_4',
                      width = figure_double_column_width,
                      height = figure_depth,
                      plot.in = fig.N.patchwork)
```

# Fig_S4
## Fig_S4A
```{r}
# bind_rows(list('panc' = salmon_quant$analysis_set$ctrl_panc_covid$deseq$fgsea$panc_v_ctrl,
#           'covid' = salmon_quant$analysis_set$ctrl_panc_covid$deseq$fgsea$covid_v_ctrl),
#           .id = 'condition') %>% 
#   select(condition, 'enriched pathway' = pathway, padj, NES) %>% 
#   mutate(padj = round(padj, 10),
#          NES = round(NES, 3)) %>% 
#   dplyr::rename('GSEA padj' = padj, 'GSEA NES' = NES) %>% 
#   gridExtra::tableGrob(theme = gridExtra::ttheme_minimal(), rows = NULL) -> fig.S2A.fgsea_table.grob

bind_rows(list('panc' = salmon_quant$analysis_set$ctrl_panc_covid$deseq$fgsea$panc_v_ctrl,
          'covid' = salmon_quant$analysis_set$ctrl_panc_covid$deseq$fgsea$covid_v_ctrl),
          .id = 'condition') %>% 
  select(condition, pathway, padj, NES, leadingEdge, size) %>% 
  arrange(NES) %>% 
  mutate(rank = row_number()) %>% 
  group_by(pathway) %>% 
  mutate(pathway = str_replace_all(pathway, 'HALLMARK ', ''),
         # pathway = str_replace_all(pathway, ' ', '\n'),
         observed = length(unlist(leadingEdge)),
         NES = round(NES, 3)) %>% 
  # arrange(-NES, padj) %>% 
  ggplot(aes(NES, rank, color = -log10(padj))) + 
  geom_point(size = 2.5) +
  geom_segment(aes(x=0, xend=NES, y=rank, yend=rank)) +
  geom_text(aes(label = paste0(condition),
                x = NES - 0.75 * sign(NES)),
            nudge_y = 0.15,
            color = 'black',
            size = txt.mm_to_pts(1)) +
  geom_text(aes(label = pathway, 
              x = - .55 * sign(NES)), 
              nudge_y = 0.025,
              color = 'black', 
              size = txt.mm_to_pts(1)) + 
  scale_color_viridis_c() +
  xlim(-1,2.5) +
  # ggforce::facet_col(~condition, scales = 'free_y') +
  ylab('gene set') + xlab('GSEA NES') +
  geom_vline(xintercept = 0, alpha = 0.3, linetype = 'dashed') +
  theme(legend.position = c(0.99, 0.17),
        axis.ticks.y = element_blank(),
        axis.text.y = element_blank()) -> fig.S4A.plt


```
## Fig_S4B
```{r}
de_compare_list <- list('panc' = ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out,
                        'covid' = ucsc.rmsk.salmon_quant$analysis_set$covid_ctrl$deseq$de_out)

purrr::imap(de_compare_list, function(de, de_name) {
  
  de %>% 
    filter(gene %in% msig.df$HALLMARK_MYC_TARGETS_V1,
           padj < 0.05,
           abs(log2FoldChange) > 0.5) %>% 
    mutate(id = de_name)
  
}) %>% bind_rows() -> de_upset.df

de_upset.df %>% 
  group_by(gene) %>% 
  summarize(context = list(id)) %>% 
  ggplot(aes(context)) + 
  geom_bar(fill = 'white', color = 'black') + 
  ggupset::scale_x_upset() +
  geom_text(stat='count', aes(label=after_stat(count)), nudge_y = -2) + 
  xlab('') + ylab('DE MYC Target V1') -> fig.S2B.plt


de_upset.df %>% 
  group_by(id) %>% 
  top_n(5, wt = log2FoldChange) %>% 
  ggplot(aes(reorder(gene, -log2FoldChange), log2FoldChange)) + 
  geom_col(fill = NA, color = 'black') + 
  coord_flip() + 
  ggforce::facet_col(~id, scales = 'free_y') + 
  geom_hline(yintercept = 1.0, linetype = 'dashed', alpha = 0.6) + 
  xlab('')

de_upset.df %>% 
  select(gene, log2FoldChange, id) %>% 
  spread(id, log2FoldChange) %>% 
  ggplot(aes(covid, panc, label = gene)) + 
  geom_point() + 
  geom_hline(yintercept = 0, linetype = 'dashed', alpha = 0.6) +
  geom_vline(xintercept = 0, linetype = 'dashed', alpha = 0.6) + 
  xlab('covid MYC Targets V1 log2FC') + ylab('panc MYC Targets V1 log2FC') +
  ggrepel::geom_text_repel() -> fig.S2B_2.plt

```
## Fig_S4C
```{r}
import::from(.from = 'SummarizedExperiment',
             rowRanges)

rowRanges(ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$gxi) %>% 
  as.data.frame() %>% 
  filter(seqnames == 'chrM') %>% 
  select(group_name) %>% 
  unlist() %>% unname() -> chrM.filter.list

ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out %>% 
# ucsc.rmsk.salmon_quant$analysis_set$ctrl_panc_covid$deseq$panc_vs_ctrl %>% 
  merge(reference_meta_data$ucsc_rmsk_insert_info.df, by = 'gene', all.x = T) %>% 
  mutate(family = ifelse(ensg %in% reference_meta_data$gencode_lncRNA_gene_names.df$ensg,
                         'GENCODE_lncRNA', family),
         family = ifelse(is.na(family), 'GENCODE_coding', family),
         clade = ifelse(grepl('GENCODE', family), family, clade),
         clade = ifelse(clade %in% c('GENCODE_coding',
                                     'GENCODE_lncRNA',
                                     'LINE',
                                     'SINE',
                                     'LTR',
                                     'DNA',
                                     'Simple_repeat'), clade, 'other'),
         type = ifelse(grepl('GENCODE',family), 'GENCODE', 'TE'),
         clade = factor(clade, levels = te_aware_annotation_levels[c(1,8)])) %>% 
  filter(ensg %in% chrM.filter.list | grepl('^MT', gene)) -> mt_panc_annotated_rmsk_de

mt_panc_label_list <- mt_panc_annotated_rmsk_de %>% 
  top_n(6, wt = abs(log2FoldChange))

mt_panc_annotated_rmsk_de %>%
  filter(clade == 'GENCODE_coding') %>% 
  ggplot(aes(log2FoldChange, -log10(padj), color = clade)) + 
  geom_point(size = 2, alpha = 0.6, show.legend = FALSE) -> gencode_layer.plt

gencode_layer.plt + 
  geom_point(data = filter(mt_panc_annotated_rmsk_de, clade != 'GENCODE_coding') %>% 
               mutate(clade = factor(clade, levels = te_aware_annotation_levels[c(1,8)])),
             aes(log2FoldChange, -log10(padj), color = clade), size = 2, alpha = 0.8) +
  scale_color_manual(values = te_color_pal[c(1,8)], name = 'biotype/clade') + 
  ggrepel::geom_text_repel(data = mt_panc_annotated_rmsk_de %>% 
             mutate(clade = factor(clade, levels = te_aware_annotation_levels[c(1,8)])),
                    aes(label = ifelse(gene %in% mt_panc_label_list$gene, gene, '')), 
                         show.legend = F, color = 'black', max.overlaps = 1000, size = txt.mm_to_pts(1)) + 
  xlim(-8.5,8.5) + 
  ggtitle('chrM panc DE') -> fig.S2C.plt

fig.S2C.plt
```
## Fig_S4D
```{r}
ucsc.rmsk.salmon_quant$analysis_set$covid_ctrl$deseq$de_out %>% 
  merge(reference_meta_data$ucsc_rmsk_insert_info.df, by = 'gene', all.x = T) %>% 
  mutate(family = ifelse(ensg %in% reference_meta_data$gencode_lncRNA_gene_names.df$ensg,
                         'GENCODE_lncRNA', family),
         family = ifelse(is.na(family), 'GENCODE_coding', family),
         clade = ifelse(grepl('GENCODE', family), family, clade),
         clade = ifelse(clade %in% c('GENCODE_coding',
                                     'GENCODE_lncRNA',
                                     'LINE',
                                     'SINE',
                                     'LTR',
                                     'DNA',
                                     'Simple_repeat'), clade, 'other'),
         type = ifelse(grepl('GENCODE',family), 'GENCODE', 'TE'),
         clade = factor(clade, levels = te_aware_annotation_levels)) %>% 
  filter(ensg %in% chrM.filter.list | grepl('^MT', gene)) -> mt_covid_annotated_rmsk_de

mt_covid_label_list <- mt_covid_annotated_rmsk_de %>% 
  top_n(6, wt = abs(log2FoldChange))

mt_covid_annotated_rmsk_de %>%
  filter(clade == 'GENCODE_coding') %>% 
  ggplot(aes(log2FoldChange, -log10(padj), color = clade)) + 
  geom_point(size = 2, alpha = 0.6, show.legend = FALSE) -> gencode_layer.plt

gencode_layer.plt + 
  geom_point(data = filter(mt_covid_annotated_rmsk_de, clade != 'GENCODE_coding') %>% 
               mutate(clade = factor(clade, levels = te_aware_annotation_levels)),
             aes(log2FoldChange, -log10(padj), color = clade), size = 2, alpha = 0.8) +
  scale_color_manual(values = te_color_pal[c(1,8)], name = 'biotype/clade') + 
  ggrepel::geom_text_repel(data = mt_covid_annotated_rmsk_de %>% 
             mutate(clade = factor(clade, levels = te_aware_annotation_levels)),
                    aes(label = ifelse(gene %in% mt_covid_label_list$gene, gene, '')), 
                         show.legend = F, color = 'black', max.overlaps = 1000, size = txt.mm_to_pts(1)) + 
  xlim(-7.5,7.5) + 
  ggtitle('chrM covid DE') -> fig.S2D.plt

fig.S2D.plt
```
### Fig_S4_patchwork
```{r Fig1}
import::from(.from = 'patchwork', plot_annotation, wrap_plots, guide_area, plot_spacer)


fig_S2_layout <- "
AA
DD
BC
EF
"

wrap_plots(fig.S4A.plt,
           patchwork::wrap_elements(full = fig.S2B.plt),
           fig.S2B_2.plt,
           guide_area(),
           fig.S2C.plt + theme(legend.background = element_blank()) + coord_cartesian(clip = 'off'),
           fig.S2D.plt + theme(legend.background = element_blank()) + coord_cartesian(clip = 'off'),
           design = fig_S2_layout, 
           guides = 'collect',
           heights = c(0.6,0.2,1.2,1)) + plot_annotation(tag_levels = 'A') & 
  theme(plot.tag.position = c(0.01, 0.99), 
        legend.position = 'bottom',
        legend.title = element_text(vjust = 0.5),
        legend.justification = c(0.5, 0.5)) -> fig.S2.patchwork

save.manuscript.panel(figure = 4,
                      figure.dir = figure.dir,
                      name = 'patchwork_full_fig_S4',
                      width = figure_double_column_width,
                      height = figure_depth,
                      plot.in = fig.S2.patchwork)


```

# Fig_S5
## Fig_S5_patchwork
```{r}

panc_annotated_rmsk_de %>% 
  filter(clade %in% c('gencode_coding', 'gencode_lncRNA')) %>% 
  ggplot(aes(log2FoldChange, -log10(padj), color = clade)) + 
  geom_point(size = 2, alpha = 0.6) + 
  scale_color_manual(values = te_color_pal[1:2], name = 'biotype') + 
  ggtitle('panc GENCODE DE') +
  ggrepel::geom_text_repel(aes(label = ifelse(-log10(padj) > 2.5, gene, '')), 
                         show.legend = F, color = 'black', 
             max.overlaps = 30, box.padding = unit(4.5, 'mm'), segment.colour = 'grey') -> plt_S2_2A

covid_annotated_rmsk_de %>% 
  filter(clade %in% c('gencode_coding', 'gencode_lncRNA')) %>% 
  ggplot(aes(log2FoldChange, -log10(padj), color = clade)) + 
  geom_point(size = 2, alpha = 0.6) + 
  scale_color_manual(values = te_color_pal[1:2], name = 'biotype') + 
  ggtitle('covid GENCODE DE') +
  ggrepel::geom_text_repel(aes(label = ifelse(-log10(padj) > 2.5, gene, '')), 
                         show.legend = F, color = 'black', 
             max.overlaps = 60, box.padding = unit(4.5, 'mm'), segment.colour = 'grey') -> plt_S2_2B

import::from(.from = 'patchwork', plot_annotation, wrap_plots, guide_area, plot_spacer)


fig_S2_2_layout <- "
AA
BB
CC
"

wrap_plots(plt_S2_2A,
           guide_area(),
           plt_S2_2B,
           design = fig_S2_2_layout, 
           guides = 'collect',
           heights = c(1,0.125,1)) + plot_annotation(tag_levels = 'A') & 
  theme(plot.tag.position = c(0.01, 0.99), 
        legend.position = 'bottom',
        legend.title = element_text(vjust = 0.5),
        legend.justification = c(0.5, 0.5)) -> fig.S2_2.patchwork

save.manuscript.panel(figure = 5,
                      figure.dir = figure.dir,
                      name = 'patchwork_full_fig_S5',
                      width = figure_double_column_width,
                      height = figure_depth,
                      plot.in = fig.S2_2.patchwork)


```

# Fig_S6
## Fig_SNA
```{r}
import::from(.from = 'patchwork', plot_annotation, wrap_plots, guide_area, plot_spacer)

ctrl.6.1.5_nanopore_hex_to.plt %>%
  group_by(tx, len, biotype.x, sample) %>%
  summarize(read_length = mean(read_length)) %>%
  merge(nanopore_biotype_labels, by.x = 'biotype.x', by.y = 'biotype') %>%
  mutate(biolabel = factor(biolabel, levels = te_aware_annotation_levels),
         condition = ifelse(grepl('ctrl', sample), 'ctrl', 'panc')) %>%
  ggplot(aes(read_length, len, fill = scales::rescale((..count../(sum(..count..) * 1e6))))) +
  geom_hex(bins = 120) +
  annotate(geom = 'text', x = 275, y = 150, label = 'ctrl 6') +
  # geom_vline(xintercept = median(ctrl.6.1.5_nanopore_hex_to.plt$read_length), 
  #            color = 'grey', linetype = 'dashed') +
  xlab('avg. observed fragment size (bp)') + ylab('expected size (bp)') +
  # ggforce::facet_col(~sample) +
  ggtitle('nanopore exp. vs obs. size -- ctrl 6') + 
  scale_fill_viridis_c(option = 'viridis', direction = 1, name = 'scaled cpm', breaks = c(0,1)) +
  theme(legend.position = c(0.95,1), legend.direction = 'horizontal', legend.key.width = unit(1, 'mm'))  +
  guides(fill = guide_colorbar(title.position = 'left', title.vjust = 1)) +
  scale_y_continuous(breaks = c(100,200,300,400), limits = c(75,425)) +
  scale_x_continuous(breaks = c(100,200,300,400), limits = c(75,370)) -> plt_SNA_1.plt


panc.7.1.1_nanopore_hex_to.plt %>% 
  group_by(tx, len, biotype.x, sample) %>% 
  summarize(read_length = mean(read_length)) %>% 
  merge(nanopore_biotype_labels, by.x = 'biotype.x', by.y = 'biotype') %>% 
  mutate(biolabel = factor(biolabel, levels = te_aware_annotation_levels),
         condition = ifelse(grepl('ctrl', sample), 'ctrl', 'panc')) %>% 
  ggplot(aes(read_length, len, fill = scales::rescale((..count../(sum(..count..) * 1e6))))) + 
  geom_hex(bins = 120) +
  # geom_vline(xintercept = median(panc.7.1.1_nanopore_hex_to.plt$read_length), color = 'grey', linetype = 'dashed') +
  annotate(geom = 'text', x = 275, y = 150, label = 'panc 7') +
  xlab('avg. observed fragment size (bp)') + ylab('expected size (bp)') +
  # ggforce::facet_col(~sample) +
    ggtitle('nanopore exp. vs obs. size -- ctrl 6') + 
  scale_fill_viridis_c(option = 'viridis', direction = 1, name = 'scaled cpm', guide = 'none') +
  scale_y_continuous(breaks = c(100,200,300,400), limits = c(75,425)) +
  scale_x_continuous(breaks = c(100,200,300,400), limits = c(75,370)) -> plt_SNA_3.plt
  
panel_SNA_layout <- "
AB
"

wrap_plots(plt_SNA_1.plt + xlab(''),
           plt_SNA_3.plt + ylab('') + xlab(''),
           design = panel_SNA_layout) -> panel.SNA.patchwork
```
## Fig_SNA_dep
```{r}
illumina_biotype_to.plt %>% 
  filter(biotype == 'SINE') %>% 
  group_by(tx, len, biotype, condition) %>% 
  mutate(len = as.numeric(len)) %>%
  summarize(read_length = median(abs(read_length))) -> tmp

tmp %>% 
  filter(condition == 'ctrl') -> ctrl_hex_to.plt

ctrl_hex_to.plt %>% 
  ggplot(aes(read_length, len, fill = scales::rescale((..count../(sum(..count..) * 1e6))))) + 
  geom_hex(bins = 120) +
  # geom_vline(xintercept = median(ctrl_hex_to.plt$read_length), color = 'grey', linetype = 'dashed') +
  annotate(geom = 'text', x = 275, y = 175, label = 'ctrl') +
  xlab('median observed SINE fragment size (bp)') + ylab('expected size (bp)') +
  # ggforce::facet_col(~sample) +
  scale_fill_viridis_c(option = 'viridis', direction = 1, name = 'scaled cpm', breaks = c(0,1), guide = 'none') +
  scale_y_continuous(breaks = c(50,100,200,300,400, 500), limits = c(0,500)) +
  scale_x_continuous(breaks = c(50,100,200,300,400, 500), limits = c(0,350)) -> plt_SNA_1.plt

tmp %>% 
  filter(condition == 'panc') -> panc_hex_to.plt

panc_hex_to.plt %>% 
  ggplot(aes(read_length, len, fill = scales::rescale((..count../(sum(..count..) * 1e6))))) + 
  geom_hex(bins = 120) +
  # geom_vline(xintercept = median(panc_hex_to.plt$read_length), color = 'grey', linetype = 'dashed') +
  annotate(geom = 'text', x = 275, y = 175, label = 'panc') +
  xlab('avg. observed fragment length') + ylab('expected insertion length') +
  # ggforce::facet_col(~sample) +
  scale_fill_viridis_c(option = 'viridis', direction = 1, name = 'scaled cpm', breaks = c(0,1)) +
  scale_y_continuous(breaks = c(50,100,200,300,400, 500), limits = c(0,500)) +
  scale_x_continuous(breaks = c(50,100,200,300,400, 500), limits = c(0,350)) +
  theme(legend.position = c(0.95,0.25), legend.direction = 'horizontal', legend.key.width = unit(1, 'mm')) +
  guides(fill = guide_colorbar(title.position = 'left', title.vjust = 1)) -> plt_SNA_2.plt

tmp %>% 
  filter(condition == 'covid') -> covid_hex_to.plt

covid_hex_to.plt %>% 
  ggplot(aes(read_length, len, fill = scales::rescale((..count../(sum(..count..) * 1e6))))) + 
  geom_hex(bins = 120) +
  # geom_vline(xintercept = median(covid_hex_to.plt$read_length), color = 'grey', linetype = 'dashed') +
  annotate(geom = 'text', x = 275, y = 175, label = 'covid') +
  xlab('avg. observed fragment length') + ylab('expected insertion length') +
  # ggforce::facet_col(~sample) +
  scale_fill_viridis_c(option = 'viridis', direction = 1, name = 'scaled cpm', breaks = c(0,1), guide = 'none') +
  scale_y_continuous(breaks = c(50,100,200,300,400, 500), limits = c(0,500)) +
  scale_x_continuous(breaks = c(50,100,200,300,400, 500), limits = c(0,350)) -> plt_SNA_3.plt

p4 <- ggplot(data.frame(l = plt_SNA_1.plt$labels$x, x = 1, y = 1)) +
      geom_text(aes(x, y, label = l), angle = 0, size = txt.mm_to_pts(0.95)) + 
      theme_void() +
      coord_cartesian(clip = "off")
  
panel_SNA_layout <- "
ABC
"

wrap_plots(plt_SNA_1.plt + xlab(''),
           plt_SNA_2.plt + xlab('') + ylab(''),
           plt_SNA_3.plt + xlab('') + ylab(''),
           heights = c(1),
           design = panel_SNA_layout) -> panel.SNA.patchwork
```
## Fig_SNB
```{r}
illumina_biotype_to.plt %>% 
  filter(biotype == 'SINE',
         sample_name %in% c('panc.6.1.1', 'panc.7.1.1', 'ctrl.6.1.5')) %>% 
  group_by(tx, len, biotype, sample_name) %>% 
  mutate(len = as.numeric(len)) %>% 
  summarize(read_length = median(abs(read_length))) -> tmp

tmp %>%
  filter(sample_name == 'ctrl.6.1.5') -> ctrl_hex_to.plt

ctrl_hex_to.plt %>%
  ggplot(aes(read_length, len, fill = scales::rescale((..count../(sum(..count..) * 1e6))))) +
  geom_hex(bins = 120) +
  annotate(geom = 'text', x = 250, y = 175, label = 'ctrl 6') +
  # geom_vline(xintercept = median(ctrl_hex_to.plt$read_length), color = 'grey', linetype = 'dashed') +
  xlab('avg. observed fragment size (bp)') + ylab('expected size (bp)') +
  ggtitle('illumina exp. vs obs. size') + 
  scale_y_continuous(breaks = c(100,200,300,400), limits = c(75,425)) +
  scale_x_continuous(breaks = c(100,200,300,400), limits = c(75,370)) +
  scale_fill_viridis_c(option = 'viridis', direction = 1, name = 'scaled cpm', breaks = c(0,1), guide = 'none') -> plt_SNB_1.plt
  

tmp %>% 
  filter(sample_name == 'panc.6.1.1') -> panc.6.1.1_hex_to.plt

panc.6.1.1_hex_to.plt %>% 
  mutate(bin = ifelse(read_length > 180, 'large', 'small')) %>% 
  group_by(tx) %>%
  ggplot(aes(read_length, len, fill = scales::rescale((..count../(sum(..count..) * 1e6))))) + 
  geom_hex(bins = 120) +
  annotate(geom = 'text', x = 250, y = 175, label = 'panc 6') +
  # geom_vline(xintercept = median(panc.6.1.1_hex_to.plt$read_length), color = 'grey', linetype = 'dashed') +
  xlab('avg. observed fragment size (bp)') + ylab('expected size (bp)') +
  scale_y_continuous(breaks = c(100,200,300,400), limits = c(75,425)) +
  scale_x_continuous(breaks = c(100,200,300,400), limits = c(75,370)) +
  scale_fill_viridis_c(option = 'viridis', direction = 1, name = 'scaled cpm', breaks = c(0,1), guide = 'none') -> plt_SNB_2.plt


tmp %>%
  filter(sample_name == 'panc.7.1.1') -> panc.7.1.1_hex_to.plt

panc.7.1.1_hex_to.plt %>% 
  ggplot(aes(read_length, len, fill = scales::rescale((..count../(sum(..count..) * 1e6))))) + 
  geom_hex(bins = 120) +
  annotate(geom = 'text', x = 250, y = 175, label = 'panc 7') +
  # geom_vline(xintercept = median(panc.7.1.1_hex_to.plt$read_length), color = 'grey', linetype = 'dashed') +
  xlab('avg. observed fragment size (bp)') + ylab('expected size (bp)') +
  scale_y_continuous(breaks = c(100,200,300,400), limits = c(75,425)) +
  scale_x_continuous(breaks = c(100,200,300,400), limits = c(75,370)) +
  scale_fill_viridis_c(option = 'viridis', direction = 1, name = 'scaled cpm', breaks = c(0,1), guide = 'none') -> plt_SNB_3.plt

p4 <- ggplot(data.frame(l = plt_SNB_3.plt$labels$x, x = 1, y = 1)) +
      geom_text(aes(x, y, label = l), angle = 0, size = txt.mm_to_pts(0.95)) + 
      theme_void() +
      coord_cartesian(clip = "off")
  
panel_SNB_layout <- "
ABC
DDD
"

wrap_plots(plt_SNB_1.plt + xlab(''),
           plt_SNB_2.plt + xlab('') + ylab(''),
           plt_SNB_3.plt + xlab('') + ylab(''),
           p4,
           heights = c(1,0.06),
           design = panel_SNB_layout) -> panel.SNB.patchwork

```
## Fig_SNC
```{r}
tmp_mean_read_len.df %>%
  group_by(biotype, condition) %>%
  filter(n() == 10, ! biotype %in% c('polymorphic_pseudogene', 'TR_C_gene', 'TR_V_gene', 'IG_V_gene', 'IG_C_gene',
                                   'TEC', 'non_stop_decay', 'nonsense_mediated_decay'),
         condition != 'panc') %>% 
  mutate(biotype = ifelse(biotype == 'lncRNA', 'gencode_lncRNA', biotype),
         biotype = ifelse(biotype == 'protein_coding', 'gencode_coding', biotype)) %>% 
  mutate(biotype = factor(biotype, levels = c('SINE', 'LINE', 'LTR', 'gencode_lncRNA', 'retained_intron', 'gencode_coding'))) %>% 
  ggplot(aes(condition, read_length, color = condition)) +
  geom_boxplot(outlier.colour = NA) + 
  geom_jitter(width = 0.24, alpha = 0.2) +
  facet_wrap(~biotype, scales = 'free', ncol = 3, strip.position = 'bottom') +
  scale_color_manual(values = identity_color_pal, guide = 'none') + 
  ggpubr::stat_compare_means(method = 'wilcox', comparisons = list(c('covid', 'ctrl')), size = txt.mm_to_pts(0.8)) +
  coord_cartesian(clip = 'off') + 
  theme(strip.placement = 'inside') +
  xlab('') + ylab('median fragment size (bp)') -> plt_SNC.plt
```
## Fig_SND
```{r}
tmp_mean_read_len.df %>% 
  filter(condition != 'panc',
         ! sample %in% qc_fails) %>% 
  pivot_wider(names_from = biotype, values_from = read_length) %>% 
  select(-condition) %>% 
  column_to_rownames('sample') -> read_len_pca.tmp
  
pca_tmp_in_sd <- apply(read_len_pca.tmp, 2, sd)
pca_tmp_in_zero_sd <- read_len_pca.tmp[,pca_tmp_in_sd!=0 & !is.na(pca_tmp_in_sd)]
non_zero_pca <- prcomp(pca_tmp_in_zero_sd, center = T, scale. = T, rank. = 50)

summary(non_zero_pca)
pcs.of.interest <- apply(non_zero_pca$x, 2, var)
pcs.props <-  pcs.of.interest/sum(pcs.of.interest)
cumsum(pcs.props)[c(1,2)]
print(as.data.frame(pcs.props) %>% 
        rownames_to_column('pc') %>% 
        mutate(pc = as.numeric(sub('PC', '', pc))) %>% 
        ggplot(aes(pc, pcs.props)) + 
        geom_col())

pca.out <- as.data.frame(non_zero_pca$x) %>% 
  rownames_to_column('sample') %>% 
  merge(tmp_mean_read_len.df %>% select(sample, condition) %>% ungroup() %>% distinct(), by = 'sample') %>% 
  ungroup() %>% 
  select(-biotype) %>% 
  distinct()

pca.out.summary <- 
  as.data.frame(summary(non_zero_pca)$importance) %>% 
  rownames_to_column('metric') %>% 
  gather(pc, value, -metric) %>% 
  spread(metric, value)

pca.out %>% 
  ggplot(aes(PC1, PC2,
             color = condition)) +
  geom_point(size = rel(2), alpha = 1) + 
  # scale_shape_manual(values = c(21,25)) +
  xlab(paste('PC1', round(cumsum(pcs.props)[1], digits = 3), sep = ' ')) +
  ylab(paste('PC2', round(cumsum(pcs.props)[2] - cumsum(pcs.props)[1], digits = 3), sep = ' ')) +
  geom_vline(xintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) +
  geom_hline(yintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) + 
  scale_color_manual(values = identity_color_pal[c(1,3)]) + 
  annotate(geom = 'text', x = -5.5, y = 2.25, label = 'size-aware', size = txt.mm_to_pts(1)) +
  theme(legend.position ='top', legend.direction = 'horizontal') -> plt_SND.plt
```
### Fig_SNanopore_patchwork
```{r}
import::from(.from = 'patchwork', plot_annotation, wrap_plots, guide_area, plot_spacer, wrap_elements)


fig_SN_layout <- "
AA
BB
CC
DD
"

wrap_plots(wrap_elements(full = panel.SNA.patchwork),
           wrap_elements(full = panel.SNB.patchwork),
           plt_SNC.plt,
           plt_SND.plt,
           heights = c(1.25,1.25,1,0.5),
           design = fig_SN_layout) +
  plot_annotation(tag_levels = 'A') -> fig.SN.patchwork

save.manuscript.panel(figure = 6,
                      figure.dir = figure.dir,
                      name = 'patchwork_full_fig_S6',
                      width = figure_double_column_width,
                      height = figure_depth,
                      plot.in = fig.SN.patchwork)
```

# Fig_S6
## Fig_SN2_patchwork
```{r}
panc.7.1.1_nanopore_tpm <- 
  read_tsv(file.path(output_data.dir, 
                     'nanopore', 'gencode_39_rmsk_nanopore_abundance', 'panc.7.1.1_gene_abundance.tsv'))

panc.6.1.1_nanopore_tpm <- 
  read_tsv(file.path(output_data.dir, 
                     'nanopore', 'gencode_39_rmsk_nanopore_abundance', 'panc.6.1.1_gene_abundance.tsv'))

ctrl.6.1.5_nanopore_tpm <- 
  read_tsv(file.path(output_data.dir, 
                     'nanopore', 'gencode_39_rmsk_nanopore_abundance', 'ctrl.6.1.5_gene_abundance.tsv'))

import::from(.from = 'SummarizedExperiment', assay)

count_matrix.df <- 
  as.data.frame(as.matrix(assay(ucsc.rmsk.salmon_quant$analysis_set$ctrl_panc_covid$gxi, 'counts')))

count_matrix.df[, 'panc.7.1.1', drop = F] %>% rownames_to_column('Gene ID') %>% 
  mutate(panc.7.1.1_tpm = panc.7.1.1*1e6/(sum(panc.7.1.1))) %>% 
  merge(panc.7.1.1_nanopore_tpm %>% select(`Gene ID`, TPM), by = 'Gene ID') %>% 
  dplyr::rename('panc.7.1.1_nanopore_tpm' = TPM) %>%
  group_by(`Gene ID`) %>% 
  summarize(panc_7_tpm = mean(panc.7.1.1_tpm),
            panc_7_nanopore_tpm = mean(panc.7.1.1_nanopore_tpm)) %>% 
  distinct() -> panc_7.1.1_tpm

count_matrix.df[, 'panc.6.1.1', drop = F] %>% rownames_to_column('Gene ID') %>%
  mutate(panc.6.1.1_tpm = panc.6.1.1 *1e6/(sum(panc.6.1.1))) %>% 
  merge(panc.6.1.1_nanopore_tpm %>% select(`Gene ID`, TPM), by = 'Gene ID') %>% 
  dplyr::rename('panc.6.1.1_nanopore_tpm' = TPM) %>% 
  group_by(`Gene ID`) %>% 
  summarize(panc_6_tpm = mean(panc.6.1.1_tpm),
            panc_6_nanopore_tpm = mean(panc.6.1.1_nanopore_tpm)) %>% 
  distinct() -> panc_6.1.1_tpm

count_matrix.df[, 'ctrl.6.1.5', drop = F] %>% rownames_to_column('Gene ID') %>%
  mutate(ctrl.6.1.5_tpm = ctrl.6.1.5 *1e6/(sum(ctrl.6.1.5))) %>% 
  merge(ctrl.6.1.5_nanopore_tpm %>% select(`Gene ID`, TPM), by = 'Gene ID') %>% 
  dplyr::rename('ctrl.6.1.5_nanopore_tpm' = TPM) %>% 
  group_by(`Gene ID`) %>% 
  summarize(ctrl_6_tpm = mean(ctrl.6.1.5_tpm),
            ctrl_6_nanopore_tpm = mean(ctrl.6.1.5_nanopore_tpm)) %>% 
  distinct() -> ctrl.6.1.5_tpm


merge(panc_7.1.1_tpm, panc_6.1.1_tpm, by = 'Gene ID') %>% 
  merge(ctrl.6.1.5_tpm, by = 'Gene ID') -> panc_compare.df


panc_compare.df %>% 
  merge(reference_meta_data$ucsc_rmsk_insert_info.df, by.x = 'Gene ID', by.y = 'gene', all.x = T) %>% 
  mutate(family = ifelse(`Gene ID` %in% reference_meta_data$gencode_lncRNA_gene_names.df$ensg,
                         'GENCODE_lncRNA', family),
         family = ifelse(is.na(family), 'GENCODE_coding', family),
         clade = ifelse(grepl('GENCODE', family), family, clade),
         clade = ifelse(clade %in% c('GENCODE_coding',
                                     'GENCODE_lncRNA',
                                     'LINE',
                                     'SINE',
                                     'LTR',
                                     'DNA',
                                     'Simple_repeat'), clade, 'other'),
         type = ifelse(grepl('GENCODE',family), 'GENCODE', 'TE'),
         clade = factor(clade, levels = te_aware_annotation_levels)) -> panc_compare_annotated.df

# lm(panc_6_tpm ~ panc_7_tpm, data = panc_compare_annotated.df) %>% 
#   summary() -> panc_lm_out
# 
# b <- round(panc_lm_out$coefficients[1,1], 2)
# m <- round(panc_lm_out$coefficients[2,1], 2)
# r2 <- round(panc_lm_out$adj.r.squared, 2)
# 
# panc_compare_annotated.df %>% 
#   ggplot(aes(log2(panc_6_tpm + 1), log2(panc_7_tpm + 1), color = clade, fill = clade)) + 
#   geom_hex(alpha = 0.4, bins = 250) +
#   annotate(geom = 'text', x = 3, y = 15, fontface = 'italic',
#            label = paste0('y', ' ', ' = ',b,' + ',m,'x; adj. R2 = ', r2), 
#            size = rel(3.75), alpha = 0.8) + 
#   scale_color_manual(values = te_color_pal) + 
#   scale_fill_manual(values = te_color_pal) -> plt_SN2A.plt

lm(panc_7_nanopore_tpm ~ panc_7_tpm, data = panc_compare_annotated.df) %>% 
  summary() -> panc_lm_out

b <- round(panc_lm_out$coefficients[1,1], 2)
m <- round(panc_lm_out$coefficients[2,1], 2)
r2 <- round(panc_lm_out$adj.r.squared, 2)

panc_compare_annotated.df %>% 
  ggplot(aes(log2(panc_7_nanopore_tpm + 1), log2(panc_7_tpm + 1), color = clade, fill = clade)) + 
  geom_hex(alpha = 0.4, bins = 250) +
  annotate(geom = 'text', x = 6, y = 16, fontface = 'italic',
         label = paste0('y', ' ', ' = ',b,' + ',m,'x; adj. R2 = ', r2), 
         size = rel(3.75), alpha = 0.8) + 
  scale_color_manual(values = te_color_pal) + 
  scale_fill_manual(values = te_color_pal)-> plt_SN2B.plt

lm(panc_6_nanopore_tpm ~ panc_6_tpm, data = panc_compare_annotated.df) %>% 
  summary() -> panc_lm_out

b <- round(panc_lm_out$coefficients[1,1], 2)
m <- round(panc_lm_out$coefficients[2,1], 2)
r2 <- round(panc_lm_out$adj.r.squared, 2)

panc_compare_annotated.df %>% 
  ggplot(aes(log2(panc_6_nanopore_tpm + 1), log2(panc_6_tpm + 1), color = clade, fill = clade)) + 
  geom_hex(alpha = 0.4, bins = 250) +
  annotate(geom = 'text', x = 6, y = 16, fontface = 'italic',
         label = paste0('y', ' ', ' = ',b,' + ',m,'x; adj. R2 = ', r2), 
         size = rel(3.75), alpha = 0.8) +
  scale_color_manual(values = te_color_pal) + 
  scale_fill_manual(values = te_color_pal) -> plt_SN2C.plt

lm(ctrl_6_nanopore_tpm ~ ctrl_6_tpm, data = panc_compare_annotated.df) %>% 
  summary() -> panc_lm_out

b <- round(panc_lm_out$coefficients[1,1], 2)
m <- round(panc_lm_out$coefficients[2,1], 2)
r2 <- round(panc_lm_out$adj.r.squared, 2)

panc_compare_annotated.df %>% 
  ggplot(aes(log2(ctrl_6_nanopore_tpm + 1), log2(ctrl_6_tpm + 1), color = clade, fill = clade)) + 
  geom_hex(alpha = 0.4, bins = 250) +
  annotate(geom = 'text', x = 6, y = 16, fontface = 'italic',
         label = paste0('y', ' ', ' = ',b,' + ',m,'x; adj. R2 = ', r2), 
         size = rel(3.75), alpha = 0.8) +
  scale_color_manual(values = te_color_pal) + 
  scale_fill_manual(values = te_color_pal) -> plt_SN2D.plt

# lm(panc_6_nanopore_tpm ~ panc_7_nanopore_tpm, data = panc_compare_annotated.df) %>% 
#   summary() -> panc_lm_out
# 
# b <- round(panc_lm_out$coefficients[1,1], 2)
# m <- round(panc_lm_out$coefficients[2,1], 2)
# r2 <- round(panc_lm_out$adj.r.squared, 2)
# 
# panc_compare_annotated.df %>% 
#   ggplot(aes(log2(panc_6_nanopore_tpm + 1), log2(panc_7_nanopore_tpm + 1), color = clade, fill = clade)) + 
#   geom_hex(alpha = 0.4, bins = 250) +
#   annotate(geom = 'text', x = 3, y = 15, fontface = 'italic',
#          label = paste0('y', ' ', ' = ',b,' + ',m,'x; adj. R2 = ', r2), 
#          size = rel(3.75), alpha = 0.8) +
#   scale_color_manual(values = te_color_pal) + 
#   scale_fill_manual(values = te_color_pal) -> plt_SN2D.plt


panc_compare_annotated.df %>% 
  select('gene' = 'Gene ID', contains('nanopore'), clade) %>% 
  gather(sample, tpm, -gene, -clade) %>% 
  mutate(sample = str_replace(sample, '_nanopore_tpm', ''),
         sample = str_replace(sample, '_', ' ')) %>% 
  filter(tpm > 0.17) %>% 
  group_by(sample, clade) %>% 
  summarize(detected = n()) %>% 
  ggplot(aes(sample, detected, fill = sample)) + 
  geom_col() + 
  xlab('') + ylab('detected genes/elements') +
  scale_fill_manual(values = viridisLite::plasma(20)[c(1,8,10)]) +
  facet_wrap(~clade, scales = 'free_y', ncol = 2, strip.position = 'bottom') -> plt_SN2E.plt

import::from(.from = 'patchwork', plot_annotation, wrap_plots, guide_area, plot_spacer, wrap_elements)


fig_SN2_layout <- "
AA
BB
CC
DD
"

wrap_plots(plt_SN2B.plt + ggtitle('panc 7 nanopore v illumina'),
           guide_area(),
           # plt_SN2B.plt + ggtitle('nanopore v illumina'),
           plt_SN2C.plt + ggtitle('panc 6 nanopore v illumina'),
           plt_SN2D.plt + ggtitle('ctrl 6 nanopore v illumina'),
           # plt_SN2E.plt,
           guides = 'collect',
           design = fig_SN2_layout,
           heights = c(1,0.25,1,1)) +
  plot_annotation(tag_levels = 'A') & 
  theme(legend.position = 'bottom', 
        legend.justification = c(0.5, 0.5))-> fig.SN2.patchwork

save.manuscript.panel(figure = 6,
                      figure.dir = figure.dir,
                      name = 'patchwork_full_fig_S6',
                      width = figure_double_column_width,
                      height = figure_depth,
                      plot.in = fig.SN2.patchwork)

fig_S6_2_layout <- "
AA
AA
"

wrap_plots(plt_SN2E.plt,
           guides = 'collect',
           design = fig_S6_2_layout,
           heights = c(1,0.25)) &
  # plot_annotation(tag_levels = 'A') & 
  theme(legend.position = 'bottom', 
        legend.justification = c(0.5, 0.5))-> fig.S6_2.patchwork

save.manuscript.panel(figure = 6,
                      figure.dir = figure.dir,
                      name = 'patchwork_full_fig_S6_2',
                      width = figure_double_column_width,
                      height = figure_depth,
                      plot.in = fig.S6_2.patchwork)


```

# Figure_process_aware
## Fig_4A
```{r}
process.aware.salmon_quant$analysis_set$ctrl_panc_covid$unsplice_data.df %>% 
  merge(process.aware.salmon_quant$analysis_set$ctrl_panc_covid$quant_meta, by = 'sample') %>% 
  # filter(condition != 'covid') %>% 
  mutate(condition = factor(condition, levels = c('panc', 'ctrl', 'covid'))) %>% 
  ggplot(aes(condition, unsplice_rate, color = condition)) + 
  geom_boxplot(fill = NA, outlier.colour = NA) + 
  geom_point() + 
  scale_color_manual(values = identity_color_pal, guide = 'none') + 
  theme(legend.position = c(0.99,0.99)) + 
  coord_cartesian(clip = 'off') +
  ggpubr::stat_compare_means(method = 'wilcox', 
                             comparisons = list(c('panc', 'ctrl'), c('covid', 'ctrl'))) + 
  xlab('') + ylab('detection rate of un-spliced genes') -> plt_4A.plt
```
## Fig_4B
```{r}
SummarizedExperiment::assay(process.aware.salmon_quant$analysis_set$ctrl_panc_covid$gxi.split,
                            'unspliced') %>% as.data.frame() -> unsplice_counts_pca.df

unsplice_counts_pca.df %>% 
  select(!contains(qc_fails)) %>% 
  t() %>% 
  as.data.frame() -> unsplice_counts_pca.tmp
  
pca_tmp_in_sd <- apply(unsplice_counts_pca.tmp, 2, sd)
pca_tmp_in_zero_sd <- unsplice_counts_pca.tmp[,pca_tmp_in_sd!=0 & !is.na(pca_tmp_in_sd)]
non_zero_pca <- prcomp(pca_tmp_in_zero_sd, center = T, scale. = T, rank. = 50)

summary(non_zero_pca)
pcs.of.interest <- apply(non_zero_pca$x, 2, var)
pcs.props <-  pcs.of.interest/sum(pcs.of.interest)
cumsum(pcs.props)[c(1,2)]
print(as.data.frame(pcs.props) %>% 
        rownames_to_column('pc') %>% 
        mutate(pc = as.numeric(sub('PC', '', pc))) %>% 
        ggplot(aes(pc, pcs.props)) + 
        geom_col())

pca.out <- as.data.frame(non_zero_pca$x) %>% 
  rownames_to_column('sample') %>% 
  merge(process.aware.salmon_quant$analysis_set$ctrl_panc_covid$quant_meta %>% 
          select(sample, condition) %>% 
          ungroup() %>% distinct(), by = 'sample')

pca.out.summary <- 
  as.data.frame(summary(non_zero_pca)$importance) %>% 
  rownames_to_column('metric') %>% 
  gather(pc, value, -metric) %>% 
  spread(metric, value)

pca.out %>% 
  ggplot(aes(PC1, PC2,
             color = condition)) +
  geom_point(size = rel(2), alpha = 1) + 
  # scale_shape_manual(values = c(21,25)) +
  xlab(paste('PC1', round(cumsum(pcs.props)[1], digits = 3), sep = ' ')) +
  ylab(paste('PC2', round(cumsum(pcs.props)[2] - cumsum(pcs.props)[1], digits = 3), sep = ' ')) +
  geom_vline(xintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) +
  geom_hline(yintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) + 
  scale_color_manual(values = identity_color_pal) + 
  annotate(geom = 'text', x = -55, y = -160, label = 'intron-aware', size = txt.mm_to_pts(1)) +
  theme(legend.position = c(0.95,1), legend.direction = 'horizontal') -> plt_4B.plt
```
### Fig_4_patchwork
```{r}
import::from(.from = 'patchwork', plot_annotation, wrap_plots, guide_area, plot_spacer, wrap_elements)


fig_4_layout <- "
AAAA
CCCC
CCCC
"

wrap_plots(plt_4A.plt,
           plot_spacer(),
           design = fig_4_layout) +
  plot_annotation(tag_levels = 'A') -> fig.4.patchwork

save.manuscript.panel(figure = 4,
                      figure.dir = figure.dir,
                      name = 'patchwork_full_fig_4',
                      width = figure_double_column_width,
                      height = figure_depth,
                      plot.in = fig.4.patchwork)
```

```{r}
process.aware.salmon_quant$analysis_set$ctrl_panc_covid$unsplice_data.df %>% 
  merge(process.aware.salmon_quant$analysis_set$ctrl_panc_covid$quant_meta, by = 'sample') %>%
  mutate(condition = factor(condition, levels = c('panc', 'ctrl', 'covid'))) %>% 
  ggplot(aes(unsplice_rate, star_multimapped_percent, color = condition)) + 
  geom_point() + 
  ggforce::facet_col(~condition) + 
  geom_smooth(method = 'lm', se = F, linetype = 'dashed') + 
  scale_color_manual(values = identity_color_pal, guide = 'none')

SummarizedExperiment::assay(process.aware.salmon_quant$analysis_set$panc_ctrl$gxi, 'counts')  %>% 
  as.data.frame() %>% 
  rownames_to_column('ensg') %>% 
  filter(ensg == 'ENSG00000230021.10-I') %>% 
  gather(sample, count, -ensg) %>% 
  mutate(condition = ifelse(grepl('panc', sample), 'panc', 'ctrl')) %>% 
  ggplot(aes(condition, count)) + 
  geom_boxplot()
  


process.aware.salmon_quant$analysis_set$panc_ctrl$deseq$de_out %>% 
  merge(salmon_quant$analysis_set$covid_ctrl$deseq$de_out, by = 'gene') %>% 
  # filter(grepl('-I', ensg)) %>% View()
  ggplot(aes(log2FoldChange.x, log2FoldChange.y, label = gene)) +
  geom_point() + 
  ggrepel::geom_text_repel() + 
  geom_hline(yintercept = 0, linetype = 'dashed') +
  geom_vline(xintercept = 0, linetype = 'dashed')

unsplice_counts_pca.df %>% 
  rownames_to_column('ensg') %>% 
  merge(reference_meta_data$gencode_tx_to_gene.df %>% select(ensg, gene) %>% distinct(), by = 'ensg')

```

# Fig_modeling
## Fig_Modeling -- multi_te
```{r}
# install.packages('glmnet')
# install.packages('doMC')
# install.packages('pROC')
# install.packages('caret')

library(glmnet)
library(pROC)

softmax.logit.calc <- function(y_hat) { exp(y_hat)/sum(exp(y_hat)) }

doMC::registerDoMC(cores = 6)

unique(c(filter(ucsc.rmsk.salmon_quant$analysis_set$covid_ctrl$deseq$de_out, 
                padj < 0.01, !grepl('^ENSG', ensg), !ensg %in% chrM.filter.list)$ensg,
  filter(ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out, 
         padj < 0.01, !grepl('^ENSG', ensg), !ensg %in% chrM.filter.list)$ensg)) ->
  de_ensg_names

x_in <- t(ucsc.rmsk.salmon_quant$analysis_set$ctrl_panc_covid$deseq$norm_counts.df)

x_in_filt <- x_in[, colnames(x_in) %in% c(de_ensg_names)]

y_in <- ucsc.rmsk.salmon_quant$analysis_set$ctrl_panc_covid$deseq$scaled_quant_meta_for_de.df %>% 
  select(sample, condition) %>% 
  mutate(condition = ifelse(condition == 'ctrl', 0, 
                            ifelse(condition == 'panc', 1, 2))) %>% 
  column_to_rownames('sample')

y <- y_in[match(rownames(x_in_filt), rownames(y_in)), ]

lapply(seq(0,1,0.1), function(alpha) {
  set.seed(123)

  mult_model <- 
    cv.glmnet(x = x_in_filt,
              y = y,
              family = 'multinomial',
              type.measure = 'class',
              nfolds = nrow(x_in_filt), 
              alpha = alpha, 
              parallel = TRUE,
              keep = TRUE)
  
  
  coef_tmp <- Reduce(cbind, coef(mult_model, s = mult_model$lambda.1se))
  
  beta <- coef_tmp[apply(coef_tmp != 0, 1, any), ]
  
  colnames(beta) <- names(coef(mult_model, s = mult_model$lambda.1se))
  
  beta
  
  inverse.logit.calc <- function(y_hat) { exp(y_hat)/(1+exp(y_hat)) }
  softmax.logit.calc <- function(y_hat) { exp(y_hat)/sum(exp(y_hat)) }
  
  y_hat_matrix <- mult_model$fit.preval[,,which(mult_model$lambda == mult_model$lambda.min)]
  
  # multi-nomial so need to apply softmax to each row (sample)
  mult_response_matrix <- 
    t(apply(y_hat_matrix, 1, softmax.logit.calc))
  
  mult_response_matrix
  
  # which column is the max -- so need to pull from column names
  y_hat_predictions <- 
    colnames(mult_response_matrix)[apply(mult_response_matrix, 1, which.max)]
  
  multi_roc <- 
    pROC::multiclass.roc(response = y, 
                         predictor = mult_response_matrix, 
                         levels = c('0','1','2'))
  
  # multi_roc$rocs[[3]]$specificities
  # multi_roc$rocs[[3]]$sensitivities
  # multi_roc$rocs[[3]]$thresholds
  # plot(multi_roc$rocs[[2]])
  
  caret::confusionMatrix(data = as.factor(y_hat_predictions),
                         reference = as.factor(y)) -> cm_out
  
  ret_lst <- list(cm_out)
  names(ret_lst) <- alpha
  
  ret_lst
  
}) %>% purrr::flatten() -> alpha_list

lapply(alpha_list, function(x) {
  
  x$overall[['Accuracy']]
  
  }
) %>% bind_rows(.id = 'alpha') %>% 
  t()


set.seed(123)

mult_model <- 
  cv.glmnet(x = x_in_filt,
            y = y,
            family = 'multinomial',
            type.measure = 'class',
            nfolds = nrow(x_in_filt), 
            alpha = 0.3, 
            parallel = TRUE,
            keep = TRUE)


coef_tmp <- Reduce(cbind, coef(mult_model, s = mult_model$lambda.1se))

beta <- coef_tmp[apply(coef_tmp != 0, 1, any), ]

colnames(beta) <- names(coef(mult_model, s = mult_model$lambda.1se))

beta

y_hat_matrix <- mult_model$fit.preval[,,which(mult_model$lambda == mult_model$lambda.min)]

# multi-nomial so need to apply softmax to each row (sample)
mult_response_matrix <- 
  t(apply(y_hat_matrix, 1, softmax.logit.calc))

mult_response_matrix %>% 
  cbind(y) %>% 
  as.data.frame() %>% 
  rownames_to_column('sample') %>% 
  gather(class, probability, -y, -sample) %>% 
  mutate(y = case_when(y == "1" ~ "panc", 
                    y == "2" ~ "covid", 
                    y == "0" ~ "ctrl"),
         class = case_when(class == "1" ~ "panc", 
                    class == "2" ~ "covid", 
                    class == "0" ~ "ctrl")) %>% 
  ggplot(aes(class, probability, color = y)) + 
  geom_point() + 
  geom_boxplot(outlier.colour = NA, fill = NA) +
  geom_line(aes(group = sample), alpha = 0.3) +
  ggforce::facet_row(~y) + 
  scale_color_manual(values = identity_color_pal, guide = 'none')

# which column is the max -- so need to pull from column names
y_hat_predictions <- 
  colnames(mult_response_matrix)[apply(mult_response_matrix, 1, which.max)]


factor(case_when(y == "1" ~ "panc", 
                    y == "2" ~ "covid", 
                    y == "0" ~ "ctrl"), ordered = T, levels = c('ctrl', 'panc', 'covid')) -> ord_y

factor(case_when(y_hat_predictions == "1" ~ "panc", 
                    y_hat_predictions == "2" ~ "covid", 
                    y_hat_predictions == "0" ~ "ctrl"), ordered = T, levels = c('ctrl', 'panc', 'covid')) -> ord_yhat

multi_roc <- 
  pROC::multiclass.roc(response = ord_y, 
                       predictor = ord_yhat, 
                       levels = c('ctrl','panc','covid'))

names(multi_roc$rocs) <- c('ctrl_v_panc', 'ctrl_v_covid', 'panc_v_covid')

ggroc(multi_roc$rocs) + 
  scale_color_manual(values = RColorBrewer::brewer.pal(name = 'Dark2', n = 3)) -> multi_roc.plt

caret::confusionMatrix(data = as.factor(case_when(y_hat_predictions == "1" ~ "panc", 
                                                  y_hat_predictions == "2" ~ "covid", 
                                                  y_hat_predictions == "0" ~ "ctrl")),
                       reference = as.factor(case_when(y == "1" ~ "panc", 
                                                       y == "2" ~ "covid", 
                                                       y == "0" ~ "ctrl"))) -> cm_out

cm_out$table %>% 
  gridExtra::tableGrob(theme = gridExtra::ttheme_minimal(base_size = 8,
                                                         core = list(padding=unit(c(1, 2.5), "mm")))) ->
  multinomial_cm.tbl

ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out %>% 
  mutate(sample = 'panc') %>% 
  rbind(ucsc.rmsk.salmon_quant$analysis_set$covid_ctrl$deseq$de_out %>% 
          mutate(sample = 'covid')) %>% 
  filter(ensg %in% rownames(beta)) %>% 
  merge(reference_meta_data$ucsc_rmsk_insert_info.df, by = 'gene', all.x = T) %>% 
  mutate(family = ifelse(ensg %in% reference_meta_data$gencode_lncRNA_gene_names.df$ensg,
                         'gencode_lncRNA', family),
         family = ifelse(is.na(family), 'gencode_coding', family),
         clade = ifelse(grepl('gencode', family), family, clade),
         clade = ifelse(clade %in% c('gencode_coding',
                                     'gencode_lncRNA',
                                     'LINE',
                                     'SINE',
                                     'LTR',
                                     'DNA',
                                     'Simple_repeat'), clade, 'other'),
         type = ifelse(grepl('gencode',family), 'GENCODE', 'TE'),
         clade = factor(clade, levels = te_aware_annotation_levels)) %>%
  ggplot(aes(log2FoldChange, -log10(padj), color = clade, label = gene)) + 
  geom_point() + 
  scale_color_manual(values = te_color_pal) +
  ggforce::facet_col(~sample) +
  ggrepel::geom_text_repel(color = 'black', segment.colour = NA, size = txt.mm_to_pts(0.8)) -> multi_feature_de.plt
```
## Fig_modeling -- multi_gt
```{r}
unique(c(filter(ucsc.rmsk.salmon_quant$analysis_set$covid_ctrl$deseq$de_out, 
                padj < 0.01, !ensg %in% chrM.filter.list)$ensg,
  filter(ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out, 
         padj < 0.01, !ensg %in% chrM.filter.list)$ensg)) ->
  de_ensg_names

x_in <- t(ucsc.rmsk.salmon_quant$analysis_set$ctrl_panc_covid$deseq$norm_counts.df)

x_in_filt <- x_in[, colnames(x_in) %in% c(de_ensg_names)]

y_in <- ucsc.rmsk.salmon_quant$analysis_set$ctrl_panc_covid$deseq$scaled_quant_meta_for_de.df %>% 
  select(sample, condition) %>% 
  mutate(condition = ifelse(condition == 'ctrl', 0, 
                            ifelse(condition == 'panc', 1, 2))) %>% 
  column_to_rownames('sample')

y <- y_in[match(rownames(x_in_filt), rownames(y_in)), ]

lapply(seq(0,1,0.1), function(alpha) {
  set.seed(123)

  mult_model <- 
    cv.glmnet(x = x_in_filt,
              y = y,
              family = 'multinomial',
              type.measure = 'class',
              nfolds = nrow(x_in_filt), 
              alpha = alpha, 
              parallel = TRUE,
              keep = TRUE)
  
  
  y_hat_matrix <- mult_model$fit.preval[,,which(mult_model$lambda == mult_model$lambda.min)]
  
  # multi-nomial so need to apply softmax to each row (sample)
  mult_response_matrix <- 
    t(apply(y_hat_matrix, 1, softmax.logit.calc))
  
  mult_response_matrix
  
  # which column is the max -- so need to pull from column names
  y_hat_predictions <- 
    colnames(mult_response_matrix)[apply(mult_response_matrix, 1, which.max)]
  
  multi_roc <- 
    pROC::multiclass.roc(response = y, 
                         predictor = mult_response_matrix, 
                         levels = c('0','1','2'))
  
  caret::confusionMatrix(data = as.factor(y_hat_predictions),
                         reference = as.factor(y)) -> cm_out
  
  ret_lst <- list(cm_out)
  names(ret_lst) <- alpha
  
  ret_lst
  
}) %>% purrr::flatten() -> alpha_list

lapply(alpha_list, function(x) {
  
  x$overall[['Accuracy']]
  
  }
) %>% bind_rows(.id = 'alpha') %>% 
  t()


set.seed(123)

mult_model <- 
  cv.glmnet(x = x_in_filt,
            y = y,
            family = 'multinomial',
            type.measure = 'class',
            nfolds = nrow(x_in_filt), 
            alpha = 0.3, 
            parallel = TRUE,
            keep = TRUE)


coef_tmp <- Reduce(cbind, coef(mult_model, s = mult_model$lambda.min))

beta <- coef_tmp[apply(coef_tmp != 0, 1, any), ]

colnames(beta) <- names(coef(mult_model, s = mult_model$lambda.min))

beta

y_hat_matrix <- mult_model$fit.preval[,,which(mult_model$lambda == mult_model$lambda.min)]

# multi-nomial so need to apply softmax to each row (sample)
mult_response_matrix <- 
  t(apply(y_hat_matrix, 1, softmax.logit.calc))

mult_response_matrix %>% 
  cbind(y) %>% 
  as.data.frame() %>% 
  rownames_to_column('sample') %>% 
  gather(class, probability, -y, -sample) %>% 
  mutate(y = case_when(y == "1" ~ "panc", 
                    y == "2" ~ "covid", 
                    y == "0" ~ "ctrl"),
         class = case_when(class == "1" ~ "panc", 
                    class == "2" ~ "covid", 
                    class == "0" ~ "ctrl")) %>% 
  ggplot(aes(class, probability, color = y)) + 
  geom_point() + 
  geom_boxplot(outlier.colour = NA, fill = NA) +
  geom_line(aes(group = sample), alpha = 0.3) +
  ggforce::facet_row(~y) + 
  scale_color_manual(values = identity_color_pal, guide = 'none')

# which column is the max -- so need to pull from column names
y_hat_predictions <- 
  colnames(mult_response_matrix)[apply(mult_response_matrix, 1, which.max)]


factor(case_when(y == "1" ~ "panc", 
                    y == "2" ~ "covid", 
                    y == "0" ~ "ctrl"), ordered = T, levels = c('ctrl', 'panc', 'covid')) -> ord_y

factor(case_when(y_hat_predictions == "1" ~ "panc", 
                    y_hat_predictions == "2" ~ "covid", 
                    y_hat_predictions == "0" ~ "ctrl"), ordered = T, levels = c('ctrl', 'panc', 'covid')) -> ord_yhat

multi_roc <- 
  pROC::multiclass.roc(response = ord_y, 
                       predictor = ord_yhat, 
                       levels = c('ctrl','panc','covid'))

names(multi_roc$rocs) <- c('ctrl_v_panc', 'ctrl_v_covid', 'panc_v_covid')

ggroc(multi_roc$rocs) + 
  scale_color_manual(values = RColorBrewer::brewer.pal(name = 'Dark2', n = 3)) -> gt_multi_roc.plt

caret::confusionMatrix(data = as.factor(case_when(y_hat_predictions == "1" ~ "panc", 
                                                  y_hat_predictions == "2" ~ "covid", 
                                                  y_hat_predictions == "0" ~ "ctrl")),
                       reference = as.factor(case_when(y == "1" ~ "panc", 
                                                       y == "2" ~ "covid", 
                                                       y == "0" ~ "ctrl"))) -> cm_out

cm_out$table %>% 
  gridExtra::tableGrob(theme = gridExtra::ttheme_minimal(base_size = 8,
                                                         core = list(padding=unit(c(1, 2.5), "mm")))) ->
  gt_multinomial_cm.tbl

ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out %>% 
  mutate(sample = 'panc') %>% 
  rbind(ucsc.rmsk.salmon_quant$analysis_set$covid_ctrl$deseq$de_out %>% 
          mutate(sample = 'covid')) %>% 
  filter(ensg %in% rownames(beta)) %>% 
  merge(reference_meta_data$ucsc_rmsk_insert_info.df, by = 'gene', all.x = T) %>% 
  mutate(family = ifelse(ensg %in% reference_meta_data$gencode_lncRNA_gene_names.df$ensg,
                         'gencode_lncRNA', family),
         family = ifelse(is.na(family), 'gencode_coding', family),
         clade = ifelse(grepl('gencode', family), family, clade),
         clade = ifelse(clade %in% c('gencode_coding',
                                     'gencode_lncRNA',
                                     'LINE',
                                     'SINE',
                                     'LTR',
                                     'DNA',
                                     'Simple_repeat'), clade, 'other'),
         type = ifelse(grepl('gencode',family), 'GENCODE', 'TE'),
         clade = factor(clade, levels = te_aware_annotation_levels)) %>%
  ggplot(aes(log2FoldChange, -log10(padj), color = clade, label = gene)) + 
  geom_point() + 
  scale_color_manual(values = te_color_pal) +
  ggforce::facet_col(~sample) +
  ggrepel::geom_text_repel(color = 'black', segment.colour = NA, size = txt.mm_to_pts(0.8)) ->
  gt_multi_feature_de.plt

```

## Fig_Modeling -- binomial
```{r}
inverse.logit.calc <- function(y_hat) { exp(y_hat)/(1+exp(y_hat)) }

sens_at_spec_ROC <- function(roc, spec) { 
  
  sens <- pROC::coords(roc, input = 'specificity', spec)$sensitivity
  
  print(sens)
  
  p.thresh <- inverse.logit.calc(pROC::coords(roc, input = 'specificity', spec)$threshold)
  
  return(list('sens' = sens, 
              'spec' = spec, 
              'p.thresh' = p.thresh))
  
}

doMC::registerDoMC(cores = 6)

x_in <-
  t(internal_rmsk_quant$panc_normal$deseq$norm_counts.df[rownames(internal_rmsk_quant$panc_normal$deseq$norm_counts.df) %in%
                                                           filter(reference_meta_data$ucsc_rmsk_insert_info.df,
                                                                  grepl('Alu', gene))$gene,])

calc_entropy(
  internal_rmsk_quant$panc_normal$deseq$norm_counts.df,
  original_normal_meta_data,
  grouping_level = 'clade'
) %>%
  mutate(clade = ifelse(clade %in% names(te_color_pal), clade, 'other')) %>% 
  group_by(sample, clade) %>% 
  summarize(entropy = mean(entropy)) %>% 
  pivot_wider(names_from = 'clade', values_from = 'entropy') %>% 
  column_to_rownames('sample') %>% as.matrix() -> x_in


y_in <- internal_rmsk_quant$panc_normal$deseq$scaled_quant_meta_for_de.df %>% 
  rownames_to_column('sample') %>% 
  select(sample, condition) %>% 
  mutate(condition = factor(ifelse(condition == 'normal', 0, 1), levels = c(0,1))) %>% 
  tibble::column_to_rownames('sample')

y <- y_in[match(rownames(x_in), rownames(y_in)), ]

set.seed(123)

binom_model <- 
  glmnet::cv.glmnet(x = x_in,
            y = y_in$condition,
            family = 'binomial',
            type.measure = 'deviance',
            nfolds = 20, 
            alpha = 0, 
            parallel = TRUE,
            keep = TRUE)


coef_tmp <- cbind(coef(binom_model, s = binom_model$lambda.min))

beta <- coef_tmp[apply(coef_tmp != 0, 1, any), ]

colnames(beta) <- names(coef(binom_model, s = binom_model$lambda.min))

beta

y_hat_matrix <- binom_model$fit.preval[,which(binom_model$lambda == binom_model$lambda.min)]

binomial_model_response_prob <- inverse.logit.calc(y_hat_matrix)
# 
classifications <- ifelse(binomial_model_response_prob > perf_list$p.thresh, 1, 0)

binom_roc_out <- pROC::roc(response = y, predictor = y_hat_matrix)

#
predictions <-
  ROCR::prediction(
    labels = y,
    predictions = binomial_model_response_prob,
    label.ordering = c(0, 1)
  )
#
performances <-
  ROCR::performance(prediction.obj = predictions,
                    measure = "tpr",
                    x.measure = "fpr")



#
rocCurve <-
  as.data.frame(cbind(unlist(slot(
    performances, "x.values"
  )[[1]]),
  unlist(slot(
    performances, "y.values"
  )[[1]])))

colnames(rocCurve) <- c("X","Y")

ggplot(rocCurve, aes(x = X, y = Y)) +
  geom_line() +
  geom_abline(linetype = "dotted", color = "grey50") +
  labs(x = "False Positive Rate", y = "True Positive Rate") +
  ylim(0,1) + xlim(0,1)

binomial_model_response_prob <- inverse.logit.calc(y_hat_matrix)

predictions@cutoffs[[1]]

binomial_model_response_prob %>%
  as.data.frame() %>%
  rownames_to_column('sample') %>%
  dplyr::rename('prob' = '.') %>%
  merge(y_in %>% rownames_to_column('sample'), by = 'sample') %>%
  mutate(condition = ifelse(condition == 1, 'panc', 'normal')) %>%
  ggplot(aes(condition, 1-prob)) +
  geom_jitter(width = 0.125) +
  geom_boxplot(outlier.colour = NA, fill = NA) + 
  geom_hline(yintercept = perf_list$p.thresh, linetype = 'dashed') + 
  annotate(geom = 'text', y = perf_list$p.thresh + 0.05, x = 1.5, 
           label = '100% specificity prob. threshold', size = txt.mm_to_pts(0.8)) +
  ylab('probability') + 
  xlab('') -> binom_prob_thresh.plt

binom_prob_thresh.plt

```

## Binomial
```{r}
doMC::registerDoMC(cores = 12)

de_out_list <-
  setNames(
    list(
      elife_rmsk_quant$liver_healthy$deseq$norm_counts.df,
      elife_rmsk_quant$lung_healthy$deseq$norm_counts.df,
      elife_rmsk_quant$esoph_healthy$deseq$norm_counts.df,
      elife_rmsk_quant$colorectal_healthy$deseq$norm_counts.df,
      elife_rmsk_quant$stomach_healthy$deseq$norm_counts.df
    ),
    c('liver',
      'lung',
      'esoph',
      'colorectal',
      'stomach')
  )


calc_entropy(elife_rmsk_quant$colorectal_healthy$deseq$norm_counts.df,
             elife_meta_data,
             grouping_level = 'clade') %>%
  mutate(clade = ifelse(clade %in% names(te_color_pal), clade, 'other')) %>%
  filter(!grepl('GENCODE', clade)) %>%
  group_by(sample, clade) %>%
  summarize(entropy = mean(entropy)) %>%
  pivot_wider(names_from = 'clade', values_from = 'entropy') %>%
  select_if( ~ !any(is.na(.))) %>%
  column_to_rownames('sample') %>% as.matrix() -> binom_model_input_data_entropy

cbind(binom_model_input_data_entropy,
      t(elife_rmsk_quant$colorectal_healthy$deseq$norm_counts.df[!rownames(elife_rmsk_quant$colorectal_healthy$deseq$norm_counts.df) %in%
                                                     reference_meta_data$ucsc_rmsk_insert_info.df$gene, ])) ->
  binom_model_input_data_entropy_plus


binom_model_input_data <-
  t(elife_rmsk_quant$colorectal_healthy$deseq$norm_counts.df[!rownames(elife_rmsk_quant$colorectal_healthy$deseq$norm_counts.df) %in%
                                                 reference_meta_data$ucsc_rmsk_insert_info.df$gene, ])

binom_model_input_data_all <-
  t(elife_rmsk_quant$stomach_healthy$deseq$norm_counts.df)

calc_entropy(elife_rmsk_quant$colorectal_healthy$deseq$norm_counts.df,
             elife_meta_data,
             grouping_level = 'clade') %>%
  mutate(clade = ifelse(clade %in% names(te_color_pal), clade, 'other')) %>%
  filter(!grepl('GENCODE', clade)) %>%
  group_by(sample, clade) %>%
  summarize(entropy = mean(entropy)) %>%
  pivot_wider(names_from = 'clade', values_from = 'entropy') %>%
  select_if( ~ !any(is.na(.))) %>%
  column_to_rownames('sample') %>% as.matrix() -> binom_model_input_data_entropy

binom_model_input_meta_data <-
  elife_meta_data[match(rownames(binom_model_input_data), elife_meta_data$sample),] %>% 
  mutate(diagnosis = ifelse(diagnosis == 'Healthy donor', 0, 1))

binom_model_input_data_gencode <-
  t(elife_rmsk_quant$colorectal_healthy$deseq$norm_counts.df[!rownames(elife_rmsk_quant$colorectal_healthy$deseq$norm_counts.df) %in%
                                                 reference_meta_data$ucsc_rmsk_insert_info.df$gene, ])

binom_model_input_data_te <-
  t(elife_rmsk_quant$colorectal_healthy$deseq$norm_counts.df[rownames(elife_rmsk_quant$colorectal_healthy$deseq$norm_counts.df) %in%
                                                 reference_meta_data$ucsc_rmsk_insert_info.df$gene, ])

cbind(
  binom_model_input_data,
  t(elife_rmsk_quant$colorectal_healthy$deseq$norm_counts.df[rownames(elife_rmsk_quant$colorectal_healthy$deseq$norm_counts.df) %in%
                                                               reference_meta_data$ucsc_rmsk_insert_info.df$gene,])
) ->
  binom_model_input_data_te_only

set.seed(2022-03-08)
trainSplitIndex <-
  caret::createDataPartition(
    factor(
      binom_model_input_meta_data$diagnosis,
      levels = c(0,1)
    ),
    p = .8,
    list = FALSE,
    times = 1
  )

binom_elife_x_in <- binom_model_input_data[, colSums(binom_model_input_data != 0) > 0]

binom_elife_y_in <-
  binom_model_input_meta_data 

binom_elife_x_in_train <- binom_elife_x_in[trainSplitIndex, , drop = F]
binom_elife_y_in_train <-
  binom_elife_y_in[trainSplitIndex, , drop = F] %>% pull(diagnosis)

binom_elife_x_in_test <- binom_elife_x_in[-trainSplitIndex, , drop = F]
binom_elife_y_in_test <-
  binom_elife_y_in[-trainSplitIndex, , drop = F] %>% pull(diagnosis)


set.seed(2022 - 03 - 08)
fold_ids <-
  caret::createFolds(factor(binom_elife_y_in_train), 10, list = F)

lapply(seq(0, 1, 0.1), function(alpha) {
  binom_elife_model <-
    glmnet::cv.glmnet(
      x = binom_elife_x_in_train,
      y = binom_elife_y_in_train,
      family = 'binomial',
      type.measure = 'deviance',
      nfolds = 10,
      alpha = alpha,
      foldid = fold_ids,
      parallel = TRUE,
      keep = TRUE
    )
  
  y_hat_matrix <-
    binom_elife_model$fit.preval[, which(binom_elife_model$lambda == binom_elife_model$lambda.min)]
  
  predictions <-
    ROCR::prediction(
      labels = binom_elife_y_in_train,
      predictions = y_hat_matrix,
      label.ordering = c(0, 1)
    )
  
  performances <-
    ROCR::performance(prediction.obj = predictions,
                      measure = "tpr",
                      x.measure = "fpr")
  ROCR::performance(prediction.obj = predictions,
                      measure = "auc")@y.values -> auc
  
  cutoffs <-
    data.frame(
      cut = performances@alpha.values[[1]],
      fpr = performances@x.values[[1]],
      tpr = performances@y.values[[1]]
    )
  
  
  perf_list <- cutoffs[which.min(abs(cutoffs$fpr - (1 - 0.9))),]
  perf_list$cut <-  inverse.logit.calc(perf_list$cut)
  perf_list$auc <-  auc
  perf_list$alpha <- alpha
  
  perf_list
}) %>% bind_rows() -> alpha_list



binom_elife_model <-
  glmnet::cv.glmnet(
    x = binom_elife_x_in_train,
    y = binom_elife_y_in_train,
    family = 'binomial',
    type.measure = 'deviance',
    nfolds = 10,
    alpha = alpha_list[which.max(alpha_list$auc), ]$alpha,
    foldid = fold_ids,
    parallel = TRUE,
    keep = TRUE
  )

y_hat_matrix <- binom_elife_model$fit.preval[,which(binom_elife_model$lambda == binom_elife_model$lambda.min)]

predictions <-
  ROCR::prediction(
    labels = binom_elife_y_in_train,
    predictions = y_hat_matrix,
    label.ordering = c(0, 1)
  )

performances <-
  ROCR::performance(prediction.obj = predictions,
                    measure = "tpr",
                    x.measure = "fpr")

cutoffs <- data.frame(cut=performances@alpha.values[[1]], fpr=performances@x.values[[1]], 
                      tpr=performances@y.values[[1]])


perf_list <- cutoffs[which.min(abs(cutoffs$fpr - (1-0.9))), ]
perf_list$cut <-  inverse.logit.calc(perf_list$cut)
perf_list

coef_tmp <- cbind(coef(binom_elife_model, s = binom_elife_model$lambda.min))

beta <- coef_tmp[apply(coef_tmp != 0, 1, any), ]

colnames(beta) <- names(coef(binom_elife_model, s = binom_elife_model$lambda.min))

beta

#
rocCurve <-
  as.data.frame(cbind(unlist(slot(
    performances, "x.values"
  )[[1]]),
  unlist(slot(
    performances, "y.values"
  )[[1]])))

colnames(rocCurve) <- c("X","Y")

ggplot(rocCurve, aes(x = X, y = Y)) +
  geom_line() +
  geom_abline(linetype = "dotted", color = "grey50") +
  labs(x = "False Positive Rate", y = "True Positive Rate") +
  ylim(0,1) + xlim(0,1)

binomial_model_response_prob <- inverse.logit.calc(y_hat_matrix)

binomial_model_response_prob %>%
  as.data.frame() %>% 
  rownames_to_column('sample') %>%
  dplyr::rename('prob' = '.') %>%
  merge(binom_elife_y_in, by = 'sample') %>% 
  mutate(diagnosis = ifelse(diagnosis == 1, 'cancer', 'normal')) %>%
  ggplot(aes(diagnosis, prob)) +
  geom_jitter(width = 0.125) +
  geom_boxplot(outlier.colour = NA, fill = NA) + 
  geom_hline(yintercept = perf_list$cut, linetype = 'dashed') + 
  annotate(geom = 'text', y = perf_list$cut + 0.05, x = 1.5, 
           label = '~90% specificity prob. threshold', size = txt.mm_to_pts(0.8)) +
  ylab('probability') + 
  xlab('')



```



## LDA
```{r}
calc_entropy(
  internal_rmsk_quant$panc_normal$deseq$norm_counts.df,
  original_normal_meta_data,
  grouping_level = 'clade'
) %>%
  mutate(clade = ifelse(clade %in% names(te_color_pal), clade, 'other')) %>% 
  # filter(!grepl('GENCODE', clade)) %>%
  group_by(sample, clade) %>% 
  summarize(entropy = mean(entropy)) %>% 
  pivot_wider(names_from = 'clade', values_from = 'entropy') %>% 
  merge(original_normal_meta_data %>% select(sample, diagnosis), by = 'sample') %>% 
  mutate(diagnosis = ifelse(diagnosis == 'normal', 0, 1),
         diagnosis = factor(diagnosis)) %>% 
  select(-sample) -> internal_entropy_input

calc_entropy(elife_rmsk_quant$rmsk$deseq$norm_counts.df,
             elife_meta_data,
             grouping_level = 'clade') %>%
  mutate(clade = ifelse(clade %in% names(te_color_pal), clade, 'other')) %>%
  # filter(!grepl('GENCODE', clade)) %>%
  group_by(sample, clade) %>%
  summarize(entropy = mean(entropy)) %>%
  pivot_wider(names_from = 'clade', values_from = 'entropy') %>%
  merge(elife_meta_data %>% select(sample, diagnosis), by = 'sample') %>%
  mutate(
    diagnosis = ifelse(diagnosis == 'Healthy donor', 0, 1),
    diagnosis = factor(diagnosis)
  ) %>%
  filter(sample %in% binom_elife_y_in[trainSplitIndex, , drop = F]$sample) -> elife_entropy_train_input

calc_entropy(elife_rmsk_quant$rmsk$deseq$norm_counts.df,
             elife_meta_data,
             grouping_level = 'clade') %>%
  mutate(clade = ifelse(clade %in% names(te_color_pal), clade, 'other')) %>%
  filter(!grepl('GENCODE', clade)) %>%
  group_by(sample, clade) %>%
  summarize(entropy = mean(entropy)) %>%
  pivot_wider(names_from = 'clade', values_from = 'entropy') %>%
  merge(elife_meta_data %>% select(sample, diagnosis), by = 'sample') %>%
  mutate(
    diagnosis = ifelse(diagnosis == 'Healthy donor', 0, 1),
    diagnosis = factor(diagnosis)
  ) %>%
  filter(!sample %in% binom_elife_y_in[trainSplitIndex, , drop = F]$sample) -> elife_entropy_test_input

calc_entropy(elife_rmsk_quant$rmsk$deseq$norm_counts.df,
             elife_meta_data,
             grouping_level = 'clade') %>%
  mutate(clade = ifelse(clade %in% names(te_color_pal), clade, 'other')) %>%
  filter(!grepl('GENCODE', clade)) %>%
  group_by(sample, clade) %>%
  summarize(entropy = mean(entropy)) %>%
  pivot_wider(names_from = 'clade', values_from = 'entropy') %>%
  merge(elife_meta_data %>% select(sample, diagnosis), by = 'sample') %>%
  mutate(
    diagnosis = case_when(
      diagnosis == 'Healthy donor' ~ 0,
      diagnosis == 'Esophagus Cancer' ~ 1,
      diagnosis == 'Liver Cancer' ~ 2,
      diagnosis == 'Lung Cancer' ~ 3,
      diagnosis == 'Colorectal Cancer' ~ 4,
      diagnosis == 'Stomach Cancer' ~ 5
    ),
    diagnosis = factor(diagnosis)
  ) %>%
  filter(sample %in% binom_elife_y_in[trainSplitIndex, , drop = F]$sample) -> multi_entropy_train_input

calc_entropy(elife_rmsk_quant$rmsk$deseq$norm_counts.df,
             elife_meta_data,
             grouping_level = 'clade') %>%
  mutate(clade = ifelse(clade %in% names(te_color_pal), clade, 'other')) %>%
  filter(!grepl('GENCODE', clade)) %>%
  group_by(sample, clade) %>%
  summarize(entropy = mean(entropy)) %>%
  pivot_wider(names_from = 'clade', values_from = 'entropy') %>%
  merge(elife_meta_data %>% select(sample, diagnosis), by = 'sample') %>%
  mutate(
    diagnosis = case_when(
      diagnosis == 'Healthy donor' ~ 0,
      diagnosis == 'Esophagus Cancer' ~ 1,
      diagnosis == 'Liver Cancer' ~ 2,
      diagnosis == 'Lung Cancer' ~ 3,
      diagnosis == 'Colorectal Cancer' ~ 4,
      diagnosis == 'Stomach Cancer' ~ 5
    ),
    diagnosis = factor(diagnosis)
  ) %>%
  filter(!sample %in% binom_elife_y_in[trainSplitIndex, , drop = F]$sample) -> multi_entropy_test_input


MASS::lda(diagnosis ~ ., data = elife_entropy_train_input %>% column_to_rownames('sample')) -> lin_model

predict(lin_model, internal_entropy_input)
```

### ensemble
```{r}

```

## Multinomial
```{r}
doMC::registerDoMC(cores = 12)

softmax.logit.calc <- function(y_hat) { exp(y_hat)/sum(exp(y_hat)) }

model_input_data <-
  t(elife_rmsk_quant$rmsk$deseq$norm_counts.df)

model_input_meta_data <-
  elife_meta_data[match(rownames(model_input_data), elife_meta_data$sample),]

trainSplitIndex <-
  caret::createDataPartition(
    factor(
      model_input_meta_data$diagnosis,
      levels = c(
        'Healthy donor',
        'Esophagus Cancer',
        'Liver Cancer',
        'Lung Cancer',
        'Colorectal Cancer',
        'Stomach Cancer'
      )
    ),
    p = .8,
    list = FALSE,
    times = 1
  )

elife_x_in <- model_input_data

elife_y_in <-
  model_input_meta_data %>%
  mutate(
    diagnosis = case_when(
      diagnosis == 'Healthy donor' ~ 0,
      diagnosis == 'Esophagus Cancer' ~ 1,
      diagnosis == 'Liver Cancer' ~ 2,
      diagnosis == 'Lung Cancer' ~ 3,
      diagnosis == 'Colorectal Cancer' ~ 4,
      diagnosis == 'Stomach Cancer' ~ 5
    )
  )

elife_x_in_train <- elife_x_in[trainSplitIndex, , drop = F]
elife_y_in_train <-
  elife_y_in[trainSplitIndex, , drop = F] %>% pull(diagnosis)

elife_x_in_test <- elife_x_in[-trainSplitIndex, , drop = F]
elife_y_in_test <-
  elife_y_in[-trainSplitIndex, , drop = F] %>% pull(diagnosis)

set.seed(2022 - 03 - 07)

print(alpha_list)

mult_model <-
  cv.glmnet(
    x = elife_x_in_train,
    y = elife_y_in_train,
    family = 'multinomial',
    type.measure = 'class',
    nfolds = 20,
    alpha = 0.2,
    parallel = TRUE,
    keep = TRUE
  )


coef_tmp <-
  Reduce(cbind, coef(mult_model, s = mult_model$lambda.1se))

beta <- coef_tmp[apply(coef_tmp != 0, 1, any),]

colnames(beta) <-
  names(coef(mult_model, s = mult_model$lambda.1se))

beta

inverse.logit.calc <-
  function(y_hat) {
    exp(y_hat) / (1 + exp(y_hat))
  }
softmax.logit.calc <-
  function(y_hat) {
    exp(y_hat) / sum(exp(y_hat))
  }

y_hat_matrix <-
  mult_model$fit.preval[, , which(mult_model$lambda == mult_model$lambda.min)]

# multi-nomial so need to apply softmax to each row (sample)
mult_response_matrix <-
  t(apply(y_hat_matrix, 1, softmax.logit.calc))

mult_response_matrix

# which column is the max -- so need to pull from column names
y_hat_predictions <-
  colnames(mult_response_matrix)[apply(mult_response_matrix, 1, which.max)]

multi_roc <-
  pROC::multiclass.roc(
    response = elife_y_in_train,
    predictor = mult_response_matrix,
    levels = c('0', '1', '2', '3', '4', '5')
  )


caret::confusionMatrix(data = as.factor(y_hat_predictions),
                       reference = as.factor(elife_y_in_train)) -> cm_out
  
  

lapply(alpha_list, function(x) {
  
  x$overall[['Accuracy']]
  
  }
) %>% bind_rows(.id = 'alpha') %>% 
  t() %>% which.max()
```

###
```{r}
orig_x_in <- t(ucsc.rmsk.salmon_quant$original_data$deseq$norm_counts.df)

orig_y_in <- ucsc.rmsk.salmon_quant$original_data$deseq$scaled_quant_meta_for_de.df %>% 
  rownames_to_column('sample') %>% 
  select(sample, condition) %>% 
  mutate(condition = ifelse(condition == 'normal', 0, 
                            ifelse(condition == 'panc', 1, 2))) %>% 
  column_to_rownames('sample')

orig_y <- orig_y_in[match(rownames(orig_x_in), rownames(orig_y_in)), ]

original_y_hat_predictions <- predict(object = binom_model, s = 'lambda.min', 
                                      newx = orig_x_in, newy = orig_y_in, type = 'response')

predictions <- ROCR::prediction(labels = orig_y_in, predictions = original_y_hat_predictions, label.ordering = c(0,1,2))
# 
performances <- ROCR::performance(prediction.obj = predictions, measure = "tpr", x.measure = "fpr")
# 
rocCurve <- as.data.frame(cbind(unlist(slot(performances,"x.values")[[1]]),
                                unlist(slot(performances,"y.values")[[1]])))

colnames(rocCurve) <- c("X","Y")

ggplot(rocCurve, aes(x = X, y = Y)) +
  # geom_vline(xintercept = xint, linetype = 'dashed', alpha = 0.3) +
  # geom_hline(yintercept = yint, linetype = 'dashed', alpha = 0.3) +
  geom_line(color = "red") +
  geom_abline(linetype = "dotted", color = "grey50") +
  labs(x = "False Positive Rate", y = "True Positive Rate") +
  ylim(0,1) + xlim(0,1) +
  # scale_y_continuous(breaks = c(0, 0.25, 0.75, yint)) +
  # scale_x_continuous(breaks = c( 0.25, 0.5, 0.75, xint)) +
  theme_classic()

orig_binom_roc_out <- pROC::roc(response = orig_y, predictor = y_hat_matrix)

perf_list <- sens_at_spec_ROC(binom_roc_out, 1.0)

original_y_hat_predictions <- ifelse(original_y_hat_predictions > perf_list$p.thresh, 1, 0)

pROC::ggroc(binom_roc_out) -> binom_roc.plt

binomial_model_response_prob <- inverse.logit.calc(y_hat_matrix)

binomial_model_response_prob %>%
  as.data.frame() %>%
  rownames_to_column('sample') %>%
  dplyr::rename('prob' = '.') %>%
  merge(y_in %>% rownames_to_column('sample'), by = 'sample') %>%
  mutate(condition = ifelse(condition == 1, 'panc', 'normal')) %>%
  ggplot(aes(condition, prob)) +
  geom_jitter(width = 0.125) +
  geom_boxplot(outlier.colour = NA, fill = NA) + 
  geom_hline(yintercept = perf_list$p.thresh, linetype = 'dashed') + 
  annotate(geom = 'text', y = perf_list$p.thresh + 0.05, x = 1.5, 
           label = '100% specificity prob. threshold', size = txt.mm_to_pts(0.8)) +
  ylab('probability') + 
  xlab('') -> binom_prob_thresh.plt

```

####
```{r}
# beta %>%
#   as.matrix() %>%
#   as.data.frame() %>%
#   rownames_to_column('gene') %>%
#   merge(ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out, by = 'gene') %>%
#   ggplot(aes(V1, log2FoldChange, label = gene)) +
#   geom_point() +
#   ggrepel::geom_text_repel() 

ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out %>% 
  filter(ensg %in% names(beta)) %>% 
  merge(reference_meta_data$ucsc_rmsk_insert_info.df, by = 'gene', all.x = T) %>% 
  mutate(family = ifelse(ensg %in% reference_meta_data$gencode_lncRNA_gene_names.df$ensg,
                         'GENCODE_lncRNA', family),
         family = ifelse(is.na(family), 'GENCODE_coding', family),
         clade = ifelse(grepl('GENCODE', family), family, clade),
         clade = ifelse(clade %in% c('GENCODE_coding',
                                     'GENCODE_lncRNA',
                                     'LINE',
                                     'SINE',
                                     'LTR',
                                     'DNA',
                                     'Simple_repeat'), clade, 'other'),
         type = ifelse(grepl('GENCODE',family), 'GENCODE', 'TE'),
         clade = factor(clade, levels = te_aware_annotation_levels)) %>% 
  ggplot(aes(log2FoldChange, -log10(padj), color = clade, label = gene)) + 
  geom_point() + 
  scale_color_manual(values = te_color_pal, name = 'biotype/clade') + 
  geom_vline(xintercept = 0, linetype = 'dashed', alpha = 0.4) + 
  ggrepel::geom_text_repel(color = 'black', segment.colour = NA, size = txt.mm_to_pts(0.8)) -> feature_de.plt

caret::confusionMatrix(data = as.factor(ifelse(as.factor(y_hat_predictions) == 1, 'panc', 'ctrl')),
                       reference = as.factor(ifelse(as.factor(y) == 1, 'panc', 'ctrl')))$table %>% 
  gridExtra::tableGrob(theme = gridExtra::ttheme_minimal(base_size = 8,
                                                         core = list(padding=unit(c(1, 2.5), "mm")))) ->
  binomial_cm.tbl

tg <- grid::textGrob('Reference', x=0, hjust=0, 
                     gp = grid::gpar(fontsize = 7, fontface = 'bold'))
lg <- grid::textGrob('Prediction\n AUC = 1', x=1, hjust=1, 
                     gp = grid::gpar(fontsize = 7, fontface = 'italic'))
binomial_cm.tbl <- gtable::gtable_add_rows(binomial_cm.tbl, unit(1,"line"), 0)
binomial_cm.tbl <- gtable::gtable_add_cols(binomial_cm.tbl, grid::grobWidth(lg), 0)
binomial_cm.tbl <- gtable::gtable_add_grob(binomial_cm.tbl, tg, t = 1, l=3, r=ncol(binomial_cm.tbl))
binomial_cm.tbl <- gtable::gtable_add_grob(binomial_cm.tbl, lg, l = 1, t=2)
```

## Fig_modeling--gt_binom
```{r}
doMC::registerDoMC(cores = 6)

filter(ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out, 
       padj < 0.01, ! ensg %in% chrM.filter.list)$ensg ->
  de_ensg_names

x_in <- t(ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$norm_counts.df)

x_in_filt <- x_in[, colnames(x_in) %in% c(de_ensg_names)]

y_in <- ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$scaled_quant_meta_for_de.df %>% 
  select(sample, condition) %>% 
  mutate(condition = ifelse(condition == 'ctrl', 0, 
                            ifelse(condition == 'panc', 1, 2))) %>% 
  column_to_rownames('sample')

y <- y_in[match(rownames(x_in_filt), rownames(y_in)), ]

set.seed(123)

binom_model <- 
  cv.glmnet(x = x_in_filt,
            y = y,
            family = 'binomial',
            type.measure = 'deviance',
            nfolds = nrow(x_in_filt), 
            alpha = 0.75, 
            parallel = TRUE,
            keep = TRUE)


coef_tmp <- cbind(coef(binom_model, s = binom_model$lambda.min))

beta <- coef_tmp[apply(coef_tmp != 0, 1, any), ]

colnames(beta) <- names(coef(binom_model, s = binom_model$lambda.min))

y_hat_matrix <- binom_model$fit.preval[,which(binom_model$lambda == binom_model$lambda.min)]

y_hat_predictions <- predict(object = binom_model, s = 'lambda.min', 
                             newx = x_in_filt, newy = y_in, type = 'response')

binom_roc_out <- pROC::roc(response = y, predictor = y_hat_matrix)

perf_list <- sens_at_spec_ROC(binom_roc_out, 1.0)

y_hat_predictions <- ifelse(y_hat_predictions > perf_list$p.thresh, 1, 0)

pROC::ggroc(binom_roc_out) -> gt_binom_roc.plt

binomial_model_response_prob <- inverse.logit.calc(y_hat_matrix)

binomial_model_response_prob %>%
  as.data.frame() %>%
  rownames_to_column('sample') %>%
  dplyr::rename('prob' = '.') %>%
  merge(y_in %>% rownames_to_column('sample'), by = 'sample') %>%
  mutate(condition = ifelse(condition == 1, 'panc', 'ctrl')) %>%
  ggplot(aes(condition, prob)) +
  geom_jitter(width = 0.125) +
  geom_boxplot(outlier.colour = NA, fill = NA) + 
  geom_hline(yintercept = perf_list$p.thresh, linetype = 'dashed') + 
  annotate(geom = 'text', y = perf_list$p.thresh + 0.05, x = 1.5, 
           label = '100% specificity prob. threshold', size = txt.mm_to_pts(0.8)) +
  ylab('probability') + 
  xlab('') -> gt_binom_prob_thresh.plt

beta %>%
  as.matrix() %>%
  as.data.frame() %>%
  rownames_to_column('gene') %>%
  merge(ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out, by = 'gene') %>%
  ggplot(aes(V1, log2FoldChange, label = gene)) +
  geom_point() +
  ggrepel::geom_text_repel() 

ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out %>% 
  filter(ensg %in% names(beta)) %>% 
  merge(reference_meta_data$ucsc_rmsk_insert_info.df, by = 'gene', all.x = T) %>% 
  mutate(family = ifelse(ensg %in% reference_meta_data$gencode_lncRNA_gene_names.df$ensg,
                         'GENCODE_lncRNA', family),
         family = ifelse(is.na(family), 'GENCODE_coding', family),
         clade = ifelse(grepl('GENCODE', family), family, clade),
         clade = ifelse(clade %in% c('GENCODE_coding',
                                     'GENCODE_lncRNA',
                                     'LINE',
                                     'SINE',
                                     'LTR',
                                     'DNA',
                                     'Simple_repeat'), clade, 'other'),
         type = ifelse(grepl('GENCODE',family), 'GENCODE', 'TE'),
         clade = factor(clade, levels = te_aware_annotation_levels)) %>% 
  ggplot(aes(log2FoldChange, -log10(padj), color = clade, label = gene)) + 
  geom_point() + 
  scale_color_manual(values = te_color_pal, name = 'biotype/clade') + 
  geom_vline(xintercept = 0, linetype = 'dashed', alpha = 0.4) + 
  ggrepel::geom_text_repel(color = 'black', segment.colour = NA, size = txt.mm_to_pts(0.8)) -> gt_feature_de.plt

caret::confusionMatrix(data = as.factor(ifelse(as.factor(y_hat_predictions) == 1, 'panc', 'ctrl')),
                       reference = as.factor(ifelse(as.factor(y) == 1, 'panc', 'ctrl')))$table %>% 
  gridExtra::tableGrob(theme = gridExtra::ttheme_minimal(base_size = 8,
                                                         core = list(padding=unit(c(1, 2.5), "mm")))) ->
  gt_binomial_cm.tbl

tg <- grid::textGrob('Reference', x=0, hjust=0, 
                     gp = grid::gpar(fontsize = 7, fontface = 'bold'))
lg <- grid::textGrob('Prediction\n AUC = 1', x=1, hjust=1, 
                     gp = grid::gpar(fontsize = 7, fontface = 'italic'))
gt_binomial_cm.tbl <- gtable::gtable_add_rows(gt_binomial_cm.tbl, unit(1,"line"), 0)
gt_binomial_cm.tbl <- gtable::gtable_add_cols(gt_binomial_cm.tbl, grid::grobWidth(lg), 0)
gt_binomial_cm.tbl <- gtable::gtable_add_grob(gt_binomial_cm.tbl, tg, t = 1, l=3, r=ncol(gt_binomial_cm.tbl))
gt_binomial_cm.tbl <- gtable::gtable_add_grob(gt_binomial_cm.tbl, lg, l = 1, t=2)

```
## Fig_Modeling -- alu_binomial
```{r}
# doMC::registerDoMC(cores = 6)
# 
# filter(ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out, 
#        padj < 0.01, grepl('^Alu', ensg), ! gene %in% chrM.filter.list)$ensg ->
#   de_ensg_names
# 
# x_in <- t(ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$norm_counts.df)
# 
# x_in_filt <- x_in[, colnames(x_in) %in% c(de_ensg_names)]
# 
# y_in <- ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$scaled_quant_meta_for_de.df %>% 
#   select(sample, condition) %>% 
#   mutate(condition = ifelse(condition == 'ctrl', 0, 
#                             ifelse(condition == 'panc', 1, 2))) %>% 
#   column_to_rownames('sample')
# 
# y <- y_in[match(rownames(x_in_filt), rownames(y_in)), ]
# 
# set.seed(123)
# 
# binom_model <- 
#   cv.glmnet(x = x_in_filt,
#             y = y,
#             family = 'binomial',
#             type.measure = 'deviance',
#             nfolds = nrow(x_in_filt), 
#             alpha = 0.9, 
#             parallel = TRUE,
#             keep = TRUE)
# 
# 
# coef_tmp <- cbind(coef(binom_model, s = binom_model$lambda.min))
# 
# beta <- coef_tmp[apply(coef_tmp != 0, 1, any), ]
# 
# colnames(beta) <- names(coef(binom_model, s = binom_model$lambda.min))
# 
# y_hat_matrix <- binom_model$fit.preval[,which(binom_model$lambda == binom_model$lambda.min)]
# 
# y_hat_predictions <- predict(object = binom_model, s = 'lambda.min', 
#                              newx = x_in_filt, newy = y_in, type = 'response')
# 
# binom_roc_out <- pROC::roc(response = y, predictor = y_hat_matrix)
# 
# perf_list <- sens_at_spec_ROC(binom_roc_out, 1.0)
# 
# y_hat_predictions <- ifelse(y_hat_predictions > perf_list$p.thresh, 1, 0)
# 
# pROC::ggroc(binom_roc_out) -> binom_roc.plt
# 
# binomial_model_response_prob <- inverse.logit.calc(y_hat_matrix)
# 
# binomial_model_response_prob %>%
#   as.data.frame() %>%
#   rownames_to_column('sample') %>%
#   dplyr::rename('prob' = '.') %>%
#   merge(y_in %>% rownames_to_column('sample'), by = 'sample') %>%
#   mutate(condition = ifelse(condition == 1, 'panc', 'ctrl')) %>%
#   ggplot(aes(condition, prob)) +
#   geom_jitter(width = 0.125) +
#   geom_boxplot(outlier.colour = NA, fill = NA) + 
#   geom_hline(yintercept = perf_list$p.thresh, linetype = 'dashed') + 
#   annotate(geom = 'text', y = perf_list$p.thresh + 0.05, x = 1.5, 
#            label = '100% specificity prob. threshold', size = txt.mm_to_pts(0.8)) +
#   ylab('probability') + 
#   xlab('') -> binom_prob_thresh.plt
# 
# beta %>%
#   as.matrix() %>%
#   as.data.frame() %>%
#   rownames_to_column('gene') %>%
#   merge(ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out, by = 'gene') %>%
#   ggplot(aes(V1, log2FoldChange, label = gene)) +
#   geom_point() +
#   ggrepel::geom_text_repel() 
# 
# ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out %>% 
#   filter(ensg %in% names(beta)) %>% 
#   merge(reference_meta_data$ucsc_rmsk_insert_info.df, by = 'gene', all.x = T) %>% 
#   mutate(family = ifelse(ensg %in% reference_meta_data$gencode_lncRNA_gene_names.df$ensg,
#                          'gencode_lncRNA', family),
#          family = ifelse(is.na(family), 'gencode_coding', family),
#          clade = ifelse(grepl('gencode', family), family, clade),
#          clade = ifelse(clade %in% c('gencode_coding',
#                                      'gencode_lncRNA',
#                                      'LINE',
#                                      'SINE',
#                                      'LTR',
#                                      'DNA',
#                                      'Simple_repeat'), clade, 'other'),
#          type = ifelse(grepl('gencode',family), 'GENCODE', 'TE'),
#          clade = factor(clade, levels = te_aware_annotation_levels)) %>% 
#   ggplot(aes(log2FoldChange, -log10(padj), color = clade, label = gene)) + 
#   geom_point() + 
#   scale_color_manual(values = te_color_pal) +
#   ggrepel::geom_text_repel(color = 'black', segment.colour = NA, size = txt.mm_to_pts(0.8)) -> feature_de.plt
# 
# caret::confusionMatrix(data = as.factor(ifelse(as.factor(y_hat_predictions) == 1, 'panc', 'ctrl')),
#                        reference = as.factor(ifelse(as.factor(y) == 1, 'panc', 'ctrl')))$table %>% 
#   gridExtra::tableGrob(theme = gridExtra::ttheme_minimal(base_size = 8,
#                                                          core = list(padding=unit(c(1, 2.5), "mm")))) ->
#   binomial_cm.tbl
```
### fig_modeling_patchwork
```{r}
import::from(.from = 'patchwork', plot_annotation, wrap_plots, guide_area, plot_spacer, wrap_elements, inset_element)

fig_surv_layout <- "
ABC
KKK
DDE
FGH
IIJ
"

# wrap_plots(wrap_elements(full = binom_roc.plt + inset_element(binomial_cm.tbl, right = 0.5, 
#                                          bottom = 0.2, left = 0.6, top = 0.5, ignore_tag = T)),
#            binom_prob_thresh.plt,
#            feature_de.plt,
#            wrap_elements(full = multi_roc.plt + inset_element(multinomial_cm.tbl, right = 0.5, 
#                                          bottom = 0.2, left = 0.6, top = 0.5, ignore_tag = T)),
#            multi_feature_de.plt,
#            wrap_elements(full = gt_binom_roc.plt + inset_element(gt_binomial_cm.tbl, right = 0.5, 
#                                          bottom = 0.2, left = 0.6, top = 0.5, ignore_tag = T)),
#            gt_binom_prob_thresh.plt,
#            gt_feature_de.plt,
#            wrap_elements(full = gt_multi_roc.plt + inset_element(gt_multinomial_cm.tbl, right = 0.5, 
#                                          bottom = 0.2, left = 0.6, top = 0.5, ignore_tag = T)),
#            gt_multi_feature_de.plt,           
#            guide_area(),
#            heights = c(1,0.01,1,1,1),
#            guides = 'collect',
#            design = fig_surv_layout) +
#   plot_annotation(tag_levels = 'A') &
#   theme(plot.tag.position = c(0.01, 0.99),
#         legend.position = 'bottom', 
#         legend.justification = c(0.5, 0.5)) -> fig.model.patchwork

fig_surv_layout <- "
AABB
EEEE
CCDD
FFFF
"

wrap_plots(wrap_elements(full = gt_binom_roc.plt + inset_element(gt_binomial_cm.tbl, right = 0.5, 
                                         bottom = 0.2, left = 0.6, top = 0.5, ignore_tag = T)),
           gt_feature_de.plt + annotate(geom = 'text', label = 'Repeat-aware', x = -2, y = 7),
           wrap_elements(full = binom_roc.plt + inset_element(binomial_cm.tbl, right = 0.5, 
                                         bottom = 0.2, left = 0.6, top = 0.5, ignore_tag = T)),
           feature_de.plt + annotate(geom = 'text', label = 'Repeat-alone', x = -1, y = 3.5),
           guide_area(),
           plot_spacer(),
           heights = c(1,0.2,1,1,1),
           guides = 'collect',
           design = fig_surv_layout) +
  plot_annotation(tag_levels = 'A') &
  theme(plot.tag.position = c(0.01, 0.99),
        legend.position = 'bottom', 
        legend.justification = c(0.5, 0.5)) -> fig.model.patchwork

save.manuscript.panel(figure = 7,
                      figure.dir = figure.dir,
                      name = 'patchwork_full_figS7_model',
                      width = figure_double_column_width,
                      height = figure_depth,
                      plot.in = fig.model.patchwork)
```


## TCGA GTEx Heatmap
```{r}
tcga_paad_gene_counts %>% 
  merge(gtex_paad_gene_counts, by = 'gene') %>%
  # filter(gene %in% reference_meta_data$ucsc_rmsk_insert_info.df$gene) %>%
  # filter(gene %in% filter(ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out,
  #                         padj < 0.05 & abs(log2FoldChange) > 0.75)$gene) %>%
  # filter(!if_all(is.numeric, ~.x == 0)) %>% nrow()
  # mutate(across(where(is.numeric), ~scale(log2(. + 1), center = T)[,1])) %>%
  # mutate(across(where(is.numeric), ~scale(., center = T)[,1])) %>%
  # filter(!gene %in% c('AluYm1', 'AluJo')) %>% 
  # mutate(across(where(is.numeric), ~log2(. + 1))) %>%
  column_to_rownames('gene') %>% as.matrix() %>% t() -> hm_count_data

tcga_identity_color_pal <- identity_color_pal[1:2]
names(tcga_identity_color_pal) <- c('GTEx', 'PAAD')

hm_meta_data <- tibble('sample' = rownames(hm_count_data), 'condition' = ifelse(grepl('PAAD', sample), 'PAAD', 'GTEx'))

t(hm_count_data) %>% 
  as.data.frame() %>%
  mutate(across(where(is.numeric), ~round(.x))) -> count_data_for_DEseq

tcga_de_dds <- DESeq2::DESeqDataSetFromMatrix(countData = count_data_for_DEseq, colData = hm_meta_data, design = as.formula('~1'), tidy = F)

tcga_de_dds <- DESeq2::estimateSizeFactors(tcga_de_dds)

tcga_de_dds_norm_counts.df <- as.data.frame(DESeq2::counts(tcga_de_dds, normalized=T))






hm_meta_data %>% pull(condition) -> hm_meta_condition
names(hm_meta_condition) <- hm_meta_data$sample

left_hma <- ComplexHeatmap::rowAnnotation(df = data.frame('condition' = hm_meta_condition), 
                                          col = list(condition = tcga_identity_color_pal),
                                          show_annotation_name = F, simple_anno_size = unit(.3, "cm"))

# hm_te_data <- data.frame('gene' = colnames(hm_count_data)) %>%
#   merge(reference_meta_data$total_tx_to_gene.df %>% select(gene, biotype) %>% distinct(), by = 'gene') %>%
#   mutate(biotype = ifelse(biotype %in% c('SINE', 'Simple_repeat'), biotype, 'gencode')) %>%
#   distinct()

hm_te_data <- data.frame('gene' = colnames(hm_count_data)) %>%
  merge(reference_meta_data$ucsc_rmsk_insert_info.df %>% select(gene, clade) %>% distinct(), by = 'gene') %>%
  distinct()

top_hma <- ComplexHeatmap::columnAnnotation(df = hm_te_data %>% select(clade), 
                                            show_annotation_name = T,
                                            annotation_name_side = 'left',
                                            annotation_name_gp = grid::gpar(fontsize = 10, fontface = 'italic'),
                                            show_legend = T, 
                                            simple_anno_size = unit(.375, "cm"))


set.seed(129)
ComplexHeatmap::Heatmap(hm_count_data, 
                        show_row_names = F, 
                        show_column_names = T, 
                        heatmap_width = figure_double_column_width * 0.85,
                        left_annotation = left_hma,
                        column_dend_height = unit(.75, "cm"),
                        heatmap_legend_param = 
                          list(legend_gp = grid::gpar(fontsize = 8)),
                        # col = circlize::colorRamp2(breaks = c(-4,-2,0,2,4),
                        #                            colors = viridis::rocket(n = 5)),
                        top_annotation = top_hma,
                        column_names_gp = grid::gpar(fontsize = 8),
                        name = 'Z-score') -> tcga_hm

new_hm <- grid::grid.grabExpr(ComplexHeatmap::draw(tcga_hm), merge_legend = T)

patchwork::wrap_plots(full = new_hm)


hm_count_data %>% 
  as.data.frame() %>% 
  # select(contains('Alu')) %>% 
  rownames_to_column('sample') %>% 
  # merge(hm_meta_data, by = 'sample') %>% 
  merge(tcga_colData %>% select(submitter_id.samples, tumor_stage.diagnoses), 
        by.x = 'sample', by.y = 'submitter_id.samples') %>%
  gather(gene, tpm, -sample, -tumor_stage.diagnoses) %>%
  ggplot(aes(tumor_stage.diagnoses, tpm)) + 
  geom_boxplot(outlier.colour = NA, fill = NA) + 
  # geom_jitter(alpha = 0.4) + 
  ggforce::facet_row(~gene) 
  # ggpubr::stat_compare_means(method = 'wilcox', comparisons = list(c('GTEx', 'PAAD')))
```
## tcga matched DESeq
```{r}
matched_samples <- c(colnames(count_data_for_DEseq[,grepl('-NT', colnames(count_data_for_DEseq))]),
                     str_replace_all(colnames(count_data_for_DEseq[,grepl('-NT', colnames(count_data_for_DEseq))]), '-NT', '-TP'))

matched_samples_counts <- count_data_for_DEseq[, colnames(count_data_for_DEseq) %in% matched_samples]

tcga_de_colData <- tibble('sample' = colnames(matched_samples_counts), 'condition' = ifelse(grepl('NT', sample), 'control', 'tumor'))

tcga_de_dds <- DESeq2::DESeqDataSetFromMatrix(countData = matched_samples_counts, colData = tcga_de_colData, 
                                              design = as.formula('~condition'), tidy = F)

tcga_de_dds <- DESeq2::estimateSizeFactors(tcga_de_dds)

de_filter <- rowSums(DESeq2::counts(tcga_de_dds, normalized=T) >= 10) >= ncol(DESeq2::counts(tcga_de_dds, normalized=T))*0.75
  
input_set_dds <- DESeq2::DESeq(tcga_de_dds[de_filter,])

DESeq2::resultsNames(input_set_dds)

input_set_dds_de_out <- DESeq2::lfcShrink(input_set_dds, coef = "condition_tumor_vs_control") %>% 
  as.data.frame() %>% 
  rownames_to_column('gene') %>% 
  filter(padj < 0.05)


input_set_dds_de_out %>% 
  merge(ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out, by = 'gene')
```
## tcga_modeling
```{r}
library(glmnet)
library(pROC)



model_input_data <- t(tcga_de_dds_norm_counts.df)[!grepl('-TM$|-NT$', rownames(hm_count_data)), ]
model_input_meta_data <- hm_meta_data[match(rownames(model_input_data), hm_meta_data$sample), ]

set.seed(123)

doMC::registerDoMC(cores =12)

trainSplitIndex <- 
  caret::createDataPartition(factor(model_input_meta_data$condition, levels = c('GTEx', 'PAAD')), p = .8, list = FALSE, times = 1)

tcga_x_in <- hm_count_data

tcga_y_in <- hm_meta_data %>% mutate(condition = ifelse(condition == 'PAAD', 1, 0))

tcga_x_in_train <- tcga_x_in[trainSplitIndex, ,drop = F]
tcga_y_in_train <- tcga_y_in[trainSplitIndex, ,drop = F] %>% pull(condition)

tcga_x_in_test <- tcga_x_in[-trainSplitIndex, ,drop = F]
tcga_y_in_test <- tcga_y_in[-trainSplitIndex, ,drop = F] %>% pull(condition)

set.seed(123)

tcga_binomial_model <- 
  cv.glmnet(x = tcga_x_in_train,
            y = tcga_y_in_train,
            family = 'binomial',
            type.measure = 'deviance',
            nfolds = 20, 
            alpha = 0.75, 
            parallel = TRUE,
            keep = TRUE)

plot(tcga_binomial_model)

coef_tmp <- coef(tcga_binomial_model, s = tcga_binomial_model$lambda.min)

beta <- coef_tmp[apply(coef_tmp != 0, 1, any), ]

beta

inverse.logit.calc <- function(y_hat) { exp(y_hat)/(1+exp(y_hat)) }

y_hat_matrix <- tcga_binomial_model$fit.preval[,which(tcga_binomial_model$lambda == tcga_binomial_model$lambda.min)]

# predict(tcga_binomial_model, s = 'lambda.min', newx = tcga_x_in_train, type = 'response')

training_pred_matrix <- predict(tcga_binomial_model, s = 'lambda.min', newx = tcga_x_in_train, type = 'response')

pROC::roc(response = tcga_y_in_train, predictor = training_pred_matrix[,1])$auc


testing_pred_matrix <- predict(tcga_binomial_model, s = 'lambda.min', newx = tcga_x_in_test, type = 'response')

pROC::roc(response = tcga_y_in_test, predictor = testing_pred_matrix[,1])


testing_pred_matrix <- predict(tcga_binomial_model, s = 'lambda.min', newx = x_in, type = 'response')

pROC::roc(response = tcga_y_in_test, predictor = testing_pred_matrix[,1])


tcga_binomial_model_response_prob <- inverse.logit.calc(y_hat_matrix)


tcga_binomial_model_response_prob %>% 
  as.data.frame() %>% 
  rownames_to_column('sample') %>% 
  dplyr::rename('prob' = '.') %>% 
  merge(tcga_y_in, by = 'sample') %>% 
  mutate(condition = ifelse(condition == 1, 'PAAD', 'GTEx')) %>% 
  ggplot(aes(condition, prob)) + 
  geom_boxplot()


```
## tcga_pca
```{r}
pca_tmp_in_sd <- apply(x_in_train, 2, sd)
pca_tmp_in_zero_sd <- x_in_train[,pca_tmp_in_sd!=0 & !is.na(pca_tmp_in_sd)]
non_zero_pca <- prcomp(pca_tmp_in_zero_sd, center = T, scale. = T, rank. = 50)

summary(non_zero_pca)
pcs.of.interest <- apply(non_zero_pca$x, 2, var)
pcs.props <-  pcs.of.interest/sum(pcs.of.interest)
cumsum(pcs.props)[c(1,2)]
print(as.data.frame(pcs.props) %>% 
        rownames_to_column('pc') %>% 
        mutate(pc = as.numeric(sub('PC', '', pc))) %>% 
        ggplot(aes(pc, pcs.props)) + 
        geom_col())

pca.out <- as.data.frame(non_zero_pca$x) %>% 
  rownames_to_column('sample') %>% 
  merge(hm_meta_data %>% select(sample, condition) %>% ungroup() %>% distinct(), by = 'sample') %>% 
  ungroup() %>% 
  distinct()

pca.out.summary <- 
  as.data.frame(summary(non_zero_pca)$importance) %>% 
  rownames_to_column('metric') %>% 
  gather(pc, value, -metric) %>% 
  spread(metric, value)

pca.out %>% 
  ggplot(aes(PC1, PC2,
             color = condition)) +
  geom_point(size = rel(2), alpha = 1) + 
  # scale_shape_manual(values = c(21,25)) +
  xlab(paste('PC1', round(cumsum(pcs.props)[1], digits = 3), sep = ' ')) +
  ylab(paste('PC2', round(cumsum(pcs.props)[2] - cumsum(pcs.props)[1], digits = 3), sep = ' ')) +
  geom_vline(xintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) +
  geom_hline(yintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) + 
  scale_color_manual(values = tcga_identity_color_pal) +
  theme(legend.position = 'top', legend.direction = 'horizontal')


pca.out %>% 
  merge(tcga_colData, by.x = 'sample', by.y = 'submitter_id.samples') %>% 
    ggplot(aes(PC1, PC2,
             color = tumor_stage.diagnoses)) +
  geom_point(size = rel(2), alpha = 1) + 
  # scale_shape_manual(values = c(21,25)) +
  xlab(paste('PC1', round(cumsum(pcs.props)[1], digits = 3), sep = ' ')) +
  ylab(paste('PC2', round(cumsum(pcs.props)[2] - cumsum(pcs.props)[1], digits = 3), sep = ' ')) +
  geom_vline(xintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) +
  geom_hline(yintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) + 
  # scale_color_manual(values = tcga_identity_color_pal) +
  theme(legend.position = 'top', legend.direction = 'horizontal')
  # extract.meta.pca.correlates()
  
```



## TCGA surv heatmap
```{r}
tcga_paad_dds_norm_counts.df %>% 
  rownames_to_column('gene') %>% 
  filter(gene %in% msig.df$HALLMARK_MYC_TARGETS_V1,
         gene %in% filter(ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out, 
                          padj < 0.05, log2FoldChange > 0.75)$gene) %>% 
  column_to_rownames('gene') %>% t() -> paad_hm_count_data


hm_meta_data %>% 
  filter(sample %in% rownames(paad_hm_count_data)) %>% 
  pull(condition) -> hm_meta_condition

names(hm_meta_condition) <- hm_meta_data$sample

left_hma <- ComplexHeatmap::rowAnnotation(df = data.frame('condition' = hm_meta_condition), 
                                          col = list(condition = tcga_identity_color_pal),
                                          show_annotation_name = F, simple_anno_size = unit(.3, "cm"))

# hm_te_data <- data.frame('gene' = colnames(hm_count_data)) %>%
#   merge(reference_meta_data$total_tx_to_gene.df %>% select(gene, biotype) %>% distinct(), by = 'gene') %>%
#   mutate(biotype = ifelse(biotype %in% c('SINE', 'Simple_repeat'), biotype, 'gencode')) %>%
#   distinct()

paad_hm_count_data_filt <- data.frame('gene' = colnames(paad_hm_count_data)) %>% 
  filter(gene %in% msig.df$HALLMARK_MYC_TARGETS_V1,
         gene %in% filter(ucsc.rmsk.salmon_quant$analysis_set$panc_ctrl$deseq$de_out, 
                          padj < 0.05, log2FoldChange > 0.75)$gene) %>%
  distinct()

top_hma <- ComplexHeatmap::columnAnnotation(df = hm_te_data %>% select(clade), 
                                            show_annotation_name = T,
                                            annotation_name_side = 'left',
                                            annotation_name_gp = grid::gpar(fontsize = 10, 
                                                                            fontface = 'italic'),
                                            show_legend = T, 
                                            simple_anno_size = unit(.375, "cm"))


set.seed(129)
ComplexHeatmap::Heatmap(paad_hm_count_data, 
                        show_row_names = F, 
                        show_column_names = T, 
                        heatmap_width = figure_double_column_width * 0.85,
                        # left_annotation = left_hma,
                        column_dend_height = unit(.75, "cm"),
                        heatmap_legend_param = 
                          list(legend_gp = grid::gpar(fontsize = 8)),
                        # col = circlize::colorRamp2(breaks = c(-4,-2,0,2,4),
                        #                            colors = viridis::rocket(n = 5)),
                        # top_annotation = top_hma,
                        column_names_gp = grid::gpar(fontsize = 8),
                        name = 'Z-score') -> paad_tcga_hm

new_hm <- grid::grid.grabExpr(ComplexHeatmap::draw(paad_tcga_hm), merge_legend = T)

patchwork::wrap_plots(full = new_hm)
```




